{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import argparse\n",
    "import time\n",
    "import def_function\n",
    "from def_class import *\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "__version__ = \"V4.1(Editor) 2023-08-07\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数-debug\n",
    "ATSS_APA_cutoff = 3\n",
    "absolute_path = False\n",
    "file_path = \"F:/OneDrive/Master/Project/trans/data/\"\n",
    "input_sample_info_filename = \"0000_sample_info.tsv\"\n",
    "input_pkl_filename = \"0001_total_info.pickle\"\n",
    "output_df_filename = \"0002_TSS_TES.tsv\"\n",
    "output_check_df_filename = \"0002_result_check.tsv\"\n",
    "output_pkl_filename = \"0002_total_info.pkl\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"--ATSS_APA_cutoff\", dest=\"ATSS_APA_cutoff\", required=False, type=int, default=3, help=\"=3,\\t the cutoff value of isoform of ATSS-APA gene\")\n",
    "parser.add_argument(\"--absolute_path\", dest=\"absolute_path\", action=\"store_true\", help=\"use absolute path\")\n",
    "parser.add_argument(\"--file_path\", dest=\"file_path\", required=False, type=str, default=\"\", help=\"the path of data directory\")\n",
    "parser.add_argument(\"--input_sample_info_filename\", dest=\"input_sample_info_filename\", required=False, type=str, default=\"0000_sample_info.tsv\", help=\"=0000_sample_info.tsv,\\t the file of metadata\")\n",
    "parser.add_argument(\"--input_pkl_filename\", dest=\"input_pkl_filename\", required=False, type=str, default=\"0001_total_info.pickle\", help=\"=0001_total_info.pickle,\\t the output file of 0001.py\")\n",
    "parser.add_argument(\"--output_df_filename\", dest=\"output_df_filename\", required=False, type=str, default=\"0002_TSS_TES.tsv\", help=\"=0002_TSS_TES.tsv,\\t the output file of this script\")\n",
    "parser.add_argument(\"--output_check_df_filename\", dest=\"output_check_df_filename\", required=False, type=str, default=\"0002_result_check.tsv\", help=\"=0002_result_check.tsv,\\t the output file of this script\")\n",
    "parser.add_argument(\"--output_pkl_filename\", dest=\"output_pkl_filename\", required=False, type=str, default=\"0002_total_info.pkl\", help=\"=0002_total_info.pkl,\\t the output file of this file\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "ATSS_APA_cutoff = args.ATSS_APA_cutoff\n",
    "absolute_path =args.absolute_path\n",
    "file_path = args.file_path\n",
    "input_sample_info_filename = args.input_sample_info_filename\n",
    "input_pkl_filename = args.input_pkl_filename\n",
    "output_df_filename = args.output_df_filename\n",
    "output_check_df_filename = args.output_check_df_filename\n",
    "output_pkl_filename = args.output_pkl_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补全路径\n",
    "if absolute_path is False:\n",
    "    input_sample_info_filename = \"{}{}\".format(file_path, input_sample_info_filename)\n",
    "    input_pkl_filename = \"{}{}\".format(file_path, input_pkl_filename)\n",
    "    output_df_filename = \"{}{}\".format(file_path, output_df_filename)\n",
    "    output_check_df_filename = \"{}{}\".format(file_path, output_check_df_filename)\n",
    "    output_pkl_filename = \"{}{}\".format(file_path, output_pkl_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 打印参数\n",
    "print('\\n')\n",
    "print(\"[Date]{}\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))\n",
    "print(\"[script]{}\".format(__file__))\n",
    "print(\"[Version]{}\".format(__version__))\n",
    "print(\"[Version]def_function: {}\".format(def_function.__version__))\n",
    "print(\"[Version]def_class: {}\".format(class_version))\n",
    "print(\"[Parament]ATSS_APA_cutoff: {}\".format(ATSS_APA_cutoff))\n",
    "print(\"[Parament]input_sample_info_filename: {}\".format(input_sample_info_filename))\n",
    "print(\"[Parament]input_pkl_filename: {}\".format(input_pkl_filename))\n",
    "print(\"[Parament]output_df_filename: {}\".format(output_df_filename))\n",
    "print(\"[Parament]output_check_df_filename: {}\".format(output_check_df_filename))\n",
    "print(\"[Parament]output_pkl_filename: {}\".format(output_pkl_filename))\n",
    "print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "存在一个问题：\n",
    "    在统计ATSS-APA类型基因时，需要isoform类型>3\n",
    "    但原0002.py在统计ATSS-APA类型基因时，实际筛选条件未isoform类型>2\n",
    "    因此需要重构0002.py\n",
    "\n",
    "思路：\n",
    "    1.重构基因分类方法，应该向gene对象添加相应的基因类型属性，而非根据df进行分类\n",
    "    2.确定原0002.py都做了什么，尽量维持输出数据格式一致\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "@def_function.log\n",
    "def loadSampleInfo(filename):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        filename,\\t str,\\t file location\\n\n",
    "    change:\n",
    "        读取样本的metadata信息, 获取细胞系与GEO_accession的联系\n",
    "    output:\n",
    "        dict,\\t {<cellLine1>: [<GSM1>, <GSM2>, ...],\n",
    "                 <cellLine2>: ...}\n",
    "    \"\"\"\n",
    "    filename = filename\n",
    "\n",
    "    sample_info = pandas.read_csv(filename, sep='\\t')\n",
    "    sample_info_group = sample_info.groupby(by=\"cell_line\")\n",
    "    sample_info = {}\n",
    "    for cellLine in sample_info_group.groups.keys():\n",
    "        sample_info[cellLine] = sample_info_group.get_group(cellLine)[\"GEO_accession\"].to_list()\n",
    "\n",
    "    return sample_info\n",
    "\n",
    "\n",
    "def getTranscriptRangeExpressedInCellline(geneObject, sampleList, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        geneObject,\\t Gene object,\\t ...\\n\n",
    "        sampleList,\\t, list, 指定cellLine所包含的的样本名\n",
    "    change:\n",
    "        获取gene中于指定cellLine存在表达的transcript的TSS及TES及transcript的种类数\\n\n",
    "    output:\n",
    "        dict,\\t {\"TSS\": <transcriptStart>, \"TES\": <transcriptEnd>, \"num\": transcriptNum}\\n\n",
    "    \"\"\"\n",
    "    geneObject = geneObject\n",
    "    sampleList = sampleList\n",
    "\n",
    "    transcriptNum = 0  # 记录该gene中有多少种transcript在指定cellLine中存在表达\n",
    "    transcriptStart = set()  # 存储在cellLine中存在表达的TSS\n",
    "    transcriptEnd = set()  # 存储在cellLine中存在表达的TES\n",
    "    # 遍历每一个transcript\n",
    "    for transcriptObject in geneObject.transcript_dict.values():\n",
    "        # 确定transcript在cellLine中是否表达\n",
    "        transcriptExpression = [transcriptObject[\"countsExpression\"].get(sample, 0) for sample in sampleList]\n",
    "        transcriptExpression = map(lambda x: True if x!=0 else False, transcriptExpression)\n",
    "\n",
    "        if True not in transcriptExpression:\n",
    "            # 该transcript在cellLine中不表达\n",
    "            continue\n",
    "        else:\n",
    "            # 该transcript在cellLine中表达\n",
    "            transcriptNum += 1\n",
    "            if geneObject.strand == '-':\n",
    "                transcriptStart.add(transcriptObject.get(\"range\")[1])\n",
    "                transcriptEnd.add(transcriptObject.get(\"range\")[0])\n",
    "            else:\n",
    "                transcriptStart.add(transcriptObject.get(\"range\")[0])\n",
    "                transcriptEnd.add(transcriptObject.get(\"range\")[1])\n",
    "            # 一旦确定该transcript在cellLine中存在表达就记录该transcript的信息，并继续处理下一个transcript\n",
    "            continue\n",
    "\n",
    "    return {\"TSS\": transcriptStart, \"TES\": transcriptEnd, \"num\": transcriptNum}\n",
    "\n",
    "\n",
    "def get_check_df(total, sampleInfo, cellLine, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        total,\\t Total,\\t ...\\n\n",
    "        sampleInfo,\\t pandas.DataFrame,\\t ...\\n\n",
    "        cellLine,\\t, str,\\t ...\\n\n",
    "    change:\n",
    "        生成df, 以gene_id为index, 列名为[\"strand\", \"gene_biotype\", \"start_counts\", \"end_counts\", \"transcript_counts\", \"cell_line\", \"gene_classified\"]\n",
    "    output:\n",
    "        check_df,\\t pandas.DataFrame,\\t ...\\n\n",
    "    \"\"\"\n",
    "    total = total\n",
    "    sampleInfo = sampleInfo\n",
    "    cellLine = cellLine\n",
    "\n",
    "    check_df = {}\n",
    "    if cellLine == \"all\":\n",
    "        # 遍历每一个gene对象\n",
    "        for gene_id, gene_object in total.gene_dict.items():\n",
    "            check_df[gene_id] = [gene_object.strand,\n",
    "                                gene_object.gene_biotype,\n",
    "                                len(TSS_TES_dict.get(gene_id).get(\"range_start\")),\n",
    "                                len(TSS_TES_dict.get(gene_id).get(\"range_end\")),\n",
    "                                len(gene_object.transcript_dict),\n",
    "                                \"all\",\n",
    "                                gene_object.gene_classified\n",
    "                                ]\n",
    "        check_df = pandas.DataFrame.from_dict(check_df, orient=\"index\",\n",
    "                                            columns=[\"strand\", \"gene_biotype\", \"start_counts\", \"end_counts\",\n",
    "                                                    \"transcript_counts\", \"cell_line\", \"gene_classified\"])\n",
    "    else:\n",
    "        # 若cellLine不属于all的话, 则遍历所有gene\n",
    "        # 对gene中的每一个transcript, 检查其在指定cellLine样本中是否表达\n",
    "        #   若表达，则记录其TSS与TES, 若不存在, 则认为在指定cellLine中不表达该transcript\n",
    "        for geneId, geneObject in total.gene_dict.items():\n",
    "            # 获取gene在cellLine中表达的情况\n",
    "            transcriptInfo = getTranscriptRangeExpressedInCellline(geneObject=geneObject,\n",
    "                                                                   sampleList=sampleInfo.get(cellLine))\n",
    "            if transcriptInfo.get(\"num\") == 0:\n",
    "                # 若该gene没有于cellLine中表达的transcript, 则跳过该gene\n",
    "                continue\n",
    "            startNum = len(transcriptInfo.get(\"TSS\"))\n",
    "            endNum = len(transcriptInfo.get(\"TES\"))\n",
    "            gene_classified = \"{}-{}\".format((\"ATSS\", \"TSS\")[startNum==1],\n",
    "                                            (\"APA\", \"PAS\")[endNum==1]\n",
    "                                            )\n",
    "            check_df[geneId] = [geneObject.strand,\n",
    "                                geneObject.gene_biotype,\n",
    "                                startNum,\n",
    "                                endNum,\n",
    "                                transcriptInfo.get(\"num\"),\n",
    "                                cellLine,\n",
    "                                gene_classified]\n",
    "        check_df = pandas.DataFrame.from_dict(check_df, orient=\"index\",\n",
    "                                            columns=[\"strand\", \"gene_biotype\", \"start_counts\", \"end_counts\",\n",
    "                                                    \"transcript_counts\", \"cell_line\", \"gene_classified\"])\n",
    "\n",
    "    return check_df\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def count_TSS_TES(check_df, ATSS_APA_cutoff, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        check_df,\\t pandas.DataFrame,\\t ...\\n\n",
    "        ATSS_APA_cutoff, int, the cutoff value of isoform of ATSS-APA gene\n",
    "    change:\n",
    "        根据gene_biotype, gene_classified对四种类型gene的transcript的种类数进行统计\n",
    "    output:\n",
    "        df,\\t pandas.DataFrame,\\t columns=[\"cellLine\", \"geneOrTranscript\", \"geneBiotype\", \"TSS-PAS\", \"TSS-APA\", \"ATSS-PAS\", \"ATSS-APA\"]\\n\n",
    "    \"\"\"\n",
    "    check_df = check_df\n",
    "    ATSS_APA_cutoff = ATSS_APA_cutoff\n",
    "\n",
    "    df = []\n",
    "\n",
    "    df_group = check_df.groupby(\"cell_line\")\n",
    "    for cellLine in df_group.groups.keys():\n",
    "        temp_df = df_group.get_group(cellLine)\n",
    "        # 删除ATSS-APA中低于指定isoform类型数的gene\n",
    "        # 也就是说, ATSS-APA型gene需要具备大于等于ATSS_APA_cutoff种transcript才会被纳入统计\n",
    "        readyToDelIndex = temp_df.query(\"gene_classified=='ATSS-APA' & transcript_counts<@ATSS_APA_cutoff\").index\n",
    "        temp_df = temp_df.drop(index=readyToDelIndex)\n",
    "        # 将gene_biotype归类为三种类型: protein_coding, non_protein_coding, un_classified\n",
    "        temp_df[\"gene_biotype\"] = temp_df[\"gene_biotype\"].map(lambda x: (x, \"non_protein_coding\")[x not in [\"protein_coding\", \"un_classified\"]])\n",
    "        temp_df_group = temp_df.groupby(\"gene_biotype\")\n",
    "        # 按gene_biotype对gene进行分类统计\n",
    "        for biotype in temp_df_group.groups.keys():\n",
    "            temp_df2 = temp_df_group.get_group(biotype)\n",
    "            temp = temp_df2[\"gene_classified\"].value_counts()\n",
    "            df.append([cellLine, \"gene\", biotype,\n",
    "                    temp[\"TSS-PAS\"], temp[\"TSS-APA\"],\n",
    "                    temp[\"ATSS-PAS\"], temp[\"ATSS-APA\"]])\n",
    "\n",
    "            # 统计transcript的数量\n",
    "            temp = {}  # {\"TSS-PAS\": <int>, \"TSS-APA\": <int>,\"ATSS-PAS\": <int>, \"ATSS-APA\": <int>}\n",
    "            temp_group = temp_df2.groupby(\"gene_classified\")\n",
    "            for gene_classified in temp_group.groups.keys():\n",
    "                # 统计不同gene_classified类型gene中所具有的transcript的种类数\n",
    "                temp[gene_classified] = temp_group.get_group(gene_classified)[\"transcript_counts\"].sum()\n",
    "            df.append([cellLine, \"transcript\", biotype,\n",
    "                    temp[\"TSS-PAS\"], temp[\"TSS-APA\"],\n",
    "                    temp[\"ATSS-PAS\"], temp[\"ATSS-APA\"]])\n",
    "    \n",
    "    # 整理df格式\n",
    "    df = pandas.DataFrame(df, columns=[\"cellLine\", \"geneOrTranscript\", \"geneBiotype\", \"TSS-PAS\", \"TSS-APA\", \"ATSS-PAS\", \"ATSS-APA\"])\n",
    "    df = df.sort_values([\"cellLine\", \"geneOrTranscript\", \"geneBiotype\"], ignore_index=True, ascending=False)\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "[Function]loadSampleInfo start.\n",
      "\t[Time]2023-08-17 10:39:11\n",
      "\t[Paraments]filename: F:/OneDrive/Master/Project/trans/data/0000_sample_info.tsv\n",
      "[2023-08-17 10:39:11]loadSampleInfo finished.\n",
      "\n",
      "\n",
      "[Function]count_TSS_TES start.\n",
      "\t[Time]2023-08-17 10:39:12\n",
      "\t[Paraments]check_df: <...>\n",
      "\t[Paraments]ATSS_APA_cutoff: 3\n",
      "[2023-08-17 10:39:12]count_TSS_TES finished.\n",
      "[2023-08-17 10:39:21]All blocks finished.\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # load data\n",
    "    with open(input_pkl_filename, 'rb') as file:\n",
    "        total = pickle.load(file)\n",
    "    sampleInfo = loadSampleInfo(filename=input_sample_info_filename)\n",
    "\n",
    "    # 向所有gene添加gene_classified属性(TSS-PAS, TSS-APA, ATSS-PAS, ATSS-APA)\n",
    "    TSS_TES_dict = total.add_gene_classified()\n",
    "\n",
    "    # 整理得到check_df\n",
    "    check_df = get_check_df(total=total, sampleInfo=sampleInfo, cellLine=\"all\")\n",
    "    check_df = check_df.reset_index()\n",
    "    check_df = check_df.rename(columns={\"index\": \"gene_id\"})\n",
    "    for cellLine in sampleInfo.keys():\n",
    "        temp_check_df = get_check_df(total=total, sampleInfo=sampleInfo, cellLine=cellLine)\n",
    "        temp_check_df = temp_check_df.reset_index()\n",
    "        temp_check_df = temp_check_df.rename(columns={\"index\": \"gene_id\"})\n",
    "        check_df = pandas.concat([check_df, temp_check_df], axis=0, ignore_index=True)\n",
    "    \n",
    "    # 获得gene及transcript的数目统计表\n",
    "    df = count_TSS_TES(check_df=check_df, ATSS_APA_cutoff=ATSS_APA_cutoff)\n",
    "\n",
    "    # save data\n",
    "    df.to_csv(output_df_filename, sep='\\t', index=None)\n",
    "    check_df.to_csv(output_check_df_filename, sep='\\t', index=None)\n",
    "    with open(output_pkl_filename, 'wb') as file:\n",
    "        pickle.dump(total, file)\n",
    "    \n",
    "    print(\"[{}]All blocks finished.\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "调试"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "调试\n",
    "\n",
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
