{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import time\n",
    "import argparse\n",
    "import def_function\n",
    "from def_class import *\n",
    "import pickle\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "import math\n",
    "import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "__version__ = \"V3.0(Editor) 2023-08-11\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数-debug\n",
    "ATSS_APA_cutoff = 3\n",
    "log2FC_cutoff = 1.5\n",
    "plot_x_maxnum = 5\n",
    "check_counts = False\n",
    "absolute_path = False\n",
    "file_path = \"F:/OneDrive/Master/Project/trans/data/\"\n",
    "input_sample_info_filename = \"0000_sample_info.tsv\"\n",
    "input_df_filename = \"0001_total_info.tsv\"\n",
    "input_pkl_filename = \"0002_total_info.pkl\"\n",
    "output_length_counts_plot_filename = \"0004_length_counts.png\"\n",
    "output_check_df_filename = \"0004_check_df.tsv\"\n",
    "output_transcript_gene_plot_path = \"0004_plot/\"\n",
    "output_relative_expression_filename = \"0004_relative_expression.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"--ATSS_APA_cutoff\", dest=\"ATSS_APA_cutoff\", required=False, type=int, default=3, help=\"=3,\\t the cutoff value of isoform of ATSS-APA gene\")\n",
    "parser.add_argument(\"--log2FC_cutoff\", dest=\"log2FC_cutoff\", required=False, type=float, default=1.5, help=\"=1.5,\\t the cutoff value of log2FC\")\n",
    "parser.add_argument(\"--plot_x_maxnum\", dest=\"plot_x_maxnum\", required=False, type=int, default=5, help=\"=5,\\t the maxnum in xaxis\")\n",
    "parser.add_argument(\"--check_counts\", dest=\"check_counts\", required=False, action=\"store_true\", help=\"output the plot about the relation of length and rep value of transcript\")\n",
    "parser.add_argument(\"--absolute_path\", dest=\"absolute_path\", required=False, action=\"store_true\", help=\"use the absolute path\")\n",
    "parser.add_argument(\"--file_path\", dest=\"file_path\", required=False, type=str, default=\"./\", help=\"=the data directory path\")\n",
    "parser.add_argument(\"--input_sample_info_filename\", dest=\"input_sample_info_filename\", required=False, type=str, default=\"0000_sample_info.tsv\", help=\"=0000_sample_info.tsv,\\t the metadata of samples\")\n",
    "parser.add_argument(\"--input_df_filename\", dest=\"input_df_filename\", required=False, type=str, default=\"0001_total_info.tsv\", help=\"=0001_total_info.tsv,\\t the output file of 0001.py\")\n",
    "parser.add_argument(\"--input_pkl_filename\", dest=\"input_pkl_filename\", required=False, type=str, default=\"0002_total_info.pkl\", help=\"=0002_total_info.pkl,\\t the output file of 0002.py\")\n",
    "parser.add_argument(\"--output_length_counts_plot_filename\", dest=\"output_length_counts_plot_filename\", required=False, type=str, default=\"0004_length_counts.png\", help=\"=0004_length_counts.png,\\t to confirm rep is not related to length of transcript\")\n",
    "parser.add_argument(\"--output_check_df_filename\", dest=\"output_check_df_filename\", required=False, type=str, default=\"0004_check_df.xlsx\", help=\"=0004_check_df.xlsx,\\t the output file for checking\")\n",
    "parser.add_argument(\"--output_transcript_gene_plot_path\", dest=\"output_transcript_gene_plot_path\", required=False, type=str, default=\"0004_plot/\", help=\"=0004_plot/,\\t the path of output plot\")\n",
    "parser.add_argument(\"--output_relative_expression_filename\", dest=\"output_relative_expression_filename\", required=False, type=str, default=\"0004_relative_expression.tsv\", help=\"=0004_relative_expression.tsv,\\t the output filename of cell_line expression\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "ATSS_APA_cutoff = args.ATSS_APA_cutoff\n",
    "log2FC_cutoff = args.log2FC_cutoff\n",
    "plot_x_maxnum = args.plot_x_maxnum\n",
    "check_counts = args.check_counts\n",
    "absolute_path = args.absolute_path\n",
    "file_path = args.file_path\n",
    "input_sample_info_filename = args.input_sample_info_filename\n",
    "input_df_filename = args.input_df_filename\n",
    "input_pkl_filename = args.input_pkl_filename\n",
    "output_length_counts_plot_filename = args.output_length_counts_plot_filename\n",
    "output_check_df_filename = args.output_check_df_filename\n",
    "output_transcript_gene_plot_path = args.output_transcript_gene_plot_path\n",
    "output_relative_expression_filename = args.output_relative_expression_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补充路径\n",
    "if absolute_path is False:\n",
    "    input_sample_info_filename = \"{}{}\".format(file_path, input_sample_info_filename)\n",
    "    input_df_filename = \"{}{}\".format(file_path, input_df_filename)\n",
    "    input_pkl_filename = \"{}{}\".format(file_path, input_pkl_filename)\n",
    "    output_length_counts_plot_filename = \"{}other/{}\".format(file_path, output_length_counts_plot_filename)\n",
    "    output_check_df_filename = \"{}{}\".format(file_path, output_check_df_filename)\n",
    "    output_transcript_gene_plot_path = \"{}{}\".format(file_path, output_transcript_gene_plot_path)\n",
    "    output_relative_expression_filename = \"{}{}\".format(file_path, output_relative_expression_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 打印参数\n",
    "print('\\n')\n",
    "print(\"[Date]{}\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))\n",
    "print(\"[Script]{}\".format(__file__))\n",
    "print(\"[Version]{}\".format(__version__))\n",
    "print(\"[Version]def_function: {}\".format(def_function.__version__))\n",
    "print(\"[Version]def_class: {}\".format(class_version))\n",
    "print(\"[Parament]ATSS_APA_cutoff: {}\".format(ATSS_APA_cutoff))\n",
    "print(\"[Parament]log2FC_cutoff: {}\".format(log2FC_cutoff))\n",
    "print(\"[Parament]plot_x_maxnum: {}\".format(plot_x_maxnum))\n",
    "print(\"[Parament]check_counts: {}\".format(check_counts))\n",
    "print(\"[Parament]file_path: {}\".format(file_path))\n",
    "print(\"[Parament]input_sample_info_filename: {}\".format(input_sample_info_filename))\n",
    "print(\"[Parament]input_df_filename: {}\".format(input_df_filename))\n",
    "print(\"[Parament]input_pkl_filename: {}\".format(input_pkl_filename))\n",
    "print(\"[Parament]output_length_counts_plot_filename: {}\".format(output_length_counts_plot_filename))\n",
    "print(\"[Parament]output_check_df_filename: {}\".format(output_check_df_filename))\n",
    "print(\"[Parament]output_transcript_gene_plot_path: {}\".format(output_transcript_gene_plot_path))\n",
    "print(\"[Parament]output_relative_expression_filename: {}\".format(output_relative_expression_filename))\n",
    "print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 隐藏函数\n",
    "def getSignificantTranscriptId(total, cellLineExpression, cellLine1, cellLine2, log2Cutoff=1.5, tab_level=0):\n",
    "    '''\n",
    "    change:\n",
    "        先筛选出在cellLine1或cellLine2中存在表达的transcript\n",
    "        然后再继续筛选出于cellLine1或cellLine2中特异性表达的transcript\n",
    "    '''\n",
    "    total = total\n",
    "    cellLineExpression = cellLineExpression\n",
    "    cellLine1 = cellLine1\n",
    "    cellLine2 = cellLine2\n",
    "    log2Cutoff = log2Cutoff\n",
    "\n",
    "    # 获取在cellLine1及cellLine2中存在表达且富集于其中一种cellLine中的transcriptId\n",
    "    # 只保留在cellLine1及cellLine2中存在表达的transcript\n",
    "    df = cellLineExpression\n",
    "    df[cellLine1] = df[cellLine1].map(lambda x: None if x==0 else x)\n",
    "    df[cellLine2] = df[cellLine2].map(lambda x: None if x==0 else x)\n",
    "    df = df.dropna(axis=0, subset=[cellLine1, cellLine2], how=\"all\")\n",
    "    # 根据log2FC筛选transcript\n",
    "    df[\"log2FC_{}@{}\".format(cellLine1, cellLine2)] = df[\"log2FC_{}@{}\".format(cellLine1, cellLine2)].map(lambda x: x if abs(x)>=log2Cutoff or x==0 else None)\n",
    "    df = df.dropna(axis=0, subset=[\"log2FC_{}@{}\".format(cellLine1, cellLine2)], how=\"any\")\n",
    "\n",
    "    return df.index.to_list()\n",
    "\n",
    "\n",
    "def countsGeneNum(df, ATSS_APA_gene, transcriptIdList, geneBiotype=\"all\", transcriptBiotype=\"all\", plot_x_maxnum=5):\n",
    "    '''\n",
    "    input:\n",
    "        df, pandas.DataFrame, 具有列[\"gene_id\"]且index.name为\"transcriptId\"\n",
    "        ATSS_APA_gene, list, 存储了ATSS-APA型gene的id\n",
    "        transcriptIdList, list, 存储了transcriptId, 这些transcript在指定cellLine中存在特异性表达\n",
    "        plot_x_maxnum, int, 最大值, 高于此值的值将被修改为此值\n",
    "    change:\n",
    "        记录df中符合条件的gene的数量\n",
    "            gene具有指定数量的特异性表达transcript\n",
    "    output:\n",
    "        isoformCounts, dict, 记录了gene的数量, key为gene所具有的特异性表达transcript数量, value为gene数量\n",
    "                               额外含有key: geneBiotype, transcriptBiotype, cellLine1, cellLine2\n",
    "    '''\n",
    "    df = df\n",
    "    ATSS_APA_gene = ATSS_APA_gene\n",
    "    transcriptIdList = transcriptIdList\n",
    "    geneBiotype = geneBiotype\n",
    "    transcriptBiotype = transcriptBiotype\n",
    "    plot_x_maxnum = plot_x_maxnum\n",
    "\n",
    "    isoformCounts = {i: 0 for i in range(0, plot_x_maxnum+1)}\n",
    "\n",
    "    # 保存geneBiotype及transcriptBiotype信息\n",
    "    isoformCounts[\"geneBiotype\"] = geneBiotype\n",
    "    isoformCounts[\"transcriptBiotype\"] = transcriptBiotype\n",
    "    # 准备记录cellLine信息\n",
    "    isoformCounts[\"cellLine1\"] = None\n",
    "    isoformCounts[\"cellLine2\"] = None\n",
    "\n",
    "    # 筛选ATSS_APA型gene\n",
    "    df = df.loc[df[\"gene_id\"].isin(ATSS_APA_gene), :]\n",
    "    # 根据geneBiotype进行筛选\n",
    "    df = df if geneBiotype == \"all\" else df.query(\"gene_biotype==@geneBiotype\")\n",
    "    # 根据transcriptBiotype进行筛选\n",
    "    df = df if transcriptBiotype == \"all\" else df.query(\"transcript_biotype==@transcriptBiotype\")\n",
    "    # 记录符合geneBiotype及transcriptBiotype的gene的总数\n",
    "    geneTotalNum = len(set(df[\"gene_id\"]))\n",
    "\n",
    "    # 筛选特异性表达的transcript\n",
    "    df = df.loc[df.index.isin(transcriptIdList), :]\n",
    "    # 记录不存在特异性表达transcript的ATSS-APA型gene的数量\n",
    "    isoformCounts[0] = geneTotalNum - len(set(df[\"gene_id\"]))\n",
    "\n",
    "    df_group = df.groupby(\"gene_id\")\n",
    "    # 遍历每一个gene, 记录gene中具有特异性表达的transcript的种类数\n",
    "    for geneId,transcriptId in df_group.groups.items():\n",
    "        if len(transcriptId) > plot_x_maxnum:\n",
    "            isoformCounts[plot_x_maxnum] = isoformCounts[plot_x_maxnum] + 1\n",
    "        else:\n",
    "            isoformCounts[len(transcriptId)] = isoformCounts[len(transcriptId)] + 1\n",
    "    \n",
    "\n",
    "    return isoformCounts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 功能函数\n",
    "def compareCellLineExpression(total, tab_level=0):\n",
    "    transcriptCellLineExpression = {}\n",
    "    # 遍历每一个transcript, 获取transcript在每一个cellLine中的平均相对表达量\n",
    "    for geneId, geneObject in total.gene_dict.items():\n",
    "        for transcriptId, transcriptObject in geneObject.transcript_dict.items():\n",
    "            transcriptCellLineExpression[transcriptId] = {cellLine: transcriptObject[\"cellLineExpression\"].get(cellLine, 0) for cellLine in total.cellLineInfo.keys()}\n",
    "            transcriptCellLineExpression[transcriptId].update({\"gene_id\": geneObject.gene_id,\n",
    "                                                               \"gene_biotype\": geneObject.gene_biotype,\n",
    "                                                               \"transcript_biotype\": transcriptObject[\"transcript_biotype\"]})\n",
    "    # 准备表格的列名\n",
    "    columns = [\"gene_id\", \"gene_biotype\", \"transcript_biotype\"] + list(total.cellLineInfo.keys())\n",
    "    # 将transcript的相关信息转化为df\n",
    "    transcriptCellLineExpression = pandas.DataFrame.from_dict(transcriptCellLineExpression,\n",
    "                                                              orient=\"index\", columns=columns)\n",
    "    # 美化表格\n",
    "    transcriptCellLineExpression.index.name = \"transcript_id\"\n",
    "\n",
    "    cellLineList = list(total.cellLineInfo.keys())\n",
    "    for cellLine1Index in range(0, len(cellLineList)-1):\n",
    "        for cellLine2Index in range(cellLine1Index+1, len(cellLineList)):\n",
    "            cellLine1 = cellLineList[cellLine1Index]\n",
    "            cellLine2 = cellLineList[cellLine2Index]\n",
    "            transcriptCellLineExpression[\"log2FC_{}@{}\".format(cellLine1, cellLine2)] = transcriptCellLineExpression[cellLine1] / \\\n",
    "                                                                                        transcriptCellLineExpression[cellLine2]\n",
    "            transcriptCellLineExpression[\"log2FC_{}@{}\".format(cellLine1, cellLine2)] = transcriptCellLineExpression[\"log2FC_{}@{}\".format(cellLine1, cellLine2)].map(lambda x: 0 if numpy.isinf(x) or x==0 else math.log(x, 2))\n",
    "    \n",
    "    return transcriptCellLineExpression\n",
    "\n",
    "\n",
    "def getATSSAPAGeneId(total, ATSS_APA_cutoff=3):\n",
    "    '''\n",
    "    change:\n",
    "        获取ATSS-APA型gene\n",
    "    '''\n",
    "    total = total\n",
    "    ATSS_APA_cutoff = ATSS_APA_cutoff\n",
    "\n",
    "    geneList = []  # 存储ATSS-APA型gene的geneId\n",
    "    for geneId, geneObject in total.gene_dict.items():\n",
    "        if geneObject.gene_classified == \"ATSS-APA\":\n",
    "            # 确定该gene属于ATSS-APA型gene\n",
    "            if len(geneObject.transcript_dict) >= ATSS_APA_cutoff:\n",
    "                # 确定该gene的isoforms数量>=定值\n",
    "                geneList.append(geneId)\n",
    "\n",
    "    return geneList\n",
    "\n",
    "\n",
    "def dfCountsGeneNum(total, cellLineExpression, log2FC_cutoff, geneBiotype, transcriptBiotype, ATSS_APA_gene, plot_x_maxnum, tab_level=0):\n",
    "    '''\n",
    "    input:\n",
    "    change:\n",
    "        获取所有的cellLine组合, 并统计具有不同isoforms的gene的数量\n",
    "    output:\n",
    "        df, pandas.DataFrame, 含有列[\"cellLine1\", \"cellLine2\", \"geneBiotype\", \"transcriptBiotype\", 0, 1, 2, ...]\n",
    "    '''\n",
    "    total = total\n",
    "    cellLineExpression = cellLineExpression\n",
    "    log2FC_cutoff = log2FC_cutoff\n",
    "    geneBiotype = geneBiotype\n",
    "    transcriptBiotype = transcriptBiotype\n",
    "    ATSS_APA_gene = ATSS_APA_gene\n",
    "    plot_x_maxnum = plot_x_maxnum\n",
    "\n",
    "    num = 0\n",
    "    isoformCounts = {}\n",
    "    # 遍历所有的cellLine组合, 统计其gene数量\n",
    "    for cellLine1, cellLine2 in cellLineCombination(cl=list(total.cellLineInfo.keys())):\n",
    "        num += 1\n",
    "        transcriptIdList = getSignificantTranscriptId(total=total, cellLineExpression=cellLineExpression,\n",
    "                                                cellLine1=cellLine1, cellLine2=cellLine2,\n",
    "                                                log2Cutoff=log2FC_cutoff)\n",
    "        isoformCountsTemp = countsGeneNum(df=cellLineExpression, geneBiotype=geneBiotype, transcriptBiotype=transcriptBiotype,\n",
    "                                ATSS_APA_gene=ATSS_APA_gene, transcriptIdList=transcriptIdList,\n",
    "                                plot_x_maxnum=plot_x_maxnum)\n",
    "        isoformCountsTemp[\"cellLine1\"] = cellLine1\n",
    "        isoformCountsTemp[\"cellLine2\"] = cellLine2\n",
    "        isoformCounts[num] = isoformCountsTemp\n",
    "    # 生成df\n",
    "    columns = [\"cellLine1\", \"cellLine2\", \"geneBiotype\", \"transcriptBiotype\"]\n",
    "    columns = columns + [i for i in range(0, plot_x_maxnum+1)]\n",
    "    df = pandas.DataFrame.from_dict(isoformCounts, orient=\"index\",\n",
    "                            columns=columns)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n@def_function.log\\ndef output_transcript_gene_plot(result_counts, file_path, tab_level=0):\\n    result_counts = result_counts\\n    file_path = output_transcript_gene_plot_path\\n\\n    plot_info_df = pandas.DataFrame()\\n    for cell_line_compare, plot_info_dict in result_counts.items():\\n        for gene_transcript, value_dict in plot_info_dict.items():\\n            [cell_line_1,cell_line_2] = cell_line_compare.split(\\'@\\')\\n            [gene_biotype,transcript_biotype] = gene_transcript.split(\\'/\\')\\n            temp_df = pandas.DataFrame([[cell_line_1, cell_line_2, gene_biotype, transcript_biotype]+[-1 for i in range(0,plot_x_maxnum+1)]],\\n                                        columns=[\"cell_line_1\", \"cell_line_2\", \"gene_biotype\", \"transcript_biotype\"]+list(range(0, plot_x_maxnum+1)))\\n            for num, counts in value_dict.items():\\n                temp_df[num] = counts\\n            plot_info_df = pandas.concat([plot_info_df, temp_df], axis=0, ignore_index=True)\\n\\n    for i in plot_info_df.index:\\n        cell_line_1 = plot_info_df.at[i, \"cell_line_1\"]\\n        cell_line_2 = plot_info_df.at[i, \"cell_line_2\"]\\n        gene_biotype = plot_info_df.at[i, \"gene_biotype\"]\\n        transcript_biotype = plot_info_df.at[i, \"transcript_biotype\"]\\n        plot = make_transcript_gene_plot(df=plot_info_df,\\n                                         cell_line_1=cell_line_1,\\n                                         cell_line_2=cell_line_2,\\n                                         gene_biotype=gene_biotype,\\n                                         transcript_biotype=transcript_biotype,\\n                                         tab_level=tab_level+1)\\n        \\n        num = 0\\n        if len(transcript_biotype.split(\\'&\\')) >= 5:\\n            num += 1\\n            print(\"{}[Result]filename should be: 0004_{}_{}_{}_{}.svg\".format(\\'\\t\\'*(tab_level+1),\\n                                                                              cell_line_1, cell_line_2,\\n                                                                              gene_biotype, transcript_biotype))\\n            print(\"{}[Result]filename is: 0004_{}_{}_{}_{}.svg\".format(\\'\\t\\'*(tab_level+1),\\n                                                                              cell_line_1, cell_line_2,\\n                                                                              gene_biotype, num))\\n            plot.write_image(\"{}0004_{}_{}_{}_{}.svg\".format(file_path,\\n                                                             cell_line_1, cell_line_2,\\n                                                             gene_biotype, num))\\n        else:\\n            plot.write_image(\"{}0004_{}_{}_{}_{}.svg\".format(file_path,\\n                                                             cell_line_1, cell_line_2,\\n                                                             gene_biotype, transcript_biotype))\\n    \\n    return None\\n'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 绘图函数\n",
    "@def_function.log\n",
    "def output_length_counts_check_plot(sample_info, total_info, output_filename, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        sample_info, df\n",
    "        total_info, df\n",
    "        output_filename, str\n",
    "    change:\n",
    "        get the plot which xaxis is length of transcript and yaxis is counts of transcript\n",
    "    output:\n",
    "        None\n",
    "    \"\"\"\n",
    "    sample_info = sample_info\n",
    "    total_info = total_info\n",
    "    output_filename = output_filename\n",
    "\n",
    "    fig_info_dict = {}\n",
    "    for sample in sample_info[\"GEO_accession\"]:\n",
    "        temp_df = total_info[[sample, \"transcript_start\", \"transcript_end\"]]\n",
    "        temp_df = temp_df.loc[temp_df[sample]>0, :]  # get the rows of count>0\n",
    "        temp_df[\"length\"] = temp_df[\"transcript_end\"] - temp_df[\"transcript_start\"]\n",
    "        temp_df = temp_df[[sample, \"length\"]]\n",
    "\n",
    "        fig_info = go.Scatter(x=temp_df[\"length\"],\n",
    "                              y=temp_df[sample],\n",
    "                              name=sample,\n",
    "                              mode=\"markers\",\n",
    "                              marker={\"size\": 2.5})\n",
    "        \n",
    "        fig_info_dict[sample] = fig_info\n",
    "\n",
    "    #rows = math.ceil(math.sqrt(sample_info.shape[0]))\n",
    "    #cols = math.ceil(sample_info.shape[0]/rows)\n",
    "    fig = make_subplots(rows=1, cols=1,\n",
    "                        shared_xaxes=True, shared_yaxes=True)\n",
    "    for sample, fig_info in fig_info_dict.items():\n",
    "        fig.add_trace(fig_info)\n",
    "    \n",
    "    layout = {\"title\": {\"font\": {\"family\": \"Arial\",\n",
    "                                 \"size\": 12,\n",
    "                                 \"color\": \"black\"},\n",
    "                        },\n",
    "              \"font\": {\"family\": \"Arial\",\n",
    "                       \"size\": 10,\n",
    "                       \"color\": \"black\"},\n",
    "              \"margin\": {'l':20, 'r':20, 't':20, 'b':20},\n",
    "              \"xaxis\": {\"title\": {\"text\": \"transcripts length\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                           \"color\": \"black\",\n",
    "                                           \"size\": 12},\n",
    "                                  \"standoff\": 0.5},\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        },\n",
    "              \"yaxis\": {\"title\": {\"text\": \"transcripts counts\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                           \"color\": \"black\",\n",
    "                                           \"size\": 12},\n",
    "                                  \"standoff\": 0.5},\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        },\n",
    "              \"paper_bgcolor\": \"white\",\n",
    "              \"plot_bgcolor\": \"white\",\n",
    "              \"showlegend\": False\n",
    "              }\n",
    "\n",
    "    fig.update_layout(layout)\n",
    "\n",
    "    fig.write_image(output_length_counts_plot_filename, scale=4)\n",
    "\n",
    "    return None\n",
    "\n",
    "\n",
    "def make_transcript_gene_plot(df):\n",
    "    '''\n",
    "    input:\n",
    "        df, pandas.DataFrame, 含有列['x', 'y']\n",
    "    '''\n",
    "    df = df\n",
    "\n",
    "    plot = go.Figure(go.Bar(x=df['x'], y=df['y'],\n",
    "                            width=0.9,\n",
    "                            text=[int(df.query('x==0')['y'])],\n",
    "                            textposition=\"outside\",\n",
    "                            textfont={\"family\": \"Arial\", \"color\": \"black\", \"size\": 12},\n",
    "                            textangle=-90,\n",
    "                            marker={\"color\": \"purple\",\n",
    "                                    \"line\": {\"color\": \"black\",\n",
    "                                            \"width\": 0.5},\n",
    "                                    }))\n",
    "    layout = {\"width\": 250, \"height\": 400,\n",
    "              \"title\": {\"text\": \"{}/{}\".format(df['y'].sum()-int(df.query('x==0')['y']), df['y'].sum()),\n",
    "                        \"font\": {\"family\": \"Arial\",\n",
    "                                 \"color\": \"black\",\n",
    "                                 \"size\": 12},\n",
    "                        'x': 0.8,},\n",
    "              \"margin\": {'l':15, 'r':15, 't':25, 'b':15}, \n",
    "              \"barmode\": \"stack\",\n",
    "              \"font\": {\"family\": \"Arial\",\n",
    "                       \"color\": \"black\",\n",
    "                       \"size\": 10},\n",
    "              \"xaxis\": {\"title\": {\"text\": \"Per gene, 5'-3' isoforms <br> with a significant 5'-3' link\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                           \"color\": \"black\",\n",
    "                                           \"size\": 12},\n",
    "                                  \"standoff\": 0,},\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"dtick\": 1,\n",
    "                        \"ticksuffix\": '+',\n",
    "                        \"showticksuffix\": \"last\",\n",
    "                        },\n",
    "              \"yaxis\": {\"title\": {\"text\": \"Number of Genes\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                           \"color\": \"black\",\n",
    "                                           \"size\": 12},\n",
    "                                  \"standoff\": 0,},\n",
    "                        \"showgrid\": False,\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"tickfont\": {\"size\": 10}\n",
    "                        },\n",
    "              \"plot_bgcolor\": \"white\",\n",
    "              \"bargap\": 0, \"bargroupgap\": 0,}\n",
    "    plot = plot.update_layout(layout)\n",
    "\n",
    "    return plot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 生成器\n",
    "def cellLineCombination(cl):\n",
    "    '''\n",
    "    input:\n",
    "        cl, list, [<cellLine1>, <cellLine2>, <cellLine3>, ...]\n",
    "    generator:\n",
    "        以[cellLine1, cellLine2]形式返回给定list中的每一对组合\n",
    "    '''\n",
    "    cl = cl\n",
    "    for cellLine1Index in range(0, len(cl)-1):\n",
    "        for cellLine2Index in range(cellLine1Index+1, len(cl)):\n",
    "            # 遍历了每一对cellLine组合\n",
    "            cellLine1 = cl[cellLine1Index]\n",
    "            cellLine2 = cl[cellLine2Index]\n",
    "            yield [cellLine1, cellLine2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "[Function]load_sample_info start.\n",
      "\t[Time]2023-08-11 15:41:26\n",
      "\t[Paraments]filename: F:/OneDrive/Master/Project/trans/data/0000_sample_info.tsv\n",
      "\t[Paraments]tab_level: 0\n",
      "[2023-08-11 15:41:26]load_sample_info finished.\n",
      "[2023-08-11 15:41:36]All blocks finished.\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # load data\n",
    "    sample_info = def_function.load_sample_info(filename=input_sample_info_filename, tab_level=0)\n",
    "    total_info = pandas.read_csv(input_df_filename, sep='\\t', index_col=0)\n",
    "    with open(input_pkl_filename, \"rb\") as file:\n",
    "        total = pickle.load(file)\n",
    "    \n",
    "    # output check plot to confirm the relationship of length and counts\n",
    "    if check_counts is True:\n",
    "        output_length_counts_check_plot(sample_info=sample_info,\n",
    "                                        total_info=total_info,\n",
    "                                        output_filename=output_length_counts_plot_filename,\n",
    "                                        tab_level=0)\n",
    "\n",
    "    # compare expression of two cell line\n",
    "    # and filter transcripts according to log2FC\n",
    "    cellLineExpression = compareCellLineExpression(total=total)\n",
    "    ATSS_APA_gene = getATSSAPAGeneId(total=total, ATSS_APA_cutoff=ATSS_APA_cutoff)\n",
    "    # 将geneBiotype及transcriptBiotype归类为三种类型: [\"protein_coding\", \"nonprotein_coding\", \"un_classified\"]\n",
    "    df = cellLineExpression.copy()\n",
    "    df[\"gene_biotype\"] = df[\"gene_biotype\"].map(lambda x: \"nonprotein_coding\" if x not in [\"protein_coding\", \"un_classified\"] else x)\n",
    "    df[\"transcript_biotype\"] = df[\"transcript_biotype\"].map(lambda x: \"nonprotein_coding\" if x not in [\"protein_coding\", \"un_classified\"] else x)\n",
    "    # 获取每种geneBiotype及transcriptBiotype的gene的数目, 存储于dfCounts\n",
    "    dfCounts = pandas.DataFrame()\n",
    "    for geneBiotype in [\"all\", \"protein_coding\", \"nonprotein_coding\", \"un_classified\"]:\n",
    "        for transcriptBiotype in [\"all\", \"protein_coding\", \"nonprotein_coding\", \"un_classified\"]:\n",
    "            dfTemp = dfCountsGeneNum(total=total, cellLineExpression=df,\n",
    "                                     log2FC_cutoff=log2FC_cutoff, plot_x_maxnum=plot_x_maxnum,\n",
    "                                     geneBiotype=geneBiotype, transcriptBiotype=transcriptBiotype,\n",
    "                                     ATSS_APA_gene=ATSS_APA_gene, tab_level=0)\n",
    "            dfCounts = pandas.concat([dfCounts, dfTemp], axis=0, ignore_index=True)\n",
    "\n",
    "    # 绘图\n",
    "    plotDict = dfCounts.to_dict(\"index\")\n",
    "    for plotId, plotInfo in plotDict.items():\n",
    "        dfTemp = pandas.DataFrame({'x': [i for i in range(0, plot_x_maxnum+1)],\n",
    "                                'y': [plotInfo[i] for i in range(0, plot_x_maxnum+1)]})\n",
    "        plotDict[plotId].update({\"plot\": make_transcript_gene_plot(dfTemp)})\n",
    "\n",
    "    # 输出结果\n",
    "    columns = [\"cellLine1\", \"cellLine2\", \"geneBiotype\", \"transcriptBiotype\"]\n",
    "    columns = columns + [i for i in range(0, plot_x_maxnum+1)]\n",
    "    df = pandas.DataFrame.from_dict(plotDict, orient=\"index\",\n",
    "                                    columns=columns)\n",
    "    df.to_csv(output_check_df_filename, sep='\\t')\n",
    "    cellLineExpression.to_csv(output_relative_expression_filename, sep='\\t')\n",
    "    for plotId, plotInfo in plotDict.items():\n",
    "        plotInfo[\"plot\"].write_image(\"{}{}_{}_{}.svg\".format(output_transcript_gene_plot_path, plotId,\n",
    "                                                             plotInfo[\"cellLine1\"], plotInfo[\"cellLine2\"]))\n",
    "\n",
    "    print(\"[{}]All blocks finished.\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "1.对每一个样本中的transcript的表达量进行标准化\n",
    "  2.1不确定原始数据的rep是如何得到的, 不知道应该用什么计算方法\n",
    "  2.2姑且认为rep值已经对基因长度进行了校正, 因为rep值将近80%都<=10, 值很低\n",
    "  2.3所以认为 一个transcript在一个样本中的相对表达量 = 该transcript的rep值 / 该样本所有transcript的rep值的和\n",
    "  2.4再计算该transcript在相同疾病类型中的不同样本中的平均表达量= 该transcript在一种类型疾病中所有样本的相对表达量的加和 / 该类型疾病的样本数\n",
    "  2.4再了解一下rep值与transcript长度是否相关\n",
    "2.筛选得到在指定的两种细胞系类型的样本中都存在表达ATSS-APA基因\n",
    "  2.1根据所有样本的数据筛选ATSS-APA基因id\n",
    "  2.2获取每种细胞系中存在表达的基因id, 以list形式存储\n",
    "  2.3分别筛选出在单一一种细胞系类型中表达的基因，然后将这些基因合并后去重\n",
    "3.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "调试"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A673 Caco-2\n"
     ]
    }
   ],
   "source": [
    "cellLineList = list(total.cellLineInfo.keys())\n",
    "for i in range(0, len(cellLineList)-1):\n",
    "    for j in range(i+1, len(cellLineList)):\n",
    "        cellLine1 = cellLineList[i]\n",
    "        cellLine2 = cellLineList[j]\n",
    "        print(cellLine1, cellLine2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "调试\n",
    "\n",
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
