{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import time\n",
    "import argparse\n",
    "import def_function\n",
    "import pickle\n",
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "import math\n",
    "import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "__version__ = \"V1.2(Editor) 2023-08-05\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数-debug\n",
    "ATSS_APA_cutoff = 3\n",
    "log2FC_cutoff = 1.5\n",
    "plot_x_maxnum = 5\n",
    "absolute_path = False\n",
    "file_path = \"F:/OneDrive/Master/Project/trans/data/\"\n",
    "input_sample_info_filename = \"0000_sample_info.tsv\"\n",
    "input_df_filename = \"0001_total_info.tsv\"\n",
    "input_TSS_TES_filename = \"0002_result_check.tsv\"\n",
    "output_length_counts_plot_filename = \"0004_length_counts.png\"\n",
    "output_check_df_filename = \"0004_check_df.xlsx\"\n",
    "output_transcript_gene_plot_path = \"0004_plot/\"\n",
    "output_relative_expression_filename = \"0004_relative_expression.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] [--log2FC_cutoff LOG2FC_CUTOFF]\n",
      "                             [--plot_x_maxnum PLOT_X_MAXNUM] [--absolute_path]\n",
      "                             [--file_path FILE_PATH]\n",
      "                             [--input_sample_info_filename INPUT_SAMPLE_INFO_FILENAME]\n",
      "                             [--input_df_filename INPUT_DF_FILENAME]\n",
      "                             [--input_TSS_TES_filename INPUT_TSS_TES_FILENAME]\n",
      "                             [--output_length_counts_plot_filename OUTPUT_LENGTH_COUNTS_PLOT_FILENAME]\n",
      "                             [--output_check_df_filename OUTPUT_CHECK_DF_FILENAME]\n",
      "                             [--output_transcript_gene_plot_path OUTPUT_TRANSCRIPT_GENE_PLOT_PATH]\n",
      "                             [--output_relative_expression_filename OUTPUT_RELATIVE_EXPRESSION_FILENAME]\n",
      "ipykernel_launcher.py: error: unrecognized arguments: --ip=127.0.0.1 --stdin=9013 --control=9011 --hb=9010 --Session.signature_scheme=\"hmac-sha256\" --Session.key=b\"bcd69888-ac5f-4b4e-a9ed-b55509d1cbcf\" --shell=9012 --transport=\"tcp\" --iopub=9014\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[1;31mSystemExit\u001b[0m\u001b[1;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Administrator\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3406: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "# 前置参数\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"--ATSS_APA_cutoff\", dest=\"ATSS_APA_cutoff\", required=False, type=int, default=3, help=\"=3,\\t the cutoff value of isoform of ATSS-APA gene\")\n",
    "parser.add_argument(\"--log2FC_cutoff\", dest=\"log2FC_cutoff\", required=False, type=float, default=1.5, help=\"=1.5,\\t the cutoff value of log2FC\")\n",
    "parser.add_argument(\"--plot_x_maxnum\", dest=\"plot_x_maxnum\", required=False, type=int, default=5, help=\"=5,\\t the maxnum in xaxis\")\n",
    "parser.add_argument(\"--absolute_path\", dest=\"absolute_path\", required=False, action=\"store_true\", help=\"use the absolute path\")\n",
    "parser.add_argument(\"--file_path\", dest=\"file_path\", required=False, type=str, default=\"./\", help=\"=the data directory path\")\n",
    "parser.add_argument(\"--input_sample_info_filename\", dest=\"input_sample_info_filename\", required=False, type=str, default=\"0000_sample_info.tsv\", help=\"=0000_sample_info.tsv,\\t the metadata of samples\")\n",
    "parser.add_argument(\"--input_df_filename\", dest=\"input_df_filename\", required=False, type=str, default=\"0001_total_info.tsv\", help=\"=0001_total_info.tsv,\\t the output file of 0001.py\")\n",
    "parser.add_argument(\"--input_TSS_TES_filename\", dest=\"input_TSS_TES_filename\", required=False, type=str, default=\"0002_result_check.tsv\", help=\"=0002_result_check.tsv,\\t the output file of 0002.py\")\n",
    "parser.add_argument(\"--output_length_counts_plot_filename\", dest=\"output_length_counts_plot_filename\", required=False, type=str, default=\"0004_length_counts.png\", help=\"=0004_length_counts.png,\\t to confirm rep is not related to length of transcript\")\n",
    "parser.add_argument(\"--output_check_df_filename\", dest=\"output_check_df_filename\", required=False, type=str, default=\"0004_check_df.xlsx\", help=\"=0004_check_df.xlsx,\\t the output file for checking\")\n",
    "parser.add_argument(\"--output_transcript_gene_plot_path\", dest=\"output_transcript_gene_plot_path\", required=False, type=str, default=\"0004_plot/\", help=\"=0004_plot/,\\t the path of output plot\")\n",
    "parser.add_argument(\"--output_relative_expression_filename\", dest=\"output_relative_expression_filename\", required=False, type=str, default=\"0004_relative_expression.tsv\", help=\"=0004_relative_expression.tsv,\\t the output filename of cell_line expression\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "ATSS_APA_cutoff = args.ATSS_APA_cutoff\n",
    "log2FC_cutoff = args.log2FC_cutoff\n",
    "plot_x_maxnum = args.plot_x_maxnum\n",
    "absolute_path = args.absolute_path\n",
    "file_path = args.file_path\n",
    "input_sample_info_filename = args.input_sample_info_filename\n",
    "input_df_filename = args.input_df_filename\n",
    "input_TSS_TES_filename = args.input_TSS_TES_filename\n",
    "output_length_counts_plot_filename = args.output_length_counts_plot_filename\n",
    "output_check_df_filename = args.output_check_df_filename\n",
    "output_transcript_gene_plot_path = args.output_transcript_gene_plot_path\n",
    "output_relative_expression_filename = args.output_relative_expression_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补充路径\n",
    "if absolute_path is False:\n",
    "    input_sample_info_filename = \"{}{}\".format(file_path, input_sample_info_filename)\n",
    "    input_df_filename = \"{}{}\".format(file_path, input_df_filename)\n",
    "    input_TSS_TES_filename = \"{}{}\".format(file_path, input_TSS_TES_filename)\n",
    "    output_length_counts_plot_filename = \"{}other/{}\".format(file_path, output_length_counts_plot_filename)\n",
    "    output_check_df_filename = \"{}{}\".format(file_path, output_check_df_filename)\n",
    "    output_transcript_gene_plot_path = \"{}{}\".format(file_path, output_transcript_gene_plot_path)\n",
    "    output_relative_expression_filename = \"{}{}\".format(file_path, output_relative_expression_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 打印参数\n",
    "print('\\n')\n",
    "print(\"[Date]{}\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))\n",
    "print(\"[Script]{}\".format(__file__))\n",
    "print(\"[Version]{}\".format(__version__))\n",
    "print(\"[Version]def_function: {}\".format(def_function.__version__))\n",
    "print(\"[Parament]ATSS_APA_cutoff: {}\".format(ATSS_APA_cutoff))\n",
    "print(\"[Parament]log2FC_cutoff: {}\".format(log2FC_cutoff))\n",
    "print(\"[Parament]plot_x_maxnum: {}\".format(plot_x_maxnum))\n",
    "print(\"[Parament]file_path: {}\".format(file_path))\n",
    "print(\"[Parament]input_sample_info_filename: {}\".format(input_sample_info_filename))\n",
    "print(\"[Parament]input_df_filename: {}\".format(input_df_filename))\n",
    "print(\"[Parament]input_TSS_TES_filename: {}\".format(input_TSS_TES_filename))\n",
    "print(\"[Parament]output_length_counts_plot_filename: {}\".format(output_length_counts_plot_filename))\n",
    "print(\"[Parament]output_check_df_filename: {}\".format(output_check_df_filename))\n",
    "print(\"[Parament]output_transcript_gene_plot_path: {}\".format(output_transcript_gene_plot_path))\n",
    "print(\"[Parament]output_relative_expression_filename: {}\".format(output_relative_expression_filename))\n",
    "print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "@def_function.log\n",
    "def output_length_counts_check_plot(sample_info, total_info, output_filename, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        sample_info, df\n",
    "        total_info, df\n",
    "        output_filename, str\n",
    "    change:\n",
    "        get the plot which xaxis is length of transcript and yaxis is counts of transcript\n",
    "    output:\n",
    "        None\n",
    "    \"\"\"\n",
    "    sample_info = sample_info\n",
    "    total_info = total_info\n",
    "    output_filename = output_filename\n",
    "\n",
    "    fig_info_dict = {}\n",
    "    for sample in sample_info[\"GEO_accession\"]:\n",
    "        temp_df = total_info[[sample, \"transcript_start\", \"transcript_end\"]]\n",
    "        temp_df = temp_df.loc[temp_df[sample]>0, :]  # get the rows of count>0\n",
    "        temp_df[\"length\"] = temp_df[\"transcript_end\"] - temp_df[\"transcript_start\"]\n",
    "        temp_df = temp_df[[sample, \"length\"]]\n",
    "\n",
    "        fig_info = go.Scatter(x=temp_df[\"length\"],\n",
    "                              y=temp_df[sample],\n",
    "                              name=sample,\n",
    "                              mode=\"markers\",\n",
    "                              marker={\"size\": 2.5})\n",
    "        \n",
    "        fig_info_dict[sample] = fig_info\n",
    "\n",
    "    #rows = math.ceil(math.sqrt(sample_info.shape[0]))\n",
    "    #cols = math.ceil(sample_info.shape[0]/rows)\n",
    "    fig = make_subplots(rows=1, cols=1,\n",
    "                        shared_xaxes=True, shared_yaxes=True)\n",
    "    for sample, fig_info in fig_info_dict.items():\n",
    "        fig.add_trace(fig_info)\n",
    "    \n",
    "    layout = {\"title\": {\"font\": {\"family\": \"Arial\",\n",
    "                                 \"size\": 12,\n",
    "                                 \"color\": \"black\"},\n",
    "                        },\n",
    "              \"font\": {\"family\": \"Arial\",\n",
    "                       \"size\": 10,\n",
    "                       \"color\": \"black\"},\n",
    "              \"margin\": {'l':20, 'r':20, 't':20, 'b':20},\n",
    "              \"xaxis\": {\"title\": {\"text\": \"transcripts length\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                           \"color\": \"black\",\n",
    "                                           \"size\": 12},\n",
    "                                  \"standoff\": 0.5},\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        },\n",
    "              \"yaxis\": {\"title\": {\"text\": \"transcripts counts\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                           \"color\": \"black\",\n",
    "                                           \"size\": 12},\n",
    "                                  \"standoff\": 0.5},\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        },\n",
    "              \"paper_bgcolor\": \"white\",\n",
    "              \"plot_bgcolor\": \"white\",\n",
    "              \"showlegend\": False\n",
    "              }\n",
    "\n",
    "    fig.update_layout(layout)\n",
    "\n",
    "    fig.write_image(output_length_counts_plot_filename, scale=4)\n",
    "\n",
    "    return None\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def compute_expression_for_each_cell_line(total_info, sample_info,\n",
    "                                          cell_line_sample_dict, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        total_info, pandas.DataFrame, ...\n",
    "        sample_info, pandas.DataFrame, ...\n",
    "        cell_line_sample_dict, dict, {cell_line1: [sample1, sample2, ...], ...}\n",
    "    change:\n",
    "        一个transcript在一个样本中的相对表达量 = 该transcript的rep值 / 该样本所有transcript的rep值的和 * 10**6\n",
    "        该transcript在相同疾病类型中的不同样本中的平均表达量= 该transcript在一种类型疾病中所有样本的相对表达量的加和 / 该类型疾病的样本数\n",
    "    output:\n",
    "        total_info, pandas.DataFrame, expression in cell_line added\n",
    "    \"\"\"\n",
    "    total_info = total_info\n",
    "    sample_info = sample_info\n",
    "    cell_line_sample_dict = cell_line_sample_dict\n",
    "\n",
    "    # 计算transcript在单个样本中的相对表达量\n",
    "    for sample in sample_info[\"GEO_accession\"]:\n",
    "        sample_sum = total_info[sample].sum()\n",
    "        total_info[\"{}_expression_sample\".format(sample)] = total_info[sample].map(lambda x: x/sample_sum * 10**6)\n",
    "    # 计算transcript在同种疾病类型的不同样本中的平均相对表达量\n",
    "    for cell_line, sample_list in cell_line_sample_dict.items():\n",
    "        column_name = \"{}_expression\".format(cell_line)\n",
    "        sample_list_expression = [\"{}_expression_sample\".format(i) for i in sample_list]\n",
    "        total_info[column_name] = total_info[sample_list_expression].apply(lambda x: x.sum(), axis=1)\n",
    "        total_info[column_name] = total_info[column_name] / len(sample_list)\n",
    "    # drop <sample>_expression_sample column\n",
    "    drop_column = [\"{}_expression_sample\".format(i) for i in sample_info[\"GEO_accession\"]]\n",
    "    total_info = total_info.drop(labels=drop_column, axis=1)\n",
    "\n",
    "    return total_info\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def get_ATSS_APA_gene_in_cell_line(df, gene_biotype, cell_line_1, cell_line_2, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df, pandas.DataFrame, the output file of 0002.py\n",
    "        gene_biotype, str, ...\n",
    "        transcript_biotype, str, ...\n",
    "        cell_line_1, str, 注意尽量以下划线代替空格\n",
    "        cell_line_2, str, 注意尽量以下划线代替空格\n",
    "    change:\n",
    "        ...\n",
    "    output:\n",
    "        # result, dict, {<gene_id>: <transcript_counts>, ...}\n",
    "        result, set, 指定gene_biotype的ATSS-APA基因id\n",
    "    \"\"\"\n",
    "    df = df\n",
    "    gene_biotype = gene_biotype\n",
    "    cell_line_1 = cell_line_1\n",
    "    cell_line_2 = cell_line_2\n",
    "\n",
    "    cell_line_1 = cell_line_1.replace(' ', '_')\n",
    "    cell_line_2 = cell_line_2.replace(' ', '_')\n",
    "    df[\"cell_line\"] = df[\"cell_line\"].map(lambda x: x.replace(' ', '_'))\n",
    "\n",
    "    # 根据所有样本的数据筛选指定gene_biotype的ATSS-APA基因\n",
    "    temp_df = df.loc[df[\"cell_line\"]==\"all\", :]\n",
    "    if gene_biotype != \"all\":\n",
    "        temp_df = temp_df.loc[temp_df[\"gene_biotype\"]==gene_biotype, :]  # retain specified gene_biotype\n",
    "    temp_df = temp_df.query(\"start_counts >=2 & end_counts >=2\")  # retain ATSS-PAS gene\n",
    "    ATSS_APA_gene = set(temp_df[\"gene_id\"].to_list())\n",
    "    \n",
    "\n",
    "    temp_df = df.loc[df[\"cell_line\"]==cell_line_1, :]\n",
    "    if gene_biotype != \"all\":\n",
    "        temp_df = temp_df.loc[temp_df[\"gene_biotype\"]==gene_biotype, :]  # retain specified gene_biotype\n",
    "    cell_line_1_gene = temp_df[\"gene_id\"].to_list()\n",
    "\n",
    "    temp_df = df.loc[df[\"cell_line\"]==cell_line_2, :]\n",
    "    if gene_biotype != \"all\":\n",
    "        temp_df = temp_df.loc[temp_df[\"gene_biotype\"]==gene_biotype, :]  # retain specified gene_biotype\n",
    "    cell_line_2_gene = temp_df[\"gene_id\"].to_list()\n",
    "\n",
    "    # cell_line_ATSS_PAS_gene, set, 在两种类型疾病中都存在表达的指定gene_biotype的ATSS-APA基因\n",
    "    cell_line_gene = set(cell_line_1_gene+cell_line_2_gene)\n",
    "    cell_line_ATSS_APA_gene = ATSS_APA_gene.intersection(cell_line_gene)\n",
    "\n",
    "    \"\"\"# get result dict\n",
    "    result = {}\n",
    "    result_index = list(cell_line_ATSS_APA_gene)\n",
    "    df = df.loc[df[\"cell_line\"]==\"all\", :]\n",
    "    df = df.set_index(\"gene_id\")\n",
    "    df = df.loc[result_index, :]\n",
    "    for gene_id in df.index:\n",
    "        result[gene_id] = df.at[gene_id, \"transcript_counts\"]\"\"\"\n",
    "    \n",
    "    result = cell_line_ATSS_APA_gene\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def compute_counts_of_gene(significant_gene_transcript, max_num=5, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        significant_gene_transcript, dict, {<gene_id>: <the counts of significant transcript of gene>, ...}\n",
    "        max_num, int, the max num of compute\n",
    "    change:\n",
    "        根据significant_gene_transcript, 统计具有不同显著表达差异的transcript数目的gene数\n",
    "        如果一个基因所具有的显著表达差异transcript的数目超过了max_num, 就认为该gene具有max_num个显著表达差异transcript\n",
    "    output:\n",
    "        counts_num, dict, {<counts of significant transcript>: <counts of gene>}\n",
    "    \"\"\"\n",
    "    max_num = 5\n",
    "    counts_num = {num: 0 for num in range(1,max_num+1)}\n",
    "    for gene_id, counts in significant_gene_transcript.items():\n",
    "        if counts <= max_num:\n",
    "            counts_num[counts] = counts_num.get(counts) + 1\n",
    "        else:\n",
    "            counts_num[max_num] = counts_num.get(max_num) +1\n",
    "    \n",
    "    return counts_num\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def get_counts_of_transcript_of_gene(isoform_info_df, total_info,\n",
    "                                     cell_line_1, cell_line_2,\n",
    "                                     gene_biotype, transcript_biotype=[\"all\"],\n",
    "                                     log2FC_cutoff=1.5, max_num=5,\n",
    "                                     tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        isoform_info_df, pandas.DataFrame, the output of 0002.py\n",
    "        total_info, pandas.DataFrame, ...\n",
    "        gene_biotype, str, ...\n",
    "        transcript_biotype, list=[\"all\"], the transcript will be retained. If \"all\" in this list, the transcript won't be filtered according to the transcript_biotype.\n",
    "        log2FC_cutoff, float=1.5, the cutoff value of log2FC\n",
    "        max_num, int=5, the maxnum of significant expressed transcript in one gene, 如果值大于该阈值则会将值修改为该阈值\n",
    "    change:\n",
    "    output:\n",
    "        dict={\"check_df\": check_df, \"result\": result}, check_df is df and result is dict\n",
    "    \"\"\"\n",
    "    isoform_info_df = isoform_info_df\n",
    "    total_info = total_info\n",
    "    gene_biotype = gene_biotype\n",
    "    transcript_biotype = transcript_biotype\n",
    "    cell_line_1 = cell_line_1\n",
    "    cell_line_2 = cell_line_2\n",
    "    log2FC_cutoff = log2FC_cutoff\n",
    "    max_num = max_num\n",
    "\n",
    "\n",
    "    ATSS_APA_gene_set = get_ATSS_APA_gene_in_cell_line(df=isoform_info_df,\n",
    "                                                       gene_biotype=gene_biotype,\n",
    "                                                       cell_line_1=cell_line_1,\n",
    "                                                       cell_line_2=cell_line_2,\n",
    "                                                       tab_level=tab_level+1)\n",
    "    print(\"{}[Result]the counts of gene about ATSS-APA and {} gene: {}\".format('\\t'*(tab_level+1), gene_biotype, len(ATSS_APA_gene_set)))\n",
    "\n",
    "    # gene and transcript filter\n",
    "    if gene_biotype != \"all\":\n",
    "        check_df = total_info.loc[total_info[\"gene_biotype\"]==gene_biotype, :]  # specify gene_biotype\n",
    "    else:\n",
    "        check_df = total_info\n",
    "    if \"all\" not in transcript_biotype:\n",
    "        check_df = check_df.loc[check_df[\"transcript_biotype\"].isin(transcript_biotype), :]  # specify transcript_biotype\n",
    "    check_df = check_df.loc[check_df[\"gene_id\"].isin(ATSS_APA_gene_set), :]  # specify gene_biotype and ATSS-PAS\n",
    "\n",
    "    # change log format\n",
    "    if len(transcript_biotype) <= 5:\n",
    "        print(\"{}[Result]the counts of gene about ATSS-APA and {} gene and {} transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                                                    gene_biotype,\n",
    "                                                                                                    transcript_biotype,\n",
    "                                                                                                    len(set(check_df[\"gene_id\"]))))\n",
    "        print(\"{}[Result]the counts of transcript about ATSS-APA and {} gene and {} transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                                                        gene_biotype,\n",
    "                                                                                                        transcript_biotype,\n",
    "                                                                                                        check_df.shape[0]))\n",
    "    else:\n",
    "        print(\"{}[Result]the counts of gene about ATSS-APA and {} gene and <transcript_biotype> transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                                                    gene_biotype,\n",
    "                                                                                                    len(set(check_df[\"gene_id\"]))))\n",
    "        print(\"{}[Result]the counts of transcript about ATSS-APA and {} gene and <transcript_biotyoe> transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                                                        gene_biotype,\n",
    "                                                                                                        check_df.shape[0]))\n",
    "    \n",
    "    # 计算log2FC, 计算完成后，值有两种类型，一种是常规的log2FC值，另一种是numpy.nan, numpy.nan代表着该transcript在两种类型疾病中的一种中的相对表达量为0\n",
    "    check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)] = check_df[\"{}_expression\".format(cell_line_1)] / check_df[\"{}_expression\".format(cell_line_2)]\n",
    "    check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)] = check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)].map(lambda x: numpy.nan if numpy.isinf(x) or x==0 else x)\n",
    "    check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)] = check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)].map(lambda x: math.log(x, 2))\n",
    "\n",
    "    # 筛选出在两种疾病类型中具有显著表达差异的transcript\n",
    "    ## numpy.nan\n",
    "    temp_df = map(lambda x: numpy.isnan(x), check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)])\n",
    "    temp_df = list(map(lambda x: numpy.isnan(x), check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)]))\n",
    "    temp_df = check_df.loc[temp_df, :]\n",
    "    ## log2FC >= cutoff\n",
    "    check_df = check_df.loc[abs(\n",
    "                                check_df[\"log2FC({}/{})\".format(cell_line_1, cell_line_2)]\n",
    "                                ) >= log2FC_cutoff, :]\n",
    "    ## concat\n",
    "    check_df = pandas.concat([check_df, temp_df], axis=0, ignore_index=True)\n",
    "    if len(transcript_biotype) <= 5:\n",
    "        print(\"{}[Result]the counts of gene showing significant expression in {} and {} about ATSS-APA and {} gene and {} transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                    cell_line_1,\n",
    "                                                                    cell_line_2,\n",
    "                                                                    gene_biotype,\n",
    "                                                                    transcript_biotype,\n",
    "                                                                    len(set(check_df[\"gene_id\"]))))\n",
    "        print(\"{}[Result]the counts of transcript showing significant expression in {} and {} about ATSS-APA and {} gene and {} transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                    cell_line_1,\n",
    "                                                                    cell_line_2,\n",
    "                                                                    gene_biotype,\n",
    "                                                                    transcript_biotype,\n",
    "                                                                    check_df.shape[0]))\n",
    "    else:\n",
    "        print(\"{}[Result]the counts of gene showing significant expression in {} and {} about ATSS-APA and {} gene and <transcript_biotype> transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                                       cell_line_1,\n",
    "                                                                                       cell_line_2,\n",
    "                                                                                       gene_biotype,\n",
    "                                                                                       len(set(check_df[\"gene_id\"]))))\n",
    "        print(\"{}[Result]the counts of transcript showing significant expression in {} and {} about ATSS-APA and {} gene and <transcript_biotype> transcript: {}\".format('\\t'*(tab_level+1),\n",
    "                                                                                       cell_line_1,\n",
    "                                                                                       cell_line_2,\n",
    "                                                                                       gene_biotype,\n",
    "                                                                                       check_df.shape[0]))\n",
    "\n",
    "    significant_gene_transcript = {}  # {<gene_id>: <the num of significant transcript of gene>}\n",
    "    check_df_group = check_df.groupby(\"gene_id\")\n",
    "    for gene_id in check_df_group.groups.keys():\n",
    "        temp_df = check_df_group.get_group(gene_id)\n",
    "        transcript_num = temp_df.shape[0]\n",
    "        significant_gene_transcript[gene_id] = transcript_num\n",
    "\n",
    "    result = compute_counts_of_gene(significant_gene_transcript,\n",
    "                                    max_num=max_num,\n",
    "                                    tab_level=tab_level+1)\n",
    "    result[0] = len(ATSS_APA_gene_set) - len(set(check_df[\"gene_id\"]))\n",
    "\n",
    "    return {\"check_df\": check_df, \"result\": result}\n",
    "\n",
    "\n",
    "# output file\n",
    "@def_function.log\n",
    "def output_check_df(df_dict, filename, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df, dict, {<df_name>: <df>, ...}\n",
    "    change:\n",
    "        output all df to .xlsx file\n",
    "    output:\n",
    "        None\n",
    "    \"\"\"\n",
    "    df_dict = df_dict\n",
    "    filename = filename\n",
    "\n",
    "    num = 0\n",
    "    with pandas.ExcelWriter(filename, mode='w') as file:\n",
    "        for key, df in df_dict.items():\n",
    "            print(\"{}[Result]sheet_name should be: {}\".format('\\t'*(tab_level+1), key))\n",
    "            key = key.split('@')\n",
    "\n",
    "            num += 1\n",
    "            key[2] = str(num)\n",
    "\n",
    "            num += 1\n",
    "            key[3] = str(num)\n",
    "\n",
    "            key = '_'.join(key)\n",
    "            print(\"{}[Result]sheet_name is: {}\".format('\\t'*(tab_level+1), key))\n",
    "\n",
    "            df.to_excel(file, sheet_name=key, index=None)\n",
    "    \n",
    "    return None\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def make_transcript_gene_plot(df, cell_line_1, cell_line_2, gene_biotype, transcript_biotype, tab_level=0):\n",
    "        plot_info_df = df\n",
    "        cell_line_1 = cell_line_1\n",
    "        cell_line_2 = cell_line_2\n",
    "        gene_biotype = gene_biotype\n",
    "        transcript_biotype = transcript_biotype\n",
    "\n",
    "        temp_df = plot_info_df.query(\"cell_line_1=='{}' & cell_line_2=='{}' & gene_biotype=='{}' & transcript_biotype==\\\"{}\\\"\".format(cell_line_1, cell_line_2, gene_biotype, transcript_biotype)).copy()\n",
    "        temp_df = temp_df.drop(columns=[\"cell_line_1\", \"cell_line_2\", \"gene_biotype\", \"transcript_biotype\"])\n",
    "        temp_df[\"index\"] = 'y'\n",
    "        temp_df = temp_df.set_index(\"index\")\n",
    "        temp_df = temp_df.T\n",
    "        temp_df['x'] = temp_df.index\n",
    "        temp_num0 = temp_df.at[0, 'y']\n",
    "        temp_df.at[0, 'y'] = 0\n",
    "\n",
    "        temp_plot = go.Figure(go.Bar(x=temp_df['x'], y=temp_df['y'],\n",
    "                                width=0.9,\n",
    "                                text=[\"{}\".format(temp_num0)],\n",
    "                                textposition=\"outside\",\n",
    "                                textfont={\"family\": \"Arial\", \"color\": \"black\", \"size\": 12},\n",
    "                                textangle=-90,\n",
    "                                marker={\"color\": \"purple\",\n",
    "                                        \"line\": {\"color\": \"black\",\n",
    "                                                \"width\": 0.5},\n",
    "                                        }))\n",
    "        layout = {\"width\": 250, \"height\": 400,\n",
    "                \"title\": {\"text\": \"{}/{}\".format(temp_df['y'][1:].sum(), temp_df['y'].sum()+temp_num0),\n",
    "                        \"font\": {\"family\": \"Arial\",\n",
    "                                \"color\": \"black\",\n",
    "                                \"size\": 12},\n",
    "                        'x': 0.8,\n",
    "                        },\n",
    "                \"margin\": {'l':15, 'r':15, 't':25, 'b':15}, \n",
    "                \"barmode\": \"stack\",\n",
    "                \"font\": {\"family\": \"Arial\",\n",
    "                        \"color\": \"black\",\n",
    "                        \"size\": 10},\n",
    "                \"xaxis\": {\"title\": {\"text\": \"Per gene, 5'-3' isoforms <br> with a significant 5'-3' link\",\n",
    "                                \"font\": {\"family\": \"Arial\",\n",
    "                                        \"color\": \"black\",\n",
    "                                        \"size\": 12},\n",
    "                                \"standoff\": 0,\n",
    "                                },\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"dtick\": 1,\n",
    "                        \"ticksuffix\": '+',\n",
    "                        \"showticksuffix\": \"last\",\n",
    "                        },\n",
    "                \"yaxis\": {\"title\": {\"text\": \"Number of Genes\",\n",
    "                                \"font\": {\"family\": \"Arial\",\n",
    "                                        \"color\": \"black\",\n",
    "                                        \"size\": 12},\n",
    "                                \"standoff\": 0,},\n",
    "                        \"showgrid\": False,\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"tickfont\": {\"size\": 10}\n",
    "                        },\n",
    "                \"plot_bgcolor\": \"white\",\n",
    "                \"bargap\": 0, \"bargroupgap\": 0,\n",
    "                }\n",
    "        temp_plot = temp_plot.update_layout(layout)\n",
    "\n",
    "        return temp_plot\n",
    "\n",
    "\n",
    "@def_function.log\n",
    "def output_transcript_gene_plot(result_counts, file_path, tab_level=0):\n",
    "    result_counts = result_counts\n",
    "    file_path = output_transcript_gene_plot_path\n",
    "\n",
    "    plot_info_df = pandas.DataFrame()\n",
    "    for cell_line_compare, plot_info_dict in result_counts.items():\n",
    "        for gene_transcript, value_dict in plot_info_dict.items():\n",
    "            [cell_line_1,cell_line_2] = cell_line_compare.split('@')\n",
    "            [gene_biotype,transcript_biotype] = gene_transcript.split('/')\n",
    "            temp_df = pandas.DataFrame([[cell_line_1, cell_line_2, gene_biotype, transcript_biotype]+[-1 for i in range(0,plot_x_maxnum+1)]],\n",
    "                                        columns=[\"cell_line_1\", \"cell_line_2\", \"gene_biotype\", \"transcript_biotype\"]+list(range(0, plot_x_maxnum+1)))\n",
    "            for num, counts in value_dict.items():\n",
    "                temp_df[num] = counts\n",
    "            plot_info_df = pandas.concat([plot_info_df, temp_df], axis=0, ignore_index=True)\n",
    "\n",
    "    for i in plot_info_df.index:\n",
    "        cell_line_1 = plot_info_df.at[i, \"cell_line_1\"]\n",
    "        cell_line_2 = plot_info_df.at[i, \"cell_line_2\"]\n",
    "        gene_biotype = plot_info_df.at[i, \"gene_biotype\"]\n",
    "        transcript_biotype = plot_info_df.at[i, \"transcript_biotype\"]\n",
    "        plot = make_transcript_gene_plot(df=plot_info_df,\n",
    "                                         cell_line_1=cell_line_1,\n",
    "                                         cell_line_2=cell_line_2,\n",
    "                                         gene_biotype=gene_biotype,\n",
    "                                         transcript_biotype=transcript_biotype,\n",
    "                                         tab_level=tab_level+1)\n",
    "        \n",
    "        num = 0\n",
    "        if len(transcript_biotype.split('&')) >= 5:\n",
    "            num += 1\n",
    "            print(\"{}[Result]filename should be: 0004_{}_{}_{}_{}.svg\".format('\\t'*(tab_level+1),\n",
    "                                                                              cell_line_1, cell_line_2,\n",
    "                                                                              gene_biotype, transcript_biotype))\n",
    "            print(\"{}[Result]filename is: 0004_{}_{}_{}_{}.svg\".format('\\t'*(tab_level+1),\n",
    "                                                                              cell_line_1, cell_line_2,\n",
    "                                                                              gene_biotype, num))\n",
    "            plot.write_image(\"{}0004_{}_{}_{}_{}.svg\".format(file_path,\n",
    "                                                             cell_line_1, cell_line_2,\n",
    "                                                             gene_biotype, num))\n",
    "        else:\n",
    "            plot.write_image(\"{}0004_{}_{}_{}_{}.svg\".format(file_path,\n",
    "                                                             cell_line_1, cell_line_2,\n",
    "                                                             gene_biotype, transcript_biotype))\n",
    "    \n",
    "    return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "[Function]load_sample_info start.\n",
      "\t[Time]2023-08-05 20:40:08\n",
      "\t[Paraments]filename: F:/OneDrive/Master/Project/trans/data/0000_sample_info.tsv\n",
      "\t[Paraments]tab_level: 0\n",
      "[2023-08-05 20:40:08]load_sample_info finished.\n",
      "\n",
      "\n",
      "[Function]output_length_counts_check_plot start.\n",
      "\t[Time]2023-08-05 20:40:08\n",
      "\t[Paraments]sample_info: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]output_filename: F:/OneDrive/Master/Project/trans/data/other/0004_length_counts.png\n",
      "\t[Paraments]tab_level: 0\n",
      "[2023-08-05 20:40:24]output_length_counts_check_plot finished.\n",
      "\n",
      "\n",
      "[Function]get_cell_line_sample_dict start.\n",
      "\t[Time]2023-08-05 20:40:24\n",
      "\t[Paraments]sample_info: <...>\n",
      "\t[Paraments]GEO_index: False\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Result]cell_line_sample_dict: {'A673': ['GSM6783527'], 'Caco-2': ['GSM6782551']}\n",
      "[2023-08-05 20:40:24]get_cell_line_sample_dict finished.\n",
      "\n",
      "\n",
      "[Function]compute_expression_for_each_cell_line start.\n",
      "\t[Time]2023-08-05 20:40:24\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]sample_info: <...>\n",
      "\t[Paraments]cell_line_sample_dict: <...>\n",
      "\t[Paraments]tab_level: 0\n",
      "[2023-08-05 20:40:33]compute_expression_for_each_cell_line finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:33\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: all\n",
      "\t[Paraments]transcript_biotype: ['all']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:33\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: all\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:33]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and all gene: 8674\n",
      "\t[Result]the counts of gene about ATSS-APA and all gene and ['all'] transcript: 8674\n",
      "\t[Result]the counts of transcript about ATSS-APA and all gene and ['all'] transcript: 78267\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_14036\\1927043246.py:256: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_14036\\1927043246.py:257: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n",
      "C:\\Users\\Administrator\\AppData\\Local\\Temp\\ipykernel_14036\\1927043246.py:258: SettingWithCopyWarning:\n",
      "\n",
      "\n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and all gene and ['all'] transcript: 8650\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and all gene and ['all'] transcript: 66017\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:34\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:34]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:34]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:34\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: all\n",
      "\t[Paraments]transcript_biotype: ['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:34\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: all\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:34]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and all gene: 8674\n",
      "\t[Result]the counts of gene about ATSS-APA and all gene and <transcript_biotype> transcript: 7741\n",
      "\t[Result]the counts of transcript about ATSS-APA and all gene and <transcript_biotyoe> transcript: 17920\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and all gene and <transcript_biotype> transcript: 5325\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and all gene and <transcript_biotype> transcript: 8967\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:35\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:35]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:35]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:35\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: protein_coding\n",
      "\t[Paraments]transcript_biotype: ['all']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:35\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: protein_coding\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:35]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and protein_coding gene: 8074\n",
      "\t[Result]the counts of gene about ATSS-APA and protein_coding gene and ['all'] transcript: 8074\n",
      "\t[Result]the counts of transcript about ATSS-APA and protein_coding gene and ['all'] transcript: 75225\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and protein_coding gene and ['all'] transcript: 8052\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and protein_coding gene and ['all'] transcript: 63168\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:35\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:35]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:35]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:35\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: protein_coding\n",
      "\t[Paraments]transcript_biotype: ['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:35\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: protein_coding\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:35]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and protein_coding gene: 8074\n",
      "\t[Result]the counts of gene about ATSS-APA and protein_coding gene and <transcript_biotype> transcript: 7533\n",
      "\t[Result]the counts of transcript about ATSS-APA and protein_coding gene and <transcript_biotyoe> transcript: 17513\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and protein_coding gene and <transcript_biotype> transcript: 5147\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and protein_coding gene and <transcript_biotype> transcript: 8686\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:36]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:36\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: non_protein_coding\n",
      "\t[Paraments]transcript_biotype: ['all']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: non_protein_coding\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and non_protein_coding gene: 0\n",
      "\t[Result]the counts of gene about ATSS-APA and non_protein_coding gene and ['all'] transcript: 0\n",
      "\t[Result]the counts of transcript about ATSS-APA and non_protein_coding gene and ['all'] transcript: 0\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and non_protein_coding gene and ['all'] transcript: 0\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and non_protein_coding gene and ['all'] transcript: 0\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:36]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:36\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: non_protein_coding\n",
      "\t[Paraments]transcript_biotype: ['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: non_protein_coding\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and non_protein_coding gene: 0\n",
      "\t[Result]the counts of gene about ATSS-APA and non_protein_coding gene and <transcript_biotype> transcript: 0\n",
      "\t[Result]the counts of transcript about ATSS-APA and non_protein_coding gene and <transcript_biotyoe> transcript: 0\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and non_protein_coding gene and <transcript_biotype> transcript: 0\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and non_protein_coding gene and <transcript_biotype> transcript: 0\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:36]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:36\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: un_classified\n",
      "\t[Paraments]transcript_biotype: ['all']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: un_classified\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and un_classified gene: 187\n",
      "\t[Result]the counts of gene about ATSS-APA and un_classified gene and ['all'] transcript: 187\n",
      "\t[Result]the counts of transcript about ATSS-APA and un_classified gene and ['all'] transcript: 720\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and un_classified gene and ['all'] transcript: 187\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and un_classified gene and ['all'] transcript: 707\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:36]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]get_counts_of_transcript_of_gene start.\n",
      "\t[Time]2023-08-05 20:40:36\n",
      "\t[Paraments]isoform_info_df: <...>\n",
      "\t[Paraments]total_info: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]gene_biotype: un_classified\n",
      "\t[Paraments]transcript_biotype: ['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Paraments]log2FC_cutoff: <...>\n",
      "\t[Paraments]max_num: 5\n",
      "\t[Paraments]tab_level: 0\n",
      "\t[Function]get_ATSS_APA_gene_in_cell_line start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]gene_biotype: un_classified\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]get_ATSS_APA_gene_in_cell_line finished.\n",
      "\t[Result]the counts of gene about ATSS-APA and un_classified gene: 187\n",
      "\t[Result]the counts of gene about ATSS-APA and un_classified gene and <transcript_biotype> transcript: 3\n",
      "\t[Result]the counts of transcript about ATSS-APA and un_classified gene and <transcript_biotyoe> transcript: 8\n",
      "\t[Result]the counts of gene showing significant expression in A673 and Caco-2 about ATSS-APA and un_classified gene and <transcript_biotype> transcript: 3\n",
      "\t[Result]the counts of transcript showing significant expression in A673 and Caco-2 about ATSS-APA and un_classified gene and <transcript_biotype> transcript: 7\n",
      "\t[Function]compute_counts_of_gene start.\n",
      "\t\t[Time]2023-08-05 20:40:36\n",
      "\t\t[Paraments]max_num: 5\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:40:36]compute_counts_of_gene finished.\n",
      "[2023-08-05 20:40:36]get_counts_of_transcript_of_gene finished.\n",
      "\n",
      "\n",
      "[Function]output_check_df start.\n",
      "\t[Time]2023-08-05 20:40:36\n",
      "\t[Paraments]df_dict: <...>\n",
      "\t[Paraments]filename: F:/OneDrive/Master/Project/trans/data/0004_check_df.xlsx\n",
      "\t[Result]sheet_name should be: A673@Caco-2@all@['all']\n",
      "\t[Result]sheet_name is: A673_Caco-2_1_2\n",
      "\t[Result]sheet_name should be: A673@Caco-2@all@['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Result]sheet_name is: A673_Caco-2_3_4\n",
      "\t[Result]sheet_name should be: A673@Caco-2@protein_coding@['all']\n",
      "\t[Result]sheet_name is: A673_Caco-2_5_6\n",
      "\t[Result]sheet_name should be: A673@Caco-2@protein_coding@['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Result]sheet_name is: A673_Caco-2_7_8\n",
      "\t[Result]sheet_name should be: A673@Caco-2@non_protein_coding@['all']\n",
      "\t[Result]sheet_name is: A673_Caco-2_9_10\n",
      "\t[Result]sheet_name should be: A673@Caco-2@non_protein_coding@['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Result]sheet_name is: A673_Caco-2_11_12\n",
      "\t[Result]sheet_name should be: A673@Caco-2@un_classified@['all']\n",
      "\t[Result]sheet_name is: A673_Caco-2_13_14\n",
      "\t[Result]sheet_name should be: A673@Caco-2@un_classified@['snRNA', 'TEC', 'transcribed_unitary_pseudogene', 'transcribed_unprocessed_pseudogene', 'scaRNA', 'snoRNA', 'misc_RNA', 'lncRNA', 'processed_transcript', 'nonsense_mediated_decay', 'protein_coding_CDS_not_defined', 'rRNA_pseudogene', 'retained_intron', 'protein_coding', 'scRNA', 'transcribed_processed_pseudogene', 'non_stop_decay', 'miRNA', 'Mt_tRNA', 'Mt_rRNA', 'unprocessed_pseudogene', 'artifact', 'protein_coding_LoF', 'TR_V_gene', 'processed_pseudogene']\n",
      "\t[Result]sheet_name is: A673_Caco-2_15_16\n",
      "[2023-08-05 20:41:12]output_check_df finished.\n",
      "\n",
      "\n",
      "[Function]output_transcript_gene_plot start.\n",
      "\t[Time]2023-08-05 20:41:12\n",
      "\t[Paraments]result_counts: <...>\n",
      "\t[Paraments]file_path: F:/OneDrive/Master/Project/trans/data/0004_plot/\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:12\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: all\n",
      "\t\t[Paraments]transcript_biotype: all\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:12]make_transcript_gene_plot finished.\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:12\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: all\n",
      "\t\t[Paraments]transcript_biotype: snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Result]filename should be: 0004_A673_Caco-2_all_snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene.svg\n",
      "\t[Result]filename is: 0004_A673_Caco-2_all_1.svg\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:13\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: protein_coding\n",
      "\t\t[Paraments]transcript_biotype: all\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:13\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: protein_coding\n",
      "\t\t[Paraments]transcript_biotype: snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Result]filename should be: 0004_A673_Caco-2_protein_coding_snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene.svg\n",
      "\t[Result]filename is: 0004_A673_Caco-2_protein_coding_1.svg\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:13\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: non_protein_coding\n",
      "\t\t[Paraments]transcript_biotype: all\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:13\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: non_protein_coding\n",
      "\t\t[Paraments]transcript_biotype: snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Result]filename should be: 0004_A673_Caco-2_non_protein_coding_snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene.svg\n",
      "\t[Result]filename is: 0004_A673_Caco-2_non_protein_coding_1.svg\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:13\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: un_classified\n",
      "\t\t[Paraments]transcript_biotype: all\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Function]make_transcript_gene_plot start.\n",
      "\t\t[Time]2023-08-05 20:41:13\n",
      "\t\t[Paraments]df: <...>\n",
      "\t\t[Paraments]cell_line_1: A673\n",
      "\t\t[Paraments]cell_line_2: Caco-2\n",
      "\t\t[Paraments]gene_biotype: un_classified\n",
      "\t\t[Paraments]transcript_biotype: snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene\n",
      "\t\t[Paraments]tab_level: 1\n",
      "\t[2023-08-05 20:41:13]make_transcript_gene_plot finished.\n",
      "\t[Result]filename should be: 0004_A673_Caco-2_un_classified_snRNA&TEC&transcribed_unitary_pseudogene&transcribed_unprocessed_pseudogene&scaRNA&snoRNA&misc_RNA&lncRNA&processed_transcript&nonsense_mediated_decay&protein_coding_CDS_not_defined&rRNA_pseudogene&retained_intron&protein_coding&scRNA&transcribed_processed_pseudogene&non_stop_decay&miRNA&Mt_tRNA&Mt_rRNA&unprocessed_pseudogene&artifact&protein_coding_LoF&TR_V_gene&processed_pseudogene.svg\n",
      "\t[Result]filename is: 0004_A673_Caco-2_un_classified_1.svg\n",
      "[2023-08-05 20:41:13]output_transcript_gene_plot finished.\n",
      "[2023-08-05 20:41:14]All blocks finished.\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # load data\n",
    "    sample_info = def_function.load_sample_info(filename=input_sample_info_filename, tab_level=0)\n",
    "    total_info = pandas.read_csv(input_df_filename, sep='\\t', index_col=0)\n",
    "    isoform_info_df = pandas.read_csv(input_TSS_TES_filename, sep='\\t')\n",
    "    # 根据ATSS_APA_cutoff对ATSS-APA型gene进行过滤\n",
    "    readyToDelIndex = isoform_info_df.query(\"gene_classified=='ATSS-APA' & transcript_counts<@ATSS_APA_cutoff\").index\n",
    "    isoform_info_df = isoform_info_df.drop(index=readyToDelIndex).copy()\n",
    "    \n",
    "    # output check plot to confirm the relationship of length and counts\n",
    "    output_length_counts_check_plot(sample_info=sample_info,\n",
    "                                    total_info=total_info,\n",
    "                                    output_filename=output_length_counts_plot_filename,\n",
    "                                    tab_level=0)\n",
    "\n",
    "\n",
    "    cell_line_sample_dict = def_function.get_cell_line_sample_dict(sample_info=sample_info,\n",
    "                                                                   GEO_index=False,\n",
    "                                                                   tab_level=0)\n",
    "\n",
    "    # compute expression for each transcript\n",
    "    total_info = compute_expression_for_each_cell_line(total_info=total_info,\n",
    "                                                       sample_info=sample_info,\n",
    "                                                       cell_line_sample_dict=cell_line_sample_dict,\n",
    "                                                       tab_level=0)\n",
    "\n",
    "\n",
    "    # compare expression of two cell line\n",
    "    # and filter transcripts according to log2FC\n",
    "    transcript_biotype_remove_unclassified = list(set(total_info[\"transcript_biotype\"]))\n",
    "    transcript_biotype_remove_unclassified.remove(\"un_classified\")\n",
    "    cell_line_list = list(cell_line_sample_dict.keys())\n",
    "\n",
    "    result_df = {}  # {\"<cell_list_1>/<cell_list_2>_<gene_biotype>/<transcript_biotype>\": <check_df>, ...}\n",
    "    result_counts = {}  # {\"<cell_line_1>/<cell_line_2>\": {\"<gene_biotype>/<transcript_biotype>\": {0: <num>,\n",
    "                        #                                                                          1: <num>, ...},\n",
    "                        #                                  ...},\n",
    "                        #  ...}\n",
    "    for cell_line_1 in range(0,len(cell_line_list)-1):\n",
    "        for cell_line_2 in range(cell_line_1+1, len(cell_line_list)):\n",
    "            cell_line_1 = cell_line_list[cell_line_1]\n",
    "            cell_line_2 = cell_line_list[cell_line_2]\n",
    "            result_counts[\"{}@{}\".format(cell_line_1, cell_line_2)] = {}\n",
    "            for gene_biotype in [\"all\", \"protein_coding\", \"non_protein_coding\", \"un_classified\"]:\n",
    "                for transcript_biotype in [[\"all\"], transcript_biotype_remove_unclassified]:\n",
    "                    temp_result = get_counts_of_transcript_of_gene(isoform_info_df=isoform_info_df, total_info=total_info,\n",
    "                                                                   cell_line_1=cell_line_1, cell_line_2=cell_line_2,\n",
    "                                                                   gene_biotype=gene_biotype, transcript_biotype=transcript_biotype,\n",
    "                                                                   log2FC_cutoff=log2FC_cutoff, max_num=plot_x_maxnum,\n",
    "                                                                   tab_level=0)\n",
    "                    result_df[\"{}@{}@{}@{}\".format(cell_line_1, cell_line_2, gene_biotype, transcript_biotype)] = temp_result[\"check_df\"]\n",
    "                    result_counts[\"{}@{}\".format(cell_line_1, cell_line_2)][\"{}/{}\".format(gene_biotype, '&'.join(transcript_biotype))] = temp_result[\"result\"]\n",
    "\n",
    "    # output file\n",
    "    output_check_df(df_dict=result_df, filename=output_check_df_filename)\n",
    "    output_transcript_gene_plot(result_counts=result_counts, file_path=output_transcript_gene_plot_path)\n",
    "    total_info.to_csv(output_relative_expression_filename, sep='\\t', index=None)\n",
    "\n",
    "    print(\"[{}]All blocks finished.\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "1.对每一个样本中的transcript的表达量进行标准化\n",
    "  2.1不确定原始数据的rep是如何得到的, 不知道应该用什么计算方法\n",
    "  2.2姑且认为rep值已经对基因长度进行了校正, 因为rep值将近80%都<=10, 值很低\n",
    "  2.3所以认为 一个transcript在一个样本中的相对表达量 = 该transcript的rep值 / 该样本所有transcript的rep值的和\n",
    "  2.4再计算该transcript在相同疾病类型中的不同样本中的平均表达量= 该transcript在一种类型疾病中所有样本的相对表达量的加和 / 该类型疾病的样本数\n",
    "  2.4再了解一下rep值与transcript长度是否相关\n",
    "2.筛选得到在指定的两种疾病类型的样本中都存在表达ATSS-APA基因\n",
    "  2.1根据所有样本的数据筛选ATSS-APA基因id\n",
    "  2.2获取每种疾病中存在表达的基因id, 以list形式存储\n",
    "  2.3分别筛选出在单一一种疾病类型中表达的基因，然后将这些基因合并后去重\n",
    "3.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "调试"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "调试\n",
    "\n",
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
