{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import argparse\n",
    "import time\n",
    "from def_class import *\n",
    "import def_function\n",
    "import pandas\n",
    "import numpy\n",
    "import math\n",
    "import pickle\n",
    "import plotly.graph_objects as go\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "__version__ = \"V1.0(Editor) 2023-08-17\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数-debug\n",
    "cellLine1 = \"A673\"\n",
    "cellLine2 = \"Caco-2\"\n",
    "geneId = None\n",
    "geneName = None\n",
    "cutoffValue = 1.5\n",
    "absolute_path = False\n",
    "file_path = \"F:/OneDrive/Master/Project/trans/data/\"\n",
    "input_pickle_filename = \"0002_total_info.pkl\"\n",
    "input_transcript_expression_filename = \"0004_relative_expression.tsv\"\n",
    "output_plot_path = \"0005_plot/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "usage: ipykernel_launcher.py [-h] --cellLine1 CELLLINE1 --cellLine2 CELLLINE2\n",
      "                             [--geneId GENEID] [--geneName GENENAME]\n",
      "                             [--cutoffValue CUTOFFVALUE] [--absolute_path]\n",
      "                             [--file_path FILE_PATH]\n",
      "                             [--input_pickle_filename INPUT_PICKLE_FILENAME]\n",
      "                             [--input_transcript_expression_filename INPUT_TRANSCRIPT_EXPRESSION_FILENAME]\n",
      "                             [--output_plot_path OUTPUT_PLOT_PATH]\n",
      "ipykernel_launcher.py: error: the following arguments are required: --cellLine1, --cellLine2\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "2",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[1;31mSystemExit\u001b[0m\u001b[1;31m:\u001b[0m 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\Administrator\\AppData\\Local\\Programs\\Python\\Python310\\lib\\site-packages\\IPython\\core\\interactiveshell.py:3406: UserWarning:\n",
      "\n",
      "To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 前置参数\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"--cellLine1\", dest=\"cellLine1\", required=True, type=str, default=\"\", help=\"the cellLine 1\")\n",
    "parser.add_argument(\"--cellLine2\", dest=\"cellLine2\", required=True, type=str, default=\"\", help=\"the cellLine 2\")\n",
    "parser.add_argument(\"--geneId\", dest=\"geneId\", required=False, type=str, default=None, help=\"=None, the gene need to be displayed\")\n",
    "parser.add_argument(\"--geneName\", dest=\"geneName\", required=False, type=str, default=None, help=\"=None, the gene need to be displayed\")\n",
    "parser.add_argument(\"--cutoffValue\", dest=\"cutoffValue\", required=False, type=float, default=1.5, help=\"=1.5, the cutoff value of log2FC about the significant expression\")\n",
    "parser.add_argument(\"--absolute_path\", dest=\"absolute_path\", action=\"store_true\", help=\"use the absolute path\")\n",
    "parser.add_argument(\"--file_path\", dest=\"file_path\", required=False, type=str, default=\"\", help=\"the path of data file\")\n",
    "parser.add_argument(\"--input_pickle_filename\", dest=\"input_pickle_filename\", required=False, type=str, default=\"0002_total_info.pkl\", help=\"=0002_total_info.pkl, the output file of 0002.py\")\n",
    "parser.add_argument(\"--input_transcript_expression_filename\", dest=\"input_transcript_expression_filename\", required=False, type=str, default=\"0004_relative_expression.tsv\", help=\"=0004_relative_expression.tsv, the output file of 0004.py\")\n",
    "parser.add_argument(\"--output_plot_path\", dest=\"output_plot_path\", required=False, type=str, default=\"0005_plot/\", help=\"=0005_plot/, the output path of the plot\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "cellLine1 = args.cellLine1\n",
    "cellLine2 = args.cellLine2\n",
    "geneId = args.geneId\n",
    "geneName = args.geneName\n",
    "cutoffValue = args.cutoffValue\n",
    "absolute_path = args.absolute_path\n",
    "file_path = args.file_path\n",
    "input_pickle_filename = args.input_pickle_filename\n",
    "input_transcript_expression_filename = args.input_transcript_expression_filename\n",
    "output_plot_path = args.output_plot_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补全路径\n",
    "if absolute_path is False:\n",
    "    input_pickle_filename = \"{}{}\".format(file_path, input_pickle_filename)\n",
    "    input_transcript_expression_filename = \"{}{}\".format(file_path, input_transcript_expression_filename)\n",
    "    output_plot_path = \"{}{}\".format(file_path, output_plot_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 打印参数\n",
    "print('\\n')\n",
    "print(\"[Date]{}\".format(time.strftime(\"%Y-%m-%d, %H:%M:%S\", time.localtime(time.time()))))\n",
    "print(\"[Script]{}\".format(__file__))\n",
    "print(\"[Version]{}\".format(__version__))\n",
    "print(\"[Version]def_function: {}\".format(def_function.__version__))\n",
    "print(\"[Version]def_class: {}\".format(class_version))\n",
    "print(\"[Parament]cellLine1: {}\".format(cellLine1))\n",
    "print(\"[Parament]cellLine2: {}\".format(cellLine2))\n",
    "print(\"[Parament]geneId: {}\".format(geneId))\n",
    "print(\"[Parament]geneName: {}\".format(geneName))\n",
    "print(\"[Parament]cutoffValue: {}\".format(cutoffValue))\n",
    "print(\"[Parament]input_pickle_filename: {}\".format(input_pickle_filename))\n",
    "print(\"[Parament]input_transcript_expression_filename: {}\".format(input_transcript_expression_filename))\n",
    "print(\"[Parament]output_plot_path: {}\".format(output_plot_path))\n",
    "print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 隐藏函数\n",
    "def select_gene_to_show(df, cell_line_1, cell_line_2, cutoff_value=1, reset_value=0.1, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df, pandas.DataFrame, expression_df\n",
    "        cell_line_1, str, ...\n",
    "        cell_line_2, str, ...\n",
    "        cutoff_value, int or float, gene中必须含有在两种组织类型中特异性表达的transcript, 该阈值即为衡量特异性表达的阈值\n",
    "        reset_value, float, 若log2FC为numpy.nan, 则将nan修改为该值\n",
    "    change:\n",
    "        挑选可以展示的gene\n",
    "        选择这样的gene, 这个gene存在一些transcript分别在两种组织类型中特异性表达\n",
    "    output:\n",
    "        selected_gene, list, 建议展示的gene\n",
    "    \"\"\"\n",
    "    expression_df = df\n",
    "    cellline1 = cell_line_1\n",
    "    cellline2 = cell_line_2\n",
    "    cutoff_value = cutoff_value\n",
    "    reset_value = 0.1\n",
    "\n",
    "    selected_gene = []  # [[<gene_id>, <the sum of abs(log2FC) of the gene], ...]\n",
    "\n",
    "    expression_df_group = expression_df.groupby(\"gene_id\")\n",
    "    for gene_id in expression_df_group.groups.keys():\n",
    "        temp = expression_df_group.get_group(gene_id)\n",
    "        temp = temp[\"log2FC_{}@{}\".format(cellline1, cellline2)].to_list()\n",
    "        temp = [reset_value if numpy.isnan(i) else i for i in temp]\n",
    "        if any([i>cutoff_value for i in temp]) and any([i<-cutoff_value for i in temp]):\n",
    "            temp = [abs(i) for i in temp]\n",
    "            temp = sum(temp)\n",
    "            selected_gene.append([gene_id, temp])\n",
    "        else:\n",
    "            continue\n",
    "    selected_gene = sorted(selected_gene, key=lambda x: x[1], reverse=True)\n",
    "    #selected_gene = selected_gene[0][0]\n",
    "    selected_gene = [i[0] for i in selected_gene]\n",
    "\n",
    "    return selected_gene\n",
    "\n",
    "\n",
    "# 隐藏函数\n",
    "def exonRange_to_dict(exonRange, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        exonRange, dict, {\"<exon_start>-<exon_end>\": <relative expression>, ...}\n",
    "    change:\n",
    "        能够将exon_expression_in_cellline1中的exon_range转换为数字\n",
    "        将每一个exon_range拆分为range(<exon_start>, <exon_end>+1, by=1), 存储到df中, 对应位置的值即为含有该位点的exon的expression的加和\n",
    "    output:\n",
    "        exon_site_expression, dict, {<position in sequence>: <expression>, ...}\n",
    "    \"\"\"\n",
    "    exonRange_dict = exonRange\n",
    "\n",
    "    exon_site_expression = {}  # {<int=position>: <float=expression>, ...}\n",
    "    for exon_range, exon_expression in exonRange_dict.items():\n",
    "        exon_range = [int(i) for i in exon_range.split('-')]\n",
    "        exon_start = exon_range[0]\n",
    "        exon_end = exon_range[1]\n",
    "        \n",
    "        # add position value\n",
    "        for position in range(exon_start, exon_end+1, 1):\n",
    "            temp = exon_site_expression.get(position, 0)\n",
    "            temp = temp + exon_expression\n",
    "            exon_site_expression[position] = temp\n",
    "    return exon_site_expression\n",
    "\n",
    "\n",
    "def classify_significant_transcript(df, geneId, cutoffValue, cellLine1, cellLine2, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df, pandas.DataFrame, expression_df\n",
    "        geneId, str, 要判断的gene_id\n",
    "        cutoffValue, float, 判断transcript的显著性时用到的log2FC阈值\n",
    "        cellLine1, str, ...\n",
    "        cellLine2, str, ...\n",
    "    change:\n",
    "        判断指定geneId的transcript是否于cellLine中显著表达\n",
    "        根据log2FC对transcript进行分类\n",
    "            1. 声明一个dict, 以存储在cellLine1或cellLine2中特异性表达以及非特异性表达的transcriptId\n",
    "                {\"<cellLine1>\": [<transcriptId>, ...], \"<cellLine2>\": [<transcriptId>, ...], \"other\": [<transcriptId>, ...]}\n",
    "            2. 根据log2FC对transcript的特异性表达进行判断\n",
    "                若log2FC>=cutoff_value, 则于cellLine1中特异性表达, 若log2FC<=-cutoff_value, 则于cellLine2中特异性表达\n",
    "                若-cutoff_value<log2FC<cutoff_value, 则无特异性表达\n",
    "                !!!若log2FC is numpy.nan, 则判断cellLine1的相对表达值是否为0, 若不为0则于cellLine1特异性表达, 否则再判断cellLine2的相对表达值是否为0, 若不为0则于cellLine2特异性表达\n",
    "    output:\n",
    "        significant_info, dict, {<cellLine1>: [<transcript_id>, ...],\n",
    "                                 <cellLine2>: [<transcript_id>, ...],\n",
    "                                 \"other\": [<transcript_id>, ...]}\n",
    "    \"\"\"\n",
    "    df = df\n",
    "    geneId = geneId\n",
    "    cutoffValue = cutoffValue\n",
    "    cellLine1 = cellLine1\n",
    "    cellLine2 = cellLine2\n",
    "\n",
    "    temp_df = df.loc[df[\"gene_id\"]==geneId, :]\n",
    "    significant_info = {cellLine1: [], cellLine2: [], \"other\": []}\n",
    "\n",
    "    for i in temp_df.index:\n",
    "        transcript_id = temp_df.at[i, \"transcript_id\"]\n",
    "        log2FC = temp_df.at[i, \"log2FC_{}@{}\".format(cellLine1, cellLine2)]\n",
    "\n",
    "        if log2FC >= cutoffValue:\n",
    "            significant_info[cellLine1].append(transcript_id)\n",
    "        elif log2FC <= -cutoffValue:\n",
    "            significant_info[cellLine2].append(transcript_id)\n",
    "        else:\n",
    "            significant_info[\"other\"].append(transcript_id)\n",
    "    \n",
    "    return significant_info\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 功能函数\n",
    "def getGeneId(total, geneId, geneName, df, cellLine1, cellLine2):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        total, Total Object\n",
    "        geneId, str/None\n",
    "        geneName, str/None\n",
    "    change:\n",
    "        根据给定的geneId或geneName, 返回geneId\n",
    "        若geneId与geneName都为None, 则根据规则挑一个geneId出来\n",
    "    output:\n",
    "        geneId, str\n",
    "    \"\"\"\n",
    "    total = total\n",
    "    geneId = geneId\n",
    "    geneName = geneName\n",
    "    df = df\n",
    "    cellLine1 = cellLine1\n",
    "    cellLine2 = cellLine2\n",
    "\n",
    "    if geneId is None:\n",
    "        if geneName is None: \n",
    "            selected_gene_id = select_gene_to_show(df=df,\n",
    "                                                   cell_line_1=cellLine1,\n",
    "                                                   cell_line_2=cellLine2,\n",
    "                                                   cutoff_value=1.5,\n",
    "                                                   reset_value=0.1)\n",
    "            geneId = selected_gene_id[0]\n",
    "        else:\n",
    "            for geneId, geneObject in total.gene_dict.items():\n",
    "                if geneObject.gene_name == geneName:\n",
    "                    return geneId\n",
    "            raise ValueError(\"[Error]The geneName is not existed in the data.\")\n",
    "    else:\n",
    "        geneId = geneId\n",
    "\n",
    "    return geneId\n",
    "\n",
    "\n",
    "def getExonExpression(total, geneId, cellLine):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        total, Total Object\n",
    "        geneId, str\n",
    "        cellLine, str\n",
    "    change:\n",
    "        获取在指定gene的指定cellLine中存在表达的exon的范围以及表达量\n",
    "    output:\n",
    "        dict, {\"<geneId>\": {\"<exon1_start>-<exon1_end>\": <value>,\n",
    "                            \"<exon2_start>-<exon2_end>\": <value>, ...}}\n",
    "    \"\"\"\n",
    "    total = total\n",
    "    geneId = geneId\n",
    "    cellLine = cellLine\n",
    "\n",
    "    exonExpressionInfo = total.gene_dict[geneId].exonExpression\n",
    "\n",
    "    tempDict = {}\n",
    "    for exon, exonDict in exonExpressionInfo.items():\n",
    "        cellLineExpression = exonDict[\"cellLineExpression\"].get(cellLine, 0)\n",
    "        if cellLineExpression != 0:\n",
    "            tempDict[exon] = cellLineExpression\n",
    "\n",
    "    return {geneId: tempDict}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 绘图函数\n",
    "def make_gene_model_plot(exon_combination, gene_id, gene_name, gene_strand, gene_start, gene_end, tab_level=0):\n",
    "    exon_combination = exon_combination\n",
    "    gene_id = gene_id\n",
    "    gene_name = gene_name\n",
    "    gene_strand = gene_strand\n",
    "    gene_start = gene_start\n",
    "    gene_end = gene_end\n",
    "\n",
    "    result = {}  # {\"fig\": <fig>, \"xaxis_min\": <xaxis_min>, \"xaxis_max\": <xaxis_max>}\n",
    "\n",
    "    # 获取指定gene的所有exon的start, end, length信息\n",
    "    exon_df = pandas.DataFrame(columns=[\"exon_start\", \"exon_end\", \"exon_length\"])\n",
    "    exon_start = []\n",
    "    exon_end = []\n",
    "    exon_length = []\n",
    "    for exon_range in exon_combination[gene_id]:\n",
    "        exon_range = [int(i) for i in exon_range.split('-')]\n",
    "        exon_start.append(exon_range[0])\n",
    "        exon_end.append(exon_range[1])\n",
    "        exon_length.append(exon_range[1]-exon_range[0])\n",
    "    exon_df[\"exon_start\"] = exon_start\n",
    "    exon_df[\"exon_end\"] = exon_end\n",
    "    exon_df[\"exon_length\"] = exon_length\n",
    "\n",
    "    # 根据gene的strand, 计算exon相对于gene start的位置\n",
    "    if gene_strand =='-':\n",
    "        exon_df[\"exon_start\"] = gene_end - exon_df[\"exon_end\"]\n",
    "        exon_df[\"exon_end\"] = exon_df[\"exon_start\"] + exon_df[\"exon_length\"]\n",
    "    else:\n",
    "        exon_df[\"exon_start\"] = exon_df[\"exon_start\"] - gene_start\n",
    "        exon_df[\"exon_end\"] = exon_df[\"exon_start\"] + exon_df[\"exon_length\"]\n",
    "\n",
    "    # 整理exon_df的格式\n",
    "    # 该exon_df存储了指定gene的每一个exon相对于该gene的起始起点的位置信息(已考虑strand)\n",
    "    exon_df = exon_df.sort_values(by=\"exon_end\")\n",
    "    exon_df = exon_df.sort_values(by=\"exon_start\")\n",
    "    exon_df = exon_df.reset_index(drop=True)\n",
    "\n",
    "    # 获取x轴的范围\n",
    "    xaxis_min = min(exon_df[\"exon_start\"])\n",
    "    xaxis_max = max(exon_df[\"exon_end\"])\n",
    "    result[\"xaxis_min\"] = xaxis_min\n",
    "    result[\"xaxis_max\"] = xaxis_max\n",
    "\n",
    "    # 将exon的range转换为柱状图的坐标\n",
    "    fig_x = []\n",
    "    for i in exon_df.index:\n",
    "        temp = list(range(exon_df.at[i, \"exon_start\"], exon_df.at[i, \"exon_end\"]+1))\n",
    "        fig_x = fig_x + temp\n",
    "    fig_x = list(set(fig_x))\n",
    "\n",
    "    # 根据exon_df绘制结构图Barplot\n",
    "    fig = go.Figure()\n",
    "    fig_trace = go.Bar(x=fig_x,\n",
    "                    y=[10]*len(fig_x),\n",
    "                    width=1,\n",
    "                    marker={\"color\": \"black\",\n",
    "                            \"line\": {\"color\": \"black\",\n",
    "                                        \"width\": 0}\n",
    "                            })\n",
    "    fig = fig.add_shape(type=\"line\",\n",
    "                        x0=xaxis_min, y0=5, x1=xaxis_max, y1=5,\n",
    "                        line={\"color\": \"black\",\n",
    "                            \"width\": 1})\n",
    "    fig = fig.add_trace(fig_trace)\n",
    "    layout = {\"width\": 700, \"height\": 50,\n",
    "              \"title\": {\"text\": \"{} gene model\".format(gene_name),\n",
    "                        \"font\": {\"family\": \"Arial\",\n",
    "                                 \"color\": \"black\",\n",
    "                                 \"size\": 12,},\n",
    "                        \"x\": 0.1},\n",
    "            \"margin\": {'l':15, 'r':15, 't':25, 'b':15}, \n",
    "            \"xaxis\": {\"showline\": False,\n",
    "                        \"showticklabels\": False,\n",
    "                        \"range\": [xaxis_min, xaxis_max],\n",
    "                        \"showgrid\": False,},\n",
    "            \"yaxis\": {\"showline\": False,\n",
    "                        \"showticklabels\": False,\n",
    "                        \"showgrid\": False,\n",
    "                        },\n",
    "            \"paper_bgcolor\": \"white\",\n",
    "            \"plot_bgcolor\":\"white\",\n",
    "            }\n",
    "    fig = fig.update_layout(layout)\n",
    "    result[\"fig\"] = fig\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "def make_exon_plot(exon_cellline_1, exon_cellline_2, gene_id, title_list= [], gene_strand='+', gene_start=0, gene_end=0, xaxis_min=0, xaxis_max=0 ,tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        gene_id, str, ...\n",
    "        exon_cellline_1, dict, 在cell line 1中每个基因的每个exon的相对表达值\n",
    "        exon_cellline_2, dict, 在cell line 2中每个基因的每个exon的相对表达值\n",
    "        title_list, list, [<the name of cell line 1>, <the name of cell line 2>]\n",
    "        gene_strand, str, 指定gene的+/-链\n",
    "        gene_start, int, 指定gene的start\n",
    "        gene_end, int, 指定gene的end\n",
    "        xaxis_min, int, the min value of xaxis\n",
    "        xaxix_max, int, the max value of xaxis\n",
    "    \"\"\"\n",
    "    gene_id = gene_id\n",
    "    exon_cellline_1 = exon_cellline_1\n",
    "    exon_cellline_2 = exon_cellline_2\n",
    "    title_list = title_list\n",
    "    gene_strand = gene_strand\n",
    "    gene_start = gene_start\n",
    "    gene_end = gene_end\n",
    "    xaxis_min = xaxis_min\n",
    "    xaxis_max = xaxis_max\n",
    "\n",
    "    # 准备绘图所用的数据\n",
    "    fig1_xy = exonRange_to_dict(exon_cellline_1[gene_id], tab_level=tab_level+1)\n",
    "    fig1_x = []\n",
    "    fig1_y = []\n",
    "    # 根据链的+/-修改position为相对于gene第一个exon起始位点的位置\n",
    "    fig1_position = {}\n",
    "    if gene_strand == \"-\":\n",
    "        for position, expression in fig1_xy.items():\n",
    "            fig1_position[gene_end-position] = expression\n",
    "    else:\n",
    "        for position, expression in fig1_xy.items():\n",
    "            fig1_position[position-gene_start] = expression\n",
    "    for position, expression in fig1_position.items():\n",
    "        fig1_x.append(position)\n",
    "        fig1_y.append(expression)\n",
    "\n",
    "\n",
    "    fig2_xy = exonRange_to_dict(exon_cellline_2[gene_id], tab_level=tab_level+1)\n",
    "    fig2_x = []\n",
    "    fig2_y = []\n",
    "    # 根据链的+/-修改position为相对于gene第一个exon起始位点的位置\n",
    "    fig2_position = {}\n",
    "    if gene_strand == \"-\":\n",
    "        for position, expression in fig2_xy.items():\n",
    "            fig2_position[gene_end-position] = expression\n",
    "    else:\n",
    "        for position, expression in fig2_xy.items():\n",
    "            fig2_position[position-gene_start] = expression\n",
    "    for position, expression in fig2_position.items():\n",
    "        fig2_x.append(position)\n",
    "        fig2_y.append(expression)\n",
    "\n",
    "    # 设定y轴的最大范围\n",
    "    y_max = max(fig1_y+fig2_y)\n",
    "\n",
    "    # 开始绘图\n",
    "    fig1 = go.Figure()\n",
    "    fig1_trace = go.Bar(x=fig1_x,\n",
    "                        y=fig1_y,\n",
    "                        width=1,\n",
    "                        marker={\"color\": \"indianred\",\n",
    "                                \"line\": {\"width\": 0}})\n",
    "    fig2 = go.Figure()\n",
    "    fig2_trace = go.Bar(x=fig2_x,\n",
    "                        y=fig2_y,\n",
    "                        width=1,\n",
    "                        marker={\"color\": \"lightblue\",\n",
    "                                \"line\": {\"width\": 0}})\n",
    "\n",
    "    layout = {\"width\": 700, \"height\": 100,\n",
    "              \"title\": {\"text\": \"cell_line\",  # 在后面会更改, 这一行不必修改\n",
    "                        \"font\": {\"family\": \"Arial\",\n",
    "                                \"color\": \"black\",\n",
    "                                \"size\": 12},\n",
    "                        'x': 0.1,},\n",
    "              \"margin\": {'l':15, 'r':15, 't':25, 'b':15}, \n",
    "              \"barmode\": \"stack\",\n",
    "              \"font\": {\"family\": \"Arial\",\n",
    "                       \"color\": \"black\",\n",
    "                       \"size\": 10},\n",
    "              \"xaxis\": {\"title\": {\"text\": \"\",\n",
    "                                  \"font\": {\"family\": \"Arial\",\n",
    "                                        \"color\": \"black\",\n",
    "                                        \"size\": 12},\n",
    "                                \"standoff\": 0.1,\n",
    "                                },\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"linewidth\": 1,\n",
    "                        \"showticklabels\": False,\n",
    "                        \"range\": [xaxis_min, xaxis_max]},\n",
    "            \"yaxis\": {\"title\": {\"text\": \"\",\n",
    "                                \"font\": {\"family\": \"Arial\",\n",
    "                                         \"color\": \"black\",\n",
    "                                         \"size\": 12},\n",
    "                                \"standoff\": 0,},\n",
    "                        \"showline\": False,\n",
    "                        \"range\": [0, y_max],\n",
    "                        \"showgrid\": False,\n",
    "                        \"showticklabels\": False,\n",
    "                        },\n",
    "            \"plot_bgcolor\": \"white\",\n",
    "            \"bargap\": 0, \"bargroupgap\": 0,}\n",
    "\n",
    "    fig1 = fig1.update_layout(layout)\n",
    "    fig2 = fig2.update_layout(layout)\n",
    "\n",
    "    fig1 = fig1.update_layout({\"title\": {\"text\": title_list[0]}})\n",
    "    fig2 = fig2.update_layout({\"title\": {\"text\": title_list[1]}})\n",
    "\n",
    "    fig1 = fig1.add_trace(fig1_trace)\n",
    "    fig2 = fig2.add_trace(fig2_trace)\n",
    "\n",
    "    return [fig1, fig2]\n",
    "\n",
    "\n",
    "def make_transcript_plot(df, gene_id, cutoff_value, cell_line_1, cell_line_2, gene_strand, gene_start, gene_end, xaxis_min, xaxis_max, tab_level=0):\n",
    "    df = df\n",
    "    gene_id = gene_id\n",
    "    cutoff_value = cutoff_value\n",
    "    cell_line_1 = cell_line_1\n",
    "    cell_line_2 = cell_line_2\n",
    "    gene_strand = gene_strand\n",
    "    gene_start = gene_start\n",
    "    gene_end = gene_end\n",
    "    xaxis_min = xaxis_min\n",
    "    xaxis_max = xaxis_max\n",
    "\n",
    "    # 将transcript分为三种类型: 于cellline1特异性表达，于cellline2特异性表达，非特异性表达\n",
    "    significant_info = classify_significant_transcript(df=df, geneId=gene_id, cutoffValue=cutoff_value,\n",
    "                                                       cellLine1=cell_line_1, cellLine2=cell_line_2,\n",
    "                                                       tab_level=tab_level+1)\n",
    "    # 设定三种类型transcript的颜色\n",
    "    color_dict = {cell_line_1: \"indianred\", cell_line_2: \"lightblue\", \"other\": \"grey\"}\n",
    "\n",
    "    # 获取指定gene的全部transcript的信息\n",
    "    temp_df = df.loc[df[\"gene_id\"]==gene_id, :].copy()\n",
    "\n",
    "    # 补充transcript的start及end信息\n",
    "    temp_df[\"transcript_start\"] = temp_df[\"transcript_id\"].map(lambda x: total.gene_dict[gene_id].transcript_dict[x][\"range\"][0])\n",
    "    temp_df[\"transcript_end\"] = temp_df[\"transcript_id\"].map(lambda x: total.gene_dict[gene_id].transcript_dict[x][\"range\"][1])\n",
    "\n",
    "    num=1\n",
    "    fig = go.Figure()\n",
    "    # 绘制在cell_line特异性表达以及非特异性表达的trace\n",
    "    for name in [cell_line_2, \"other\", cell_line_1]:\n",
    "        significant_id_list = significant_info[name]\n",
    "        temp_df_name = temp_df.loc[temp_df[\"transcript_id\"].isin(significant_id_list), :].copy()\n",
    "\n",
    "        # 根据strand的方向，修改transcript的start与end为相对于gene的起点的位置\n",
    "        if gene_strand == '-':\n",
    "            temp_df_name[\"base\"] = gene_end - temp_df_name[\"transcript_end\"]\n",
    "        else:\n",
    "            temp_df_name[\"base\"] = temp_df_name[\"transcript_start\"] - gene_start\n",
    "        temp_df_name[\"length\"] = temp_df_name[\"transcript_end\"] - temp_df_name[\"transcript_start\"]\n",
    "        # 整理transcript的排序\n",
    "        temp_df_name = temp_df_name.sort_values(by=\"length\", ascending=False)\n",
    "        temp_df_name = temp_df_name.sort_values(by=\"base\", ascending=False)\n",
    "\n",
    "        fig_x = [i for i in range(num, num+temp_df_name.shape[0])]\n",
    "        num = num + temp_df_name.shape[0]\n",
    "        fig_trace = go.Bar(x=temp_df_name[\"length\"],\n",
    "                        y=fig_x,\n",
    "                        base=temp_df_name[\"base\"],\n",
    "                        name=name,\n",
    "                        width=1,\n",
    "                        orientation='h',\n",
    "                        marker={\"color\": color_dict[name]})\n",
    "        fig = fig.add_trace(fig_trace)\n",
    "\n",
    "    layout = {\"width\":700, \"height\":200,\n",
    "                \"title\": {\"text\": \"5'-3' isoforms count\",\n",
    "                            \"font\": {\"family\": \"Arial\",\n",
    "                                \"size\": 12,\n",
    "                                \"color\": \"black\"},\n",
    "                            'x': 0.1},\n",
    "                \"margin\": {'l':15, 'r':15, 't':25, 'b':15}, \n",
    "                \"xaxis\": {\"showline\": False,\n",
    "                            \"showticklabels\": False,\n",
    "                            \"range\": [xaxis_min, xaxis_max],\n",
    "                            \"showgrid\": False,},\n",
    "                \"yaxis\": {\"showline\": False,\n",
    "                            \"showticklabels\": False,\n",
    "                            \"showgrid\": False,\n",
    "                            },\n",
    "                \"showlegend\": False,\n",
    "                \"paper_bgcolor\": \"white\",\n",
    "                \"plot_bgcolor\": \"white\",\n",
    "                }\n",
    "    fig = fig.update_layout(layout)\n",
    "\n",
    "    return fig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t[Reseult]selected gene id is: ENCLB767XEKG000058938, gene name isENCLB767XEKG000058938\n",
      "[2023-08-17 20:24:32]All blocks finished.\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # load data\n",
    "    with open(input_pickle_filename, 'rb') as file:\n",
    "        total = pickle.load(file)\n",
    "    expression_df = pandas.read_csv(input_transcript_expression_filename, sep='\\t', index_col=None)\n",
    "    \n",
    "    # get exon combination for each gene\n",
    "    exon_combination = total.get_exon_combination()\n",
    "\n",
    "    # 获取展示用的gene_id\n",
    "    geneId = getGeneId(total=total, geneId=geneId, geneName=geneName,\n",
    "                       df=expression_df, cellLine1=cellLine1, cellLine2=cellLine2)\n",
    "    geneName = total.gene_dict[geneId].gene_name\n",
    "    print(\"\\t[Reseult]selected gene id is: {}, gene name is: {}\".format(geneId, geneName))\n",
    "\n",
    "    # 获取展示用的gene的strand, start, end信息\n",
    "    geneStrand = total.gene_dict[geneId].strand\n",
    "    geneStart = total.gene_dict[geneId].start\n",
    "    geneEnd = total.gene_dict[geneId].end\n",
    "\n",
    "    # make gene model plot\n",
    "    gene_model = make_gene_model_plot(exon_combination=exon_combination,\n",
    "                                      gene_id=geneId, gene_name=geneName, gene_strand=geneStrand,\n",
    "                                      gene_start=geneStart, gene_end=geneEnd, tab_level=0)\n",
    "    xaxis_min = gene_model[\"xaxis_min\"]\n",
    "    xaxis_max = gene_model[\"xaxis_max\"]\n",
    "    fig0 = gene_model[\"fig\"]\n",
    "\n",
    "    # 获取指定gene在cellLine1及cellLine2中的exon的表达信息\n",
    "    exon_expression_in_cellline1 = getExonExpression(total, geneId, cellLine1)\n",
    "    exon_expression_in_cellline2 = getExonExpression(total, geneId, cellLine2)\n",
    "    # make exon plot\n",
    "    [fig1, fig2] = make_exon_plot(exon_cellline_1=exon_expression_in_cellline1,\n",
    "                                  exon_cellline_2=exon_expression_in_cellline2,\n",
    "                                  title_list= [cellLine1, cellLine2],\n",
    "                                  gene_id=geneId,\n",
    "                                  gene_strand=geneStrand, gene_start=geneStart, gene_end=geneEnd,\n",
    "                                  xaxis_min=xaxis_min, xaxis_max=xaxis_max)\n",
    "    # make transcript plot\n",
    "    fig3 = make_transcript_plot(df=expression_df, gene_id=geneId, cutoff_value=cutoffValue,\n",
    "                                cell_line_1=cellLine1, cell_line_2=cellLine2,\n",
    "                                gene_strand=geneStrand, gene_start=geneStart, gene_end=geneEnd,\n",
    "                                xaxis_min=xaxis_min, xaxis_max=xaxis_max)\n",
    "\n",
    "    # output\n",
    "    fig0.write_image(\"{}0005_{}_gene_model.svg\".format(output_plot_path, geneName))\n",
    "    fig1.write_image(\"{}0005_{}_{}.svg\".format(output_plot_path, geneName, cellLine1))\n",
    "    fig2.write_image(\"{}0005_{}_{}.svg\".format(output_plot_path, geneName, cellLine2))\n",
    "    fig3.write_image(\"{}0005_{}_transcript.svg\".format(output_plot_path, geneName))\n",
    "\n",
    "    print(\"[{}]All blocks finished.\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "end\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "- 运行该脚本前需指定需要分析的两种组织类型\n",
    "- 该脚本分析Total对象中每一个gene分别在两种组织类型中的exon的相对表达量\n",
    "    - 先统计每个gene的exon\n",
    "        - 以dict形式返回gene所含的exon, {\"<gene_id>\": [\"<exon1_start>-<exon1_end>\", ...], ...}\n",
    "    - 再根据0004.tsv中的组织相对表达量, 分别计算exon在两种类型组织中的相对表达量\n",
    "        - 声明一个dict, {\"<gene_id>\": {\"<cell_line_1>\": {\"<exon1_start>-<exon1_end>\": <relative expression>,\n",
    "                                                        \"<exon2_start>-<exon2_end>\": <relative expression>, ...},\n",
    "                                      \"<cell_line_2>\": {\"<exon1_start>-<exon1_end>\": <relative expression>,\n",
    "                                                        \"<exon2_start>-<exon2_end>\": <relative expression>, ...}\n",
    "                                      }\n",
    "                        }\n",
    "        - 写一个函数\n",
    "            - 声明一个dict, {\"<gene_id>\": {\"<exon1_start>-<exon1_end>\": <value>,\n",
    "                                          \"<exon2_start>-<exon2_end>\": <value>, ...\n",
    "                                          }\n",
    "                            }\n",
    "            - 该函数可确定在一种组织类型中存在表达的transcript, 筛选得到这些transcript的信息\n",
    "            - 根据Total.gene_dict, 按照transcript对应的gene_id及transcript_id检索该transcript的exon组成\n",
    "                - 更新该函数中声明的dict, 对应的exon的value = 原value值 + 该transcript的相对表达值\n",
    "        - 在一种组织类型中, 根据transcript的相对表达量，计算exon的相对表达量\n",
    "            - 建立一个dict存储该样本中不同gene的exon的相对表达量, {<gene_id>: {\"<exon-start>,<exon-end>\": <相对表达量>, ...}, ...}\n",
    "            - 在Gene对象中，已记录了该gene的transcript的exon, 对transcript的每一个exon, 取上一步建立的dict中的exon, exon相对表达量=exon相对表达量+transcript的相对表达量\n",
    "            - 最终, dict中存储了一种组织类型中不同gene的不同exon的相对表达量\n",
    "    - 以df保存结果\n",
    "    - 可以以柱状图形式展示exon的相对表达量\n",
    "\n",
    "1. Gene新增方法, 整理该gene的全部exon, 并以dict形式返回{<gene_id_1>: [[<exon_1_start>, <exon_1_end>],\n",
    "                                                                   [<exon_2_start>, <exon_2_end>], ...]\n",
    "                                                                  ],\n",
    "                                                     ...}\n",
    "2. 根据0004.xlsx及0001.tsv文件, 对每一个gene的transcript的特异性表达进行归类，判断transcript富集于两种组织中的哪一种\n",
    "    2.1 根据cell_line_1及cell_line_2及log2FC的正负号进行判断, log2FC>0则表示富集于cell_line_1, log2FC<0则表示富集于cell_line_2\n",
    "3. 对两种cell_line的每一个gene，统计gene中exon的使用次数\n",
    "    3.1 对于cell_line_1的gene, 取在cell_line_1中对应gene中表达的对应transcript。在Gene对象中，已存储了该transcript所对应的exon, 所以 该exon的相对表达量=该exon的相对表达量+该transcript在该组织中的的相对表达量\n",
    "    3.2 可以以柱状图形式呈现一个gene中所有exon的相对表达量\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
