{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import time\n",
    "import argparse\n",
    "import plotly.graph_objects as go\n",
    "import numpy\n",
    "import math\n",
    "import def_function\n",
    "from def_class import *\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "__version__ = \"V2.0(Editor) 2023-08-18\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数-debug\n",
    "infinity_mode = \"del\"  # 对于仅在一种细胞系中表达的transcript的处理策略, del--不展示这种transcript, endPoint--将这种transcript的表达值定为 最大值+1 / 最小值-1\n",
    "cutoff_value = 1.5\n",
    "cell_line_1 = \"A673\"\n",
    "cell_line_2 = \"Caco-2\"\n",
    "gene_name_list = []\n",
    "gene_id_list = []\n",
    "absolute_path = False\n",
    "file_path = \"F:/OneDrive/Master/Project/trans/data/\"\n",
    "input_df_filename = \"0004_relative_expression.tsv\"\n",
    "input_pkl_filename = \"0002_total_info.pkl\"\n",
    "output_plot_filename = \"0006_transcript_expression.svg\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数\n",
    "parser = argparse.ArgumentParser()\n",
    "parser.add_argument(\"--infinity_mode\", dest=\"infinity_mode\", required=False, type=str, default=\"del\", help=\"=del, ['del', 'endPoint'], the mode to work with the transcript which only expressing in one cellLine\")\n",
    "parser.add_argument(\"--cutoff_value\", dest=\"cutoff_value\", required=False, type=float, default=1.5, help=\"=1.5, the cutoff value of the significant log2FC for transcript\")\n",
    "parser.add_argument(\"--cell_line_1\", dest=\"cell_line_1\", required=True, type=str, default=\"\", help=\"the cell line 1\")\n",
    "parser.add_argument(\"--cell_line_2\", dest=\"cell_line_2\", required=True, type=str, default=\"\", help=\"the cell line 2\")\n",
    "parser.add_argument(\"--gene_name_list\", dest=\"gene_name_list\", required=False, nargs=\"*\", type=str, default=[], help=\"the gene will be showed\")\n",
    "parser.add_argument(\"--gene_id_list\", dest=\"gene_id_list\", required=False, nargs=\"*\", type=str, default=[], help=\"the gene will be showed\")\n",
    "parser.add_argument(\"--absolute_path\", dest=\"absolute_path\", required=False, action=\"store_true\", help=\"use the absolute path\")\n",
    "parser.add_argument(\"--file_path\", dest=\"file_path\", required=False, type=str, default=\"\", help=\"the path of data directory\")\n",
    "parser.add_argument(\"--input_df_filename \", dest=\"input_df_filename\", required=False, type=str, default=\"0004_relative_expression.tsv\", help=\"=0004_relative_expression.tsv, the output file of 0005.py\")\n",
    "parser.add_argument(\"--input_pkl_filename \", dest=\"input_pkl_filename\", required=False, type=str, default=\"0002_total_info.pkl\", help=\"=0002_total_info.pkl, the pickle file of 0002.py\")\n",
    "parser.add_argument(\"--output_plot_filename\", dest=\"output_plot_filename\", required=False, type=str, default=\"0006_transcript_expression.svg\", help=\"the output filename of this script\")\n",
    "\n",
    "args = parser.parse_args()\n",
    "infinity_mode = args.infinity_mode\n",
    "cutoff_value = args.cutoff_value\n",
    "cell_line_1 = args.cell_line_1\n",
    "cell_line_2 = args.cell_line_2\n",
    "gene_name_list = args.gene_name_list\n",
    "gene_id_list = args.gene_id_list\n",
    "absolute_path = args.absolute_path\n",
    "file_path = args.file_path\n",
    "input_df_filename = args.input_df_filename\n",
    "input_pkl_filename = args.input_pkl_filename\n",
    "output_plot_filename = args.output_plot_filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补全路径\n",
    "if absolute_path is False:\n",
    "    input_df_filename = \"{}{}\".format(file_path, input_df_filename)\n",
    "    input_pkl_filename = \"{}{}\".format(file_path, input_pkl_filename)\n",
    "    output_plot_filename = \"{}{}\".format(file_path, output_plot_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 打印参数\n",
    "print(\"\\n\")\n",
    "print(\"[Date]{}\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))\n",
    "print(\"[Script]{}\".format(__file__))\n",
    "print(\"[Version]{}\".format(__version__))\n",
    "print(\"[Version]def_function {}\".format(def_function.__version__))\n",
    "print(\"[Version]def_class: {}\".format(class_version))\n",
    "print(\"[Parament]infinity_mode: {}\".format(infinity_mode))\n",
    "print(\"[Parament]cutoff_value: {}\".format(cutoff_value))\n",
    "print(\"[Parament]cell_line_1: {}\".format(cell_line_1))\n",
    "print(\"[Parament]cell_line_2: {}\".format(cell_line_2))\n",
    "print(\"[Parament]gene_name_list: {}\".format(gene_name_list))\n",
    "print(\"[Parament]gene_id_list: {}\".format(gene_id_list))\n",
    "print(\"[Parament]input_df_filename: {}\".format(input_df_filename))\n",
    "print(\"[Parament]output_plot_filename: {}\".format(output_plot_filename))\n",
    "print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 绘图函数\n",
    "@def_function.log\n",
    "def make_plot(df, cell_line_1, cell_line_2, infinity_mode, cutoff_value, gene_name_list=None, gene_id_list=None, tab_level=0):\n",
    "    \"\"\"\n",
    "    input:\n",
    "        df, pandas.DataFrame, \n",
    "        cell_line_1, str, ...\n",
    "        cell_line_2, str, ...\n",
    "        infinity_mode, str, 对于仅在一种细胞系中表达的transcript的处理方式：\"del\" or \"endPoint\"\n",
    "        cutoff_value, float, log2FC的显著性阈值\n",
    "        gene_name_list, list, 要展示的gene的name, gene_name的优先级低于gene_id\n",
    "        gene_id_list, list, 要展示的gene的id, gene_id的优先级高于gene_name\n",
    "    change:\n",
    "        1.根据读取的数据, 整理出一个df, 一列保存每个转录本的log2FC, 一列保存转录本所映射的基因\n",
    "        2.存在部分transcript, 这些transcript仅在一种细胞系中表达, 其log2FC为nan\n",
    "            2.1当参数infinity_mode为endPoint时, 将这种transcript的log2FC定为x轴 最大值+1 or 最小值-1\n",
    "            2.2当参数infinity_mode为del时, 将不展示这种transcript\n",
    "        3.根据整理得到的temp_df, 绘制图片\n",
    "    output:\n",
    "        fig\n",
    "    \"\"\"\n",
    "    df = df\n",
    "    cell_line_1 = cell_line_1\n",
    "    cell_line_2 = cell_line_2\n",
    "    infinity_mode = infinity_mode\n",
    "    cutoff_value = cutoff_value\n",
    "    gene_name_list = (gene_name_list, [])[gene_name_list is None]\n",
    "    gene_id_list = (gene_id_list, [])[gene_id_list is None]\n",
    "\n",
    "\n",
    "    # 确定x轴的显示范围\n",
    "    temp = df[\"log2FC_{}@{}\".format(cell_line_1, cell_line_2)].map(lambda x: None if numpy.isinf(x) else x)\n",
    "    xaxis_min = math.floor(min(temp))\n",
    "    xaxis_max = math.ceil(max(temp))\n",
    "\n",
    "\n",
    "    # 根据cutoff_value对log2FC进行分类并对不同类型设定相应的颜色\n",
    "    temp_list = df[\"log2FC_{}@{}\".format(cell_line_1, cell_line_2)].to_list()\n",
    "    for i in range(0, len(temp_list)):\n",
    "        value = temp_list[i]\n",
    "        # 处理仅在一种cellLine中表达的transcript\n",
    "        if numpy.isinf(value):\n",
    "            if infinity_mode == \"del\":\n",
    "                temp_list[i] = None\n",
    "                continue\n",
    "            elif infinity_mode == \"endPoint\":\n",
    "                pass\n",
    "            else:\n",
    "                raise ValueError(\"[Error]Unsupported infinity_mode: {}\".format(infinity_mode))\n",
    "\n",
    "        # 根据log2FC的值判断应对transcript标注的颜色\n",
    "        if value >= cutoff_value:\n",
    "            temp_list[i] = \"indianred\"\n",
    "        elif value <= -cutoff_value:\n",
    "            temp_list[i] = \"lightblue\"\n",
    "        elif -cutoff_value < value < cutoff_value:\n",
    "            temp_list[i] = \"grey\"\n",
    "        else:\n",
    "            raise ValueError(\"[Error]the log2FC of {} is {}\".format(df[\"transcript_id\"][i], value))\n",
    "    df[\"color\"] = temp_list\n",
    "\n",
    "    if infinity_mode == \"del\":\n",
    "        df = df.dropna(axis=0, how=\"any\", subset=\"color\")\n",
    "    elif infinity_mode == \"endPoint\":\n",
    "        # 若log2FC值为inf, 则将其更改为最大值+1; 若log2FC值为-inf, 则将其更改为最小值-1\n",
    "        df[\"log2FC_{}@{}\".format(cell_line_1, cell_line_2)] = df[\"log2FC_{}@{}\".format(cell_line_1, cell_line_2)].map(lambda x: x if not numpy.isinf(x) else xaxis_max+1 if numpy.isposinf(x) else xaxis_min-1)\n",
    "\n",
    "\n",
    "    # 可以设置要展示的基因，基因标识可以是gene_id也可以是gene_name, gene_id的优先级更高，不可混合使用gene_id与gene_name\n",
    "    id_marker = \"Default\"\n",
    "    if len(gene_id_list) == 0:\n",
    "        if len(gene_name_list) == 0:\n",
    "            pass\n",
    "        else:\n",
    "            id_marker = \"gene_name\"\n",
    "    else:\n",
    "        id_marker = \"gene_id\"\n",
    "    # 根据id_marker筛选要展示的gene\n",
    "    if id_marker == \"gene_id\":\n",
    "        temp_df = df.loc[df[\"gene_id\"].isin(gene_id_list), :]\n",
    "    elif id_marker == \"gene_name\":\n",
    "        temp_df = df.loc[df[\"gene_name\"].isin(gene_name_list), :]\n",
    "    elif id_marker == \"Default\":\n",
    "        gene_id_list = df[\"gene_id\"].value_counts()\n",
    "        gene_id_list = [[k, v] for k, v in dict(gene_id_list).items()]\n",
    "        gene_id_list = sorted(gene_id_list, key=lambda x: x[1], reverse=True)\n",
    "        gene_id_list = gene_id_list[0:33]\n",
    "        gene_id_list = [i[0] for i in gene_id_list]\n",
    "        temp_df = df.loc[df[\"gene_id\"].isin(gene_id_list), :]\n",
    "    else:\n",
    "        raise ValueError(\"[Error]id_marker is {}\".format(id_marker))\n",
    "    # 调整排序，美化排版\n",
    "    temp_df = temp_df.sort_values(by=\"gene_name\", ascending=False)\n",
    "\n",
    "\n",
    "    # 绘制图形\n",
    "    fig = go.Figure()\n",
    "    fig_trace = go.Scatter(x=temp_df[\"log2FC_{}@{}\".format(cell_line_1, cell_line_2)],\n",
    "                        y=temp_df[\"gene_name\"],\n",
    "                        mode=\"markers\",\n",
    "                        marker={\"color\": temp_df[\"color\"]})\n",
    "    fig = fig.add_trace(fig_trace)\n",
    "    # 设置图片格式\n",
    "    layout = {\"width\": 250, \"height\": 500,\n",
    "            \"title\": {\"text\": \"5'-3' isoform expression\",\n",
    "                            \"font\": {\"family\": \"Arial\",\n",
    "                                    \"color\": \"black\",\n",
    "                                    \"size\": 12,},\n",
    "                            \"x\": 0.6},\n",
    "            \"margin\": {'l':15, 'r':15, 't':25, 'b':15},\n",
    "            \"paper_bgcolor\": \"white\",\n",
    "            \"plot_bgcolor\":\"white\",\n",
    "            \"xaxis\": {\"title\": {\"text\": \"log2FC({}/{})\".format(cell_line_1, cell_line_2),\n",
    "                                \"font\": {\"family\": \"Arial\",\n",
    "                                        \"color\": \"black\",\n",
    "                                        \"size\": 12,},\n",
    "                                \"standoff\": 0.5,\n",
    "                                },\n",
    "                        \"range\": [xaxis_min-1, xaxis_max+1],\n",
    "                        \"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"showticklabels\": True,\n",
    "                        \"ticklen\": 3,\n",
    "                        \"tickcolor\": \"black\",\n",
    "                        \"ticks\": \"outside\",\n",
    "                        },\n",
    "            \"yaxis\": {\"showline\": True,\n",
    "                        \"linecolor\": \"black\",\n",
    "                        \"showticklabels\": True,\n",
    "                        \"range\": [-1, len(set(temp_df[\"gene_name\"]))],\n",
    "                        \"tickfont\": {\"family\": \"Arial\",\n",
    "                                    \"color\": \"black\",\n",
    "                                    \"size\": 10},\n",
    "                        \"dtick\":1,\n",
    "                        \"ticks\": \"outside\",\n",
    "                        \"ticklen\": 3,\n",
    "                        \"tickcolor\": \"black\",\n",
    "                        }\n",
    "            }\n",
    "    fig = fig.update_layout(layout)\n",
    "\n",
    "    return fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "[Function]make_plot start.\n",
      "\t[Time]2023-08-18 10:32:58\n",
      "\t[Paraments]df: <...>\n",
      "\t[Paraments]cell_line_1: A673\n",
      "\t[Paraments]cell_line_2: Caco-2\n",
      "\t[Paraments]infinity_mode: del\n",
      "\t[Paraments]cutoff_value: <...>\n",
      "\t[Paraments]gene_name_list: []\n",
      "\t[Paraments]gene_id_list: []\n",
      "[2023-08-18 10:32:58]make_plot finished.\n",
      "[2023-08-18 10:32:59]All blocks finished.\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # load_data\n",
    "    df = pandas.read_csv(input_df_filename, sep='\\t')\n",
    "    with open(input_pkl_filename, 'rb') as file:\n",
    "        total = pickle.load(file)\n",
    "\n",
    "    # filter columns\n",
    "    new_columns = [\"gene_id\", \"transcript_id\"] + \\\n",
    "                  [\"log2FC_{}@{}\".format(cell_line_1, cell_line_2)] + \\\n",
    "                  [\"{}\".format(cell_line_1), \"{}\".format(cell_line_2)]\n",
    "    df = df[new_columns]\n",
    "\n",
    "    # filter gene_id\n",
    "    df = df.loc[map(lambda x: x[0:5]!=\"ENCLB\" , df[\"gene_id\"]), :]\n",
    "\n",
    "    # add gene_name\n",
    "    df[\"gene_name\"] = df[\"gene_id\"].map(lambda x: total.gene_dict[x].gene_name)\n",
    "\n",
    "    # make plot\n",
    "    fig = make_plot(df=df, cell_line_1=cell_line_1, cell_line_2=cell_line_2,\n",
    "                    infinity_mode=infinity_mode, cutoff_value=cutoff_value,\n",
    "                    gene_name_list=gene_name_list, gene_id_list=gene_id_list)\n",
    "    \n",
    "    # output plot\n",
    "    fig.write_image(output_plot_filename)\n",
    "\n",
    "    # log\n",
    "    print(\"[{}]All blocks finished.\".format(time.strftime(\"%Y-%m-%d %H:%M:%S\", time.localtime(time.time()))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "Start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "End\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "1.读取含有每个转录本在不同细胞系中差异表达的表格\n",
    "2.根据读取的数据，整理出一个df，一列保存每个转录本的log2FC，一列保存转录本所映射的基因\n",
    "    2.1存在部分转录本只在一种细胞系中表达，对于这种转录本，暂且将其log2FC定为最大值/最小值+1\n",
    "3.绘图\n",
    "\"\"\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
