{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "import gzip\n",
    "import time\n",
    "import argparse\n",
    "import pickle\n",
    "from Class import *\n",
    "import function\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "__version__ = \"V0.0(Editor) 2023-10-18\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 前置参数-debug\n",
    "COUNTSCUTOFF = 2\n",
    "ABSOLUTEPATH = False\n",
    "FILEPATH = \"F:/OneDrive/Master/Project/trans/data/\"\n",
    "RAWDATAPATH = \"raw_data/\"\n",
    "INPUTTOTAL = \"10001_total.pkl\"\n",
    "INPUTREFANNOTATION = \"Homo_sapiens.GRCh38.109.chr.gtf.gz\"\n",
    "OUTPUTTOTAL = \"10002_total.pkl\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补全路径\n",
    "if ABSOLUTEPATH == False:\n",
    "    RAWDATAPATH = \"{}{}\".format(FILEPATH, RAWDATAPATH)\n",
    "    INPUTTOTAL = \"{}{}\".format(FILEPATH, INPUTTOTAL)\n",
    "    INPUTREFANNOTATION = \"{}annotation/{}\".format(RAWDATAPATH, INPUTREFANNOTATION)\n",
    "    OUTPUTTOTAL = \"{}{}\".format(FILEPATH, OUTPUTTOTAL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 引用变量\n",
    "EXONRANGE = 5\n",
    "RANGETSS = 50\n",
    "RANGETES = 150"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 一级函数\n",
    "def findId(index, chr, location, rangeMax, orientation, allowType, total=None, transcriptGeneMap=None):\n",
    "    '''\n",
    "    input:\n",
    "        total, Total Object\n",
    "        index, dict, {<chr>: {<location>: {<location>: <id>, ...}, ...}, ...}\n",
    "        chr, str\n",
    "        location, int\n",
    "        rangeMax, int\n",
    "        orientation, str, 当为+时, location递增; 当为-时, location递减\n",
    "        allowType, str, \"gene\" or \"transcript\" or \"exon\"\n",
    "        transcriptGeneMap, dict, {<transcriptId>: <geneId>, ...}\n",
    "    change:\n",
    "        寻找index中, 相距location为rangeMax的id(不包含location)\n",
    "    '''\n",
    "    total = total\n",
    "    index = index\n",
    "    chr = chr\n",
    "    location = location\n",
    "    rangeMax = rangeMax\n",
    "    orientation = orientation\n",
    "    allowType = allowType\n",
    "    transcriptGeneMap = transcriptGeneMap\n",
    "\n",
    "    temp = []  # [(<id>, <mean expression in all cellLine>)]\n",
    "\n",
    "    if orientation == '-':\n",
    "        rangeList = [i for i in range(location-1, location-rangeMax-1, -1)]\n",
    "    else:\n",
    "        rangeList = [i for i in range(location+1, location+rangeMax+1, 1)]\n",
    "\n",
    "    for i in rangeList:\n",
    "        # 寻找是否在指定范围内存在相应的location\n",
    "        if index[chr].get(i, None) is None:\n",
    "            # 如果不存在, 就直接判断下一个位置\n",
    "            continue\n",
    "        else:\n",
    "            # 寻找到了范围内的location\n",
    "            newLocation = i\n",
    "            if allowType == \"exon\":\n",
    "                # index[chr][newLocation].values()中的值为str, 而不是list\n",
    "                for id in index[chr][newLocation].values():\n",
    "                    if total is not None:\n",
    "                        expression = numpy.mean([total.exonDict[id].cellLineExpression.get(cellLine, 0) for cellLine in total.celllineInfo.keys()])\n",
    "                    else:\n",
    "                        expression = None\n",
    "                    temp.append((id, expression))\n",
    "            elif allowType == \"gene\":\n",
    "                # index[chr][newLocation].values()中的值为str, 而不是list\n",
    "                for id in index[chr][newLocation].values():\n",
    "                    if total is not None:\n",
    "                        expression = numpy.mean([total.geneDict[id].cellLineExpression.get(cellLine, 0) for cellLine in total.celllineInfo.keys()])\n",
    "                    else:\n",
    "                        expression = None\n",
    "                    temp.append((id, expression))\n",
    "            else:\n",
    "                # index[chr][newLocation].values()中的值为list, 而不是str\n",
    "                if transcriptGeneMap is None:\n",
    "                    raise ValueError(\"findId(): need transcriptGeneMap\")\n",
    "                for idList in index[chr][newLocation].values():\n",
    "                    for id in idList:\n",
    "                        if total is not None:\n",
    "                            geneId = transcriptGeneMap.get(id, None)\n",
    "                            if geneId is None:\n",
    "                                expression = None\n",
    "                            else:\n",
    "                                expression = numpy.mean([total.geneDict[geneId].transcriptDict[id].cellLineExpression.get(cellLine, 0) for cellLine in total.celllineInfo.keys()])\n",
    "                        else:\n",
    "                            expression = None\n",
    "                        temp.append((id, expression))\n",
    "            temp = temp + findId(total=total, index=index, chr=chr, location=newLocation, rangeMax=rangeMax, orientation=orientation, allowType=allowType, transcriptGeneMap=transcriptGeneMap)\n",
    "\n",
    "    # 去重\n",
    "    temp = list(set(temp))\n",
    "\n",
    "    return temp\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 一级函数\n",
    "def loadRefAnnotation(filename, allowType):\n",
    "    '''\n",
    "    input:\n",
    "        filename, str, 参考基因组注释文件位置.gz\n",
    "        allowType, str, \"transcript\" or \"gene\" or \"exon\"\n",
    "    change:\n",
    "        读取参考基因组注释数据中allowType的数据\n",
    "        Warning: 在读取name时, 如果没有name则会返回\"None\"\n",
    "    return:\n",
    "        dict, {\"positionRef\": {\"TSS\": {<chr>: {<TSS>: {<TES>: [<transcriptId>, ...],\n",
    "                                                               ...},\n",
    "                                                       ...},\n",
    "                                               ...},\n",
    "                                       ...},\n",
    "                               \"TES\": {<chr>: {<TES>: {<TSS>: [<transcriptId>, ...],\n",
    "                                                               ...},\n",
    "                                                       ...},\n",
    "                                               ...}\n",
    "                                       ...},\n",
    "                \"dataRef\": {<id>: {<key>: <value>,\n",
    "                                   ...},\n",
    "                            ...}\n",
    "    '''\n",
    "    # 读取参考基因组注释数据\n",
    "    filename = filename\n",
    "    allowType = [allowType]\n",
    "\n",
    "    positionRef = {\"TSS\": {}, \"TES\": {}}  # {\"TSS\": {<chr>: {<TSS>: {<TES>: [<id>, ...], ...}, ...}, ...}, \"TES\": {<chr>: {<TES>: {<TSS>: [<id>, ...], ...}, ...}, ...}}\n",
    "    dataRef = {}\n",
    "\n",
    "    # 读取数据\n",
    "    with gzip.open(filename, 'rt') as file:\n",
    "        if allowType == [\"transcript\"]:\n",
    "            for line in file:\n",
    "                line = function.pickRefAnnotation(info=line, allowType=allowType)\n",
    "                if line is None:\n",
    "                    continue\n",
    "                # 提取数据\n",
    "                geneId = line[\"gene_id\"]\n",
    "                geneVersion = line[\"gene_version\"]\n",
    "                geneName = line.get(\"gene_name\", \"None\")\n",
    "                geneBiotype = line[\"gene_biotype\"]\n",
    "                transcriptId = line[\"transcript_id\"]\n",
    "                transcriptId = function.strSplit(transcriptId)[0]  # 去掉版本号\n",
    "                transcriptVersion = line[\"transcript_version\"]\n",
    "                transcriptName = line.get(\"transcript_name\", \"None\")\n",
    "                transcriptBiotype = line[\"transcript_biotype\"]\n",
    "                strand = line[\"strand\"]\n",
    "                chr = line[\"chr\"]\n",
    "                chr = (chr, 'M')[chr == \"MT\"]\n",
    "                chr = \"chr{}\".format(chr)\n",
    "                start = int(line[\"start\"])\n",
    "                end = int(line[\"end\"])\n",
    "                # 判断转录起点与转录终点\n",
    "                if strand == '+':\n",
    "                    TSS, TES = start, end\n",
    "                else:\n",
    "                    TSS, TES = end, start\n",
    "                # 在dataRef中添加信息\n",
    "                if transcriptId in dataRef.keys():\n",
    "                    print(\"[Warning]loadRefAnnotation(): transcriptId {} have been existed\".format(transcriptId))\n",
    "                dataRef[transcriptId] = {\"geneId\": geneId,\n",
    "                                         \"geneVersion\": geneVersion,\n",
    "                                         \"geneName\": geneName,\n",
    "                                         \"geneBiotype\": geneBiotype,\n",
    "                                         \"transcriptId\": transcriptId,\n",
    "                                         \"transcriptVersion\": transcriptVersion,\n",
    "                                         \"transcriptName\": transcriptName,\n",
    "                                         \"transcriptBiotype\": transcriptBiotype,\n",
    "                                         \"strand\": strand,\n",
    "                                         \"chr\": chr,\n",
    "                                         \"start\": start,\n",
    "                                         \"end\": end,\n",
    "                                         \"TSS\": TSS,\n",
    "                                         \"TES\": TES}\n",
    "        elif allowType == [\"gene\"]:\n",
    "            for line in file:\n",
    "                line = function.pickRefAnnotation(info=line, allowType=allowType)\n",
    "                if line is None:\n",
    "                    continue\n",
    "                # 提取数据\n",
    "                geneId = function.strSplit(line[\"gene_id\"])[0]\n",
    "                geneVersion = line[\"gene_version\"]\n",
    "                geneName = line.get(\"gene_name\", \"None\")\n",
    "                geneBiotype = line[\"gene_biotype\"]\n",
    "                strand = line[\"strand\"]\n",
    "                chr = line[\"chr\"]\n",
    "                chr = (chr, 'M')[chr == \"MT\"]\n",
    "                chr = \"chr{}\".format(chr)\n",
    "                start = int(line[\"start\"])\n",
    "                end = int(line[\"end\"])\n",
    "                # 判断转录起点与转录终点\n",
    "                if strand == '+':\n",
    "                    TSS, TES = start, end\n",
    "                else:\n",
    "                    TSS, TES = end, start\n",
    "                # 在dataRef中添加信息\n",
    "                if geneId in dataRef.keys():\n",
    "                    print(\"[Warning]loadRefAnnotation(): geneId {} have been existed\".format(geneId))\n",
    "                dataRef[geneId] = {\"geneId\": geneId,\n",
    "                                    \"geneName\": geneName,\n",
    "                                    \"geneVersion\": geneVersion,\n",
    "                                    \"geneBiotype\": geneBiotype,\n",
    "                                    \"strand\": strand,\n",
    "                                    \"chr\": chr,\n",
    "                                    \"start\": start,\n",
    "                                    \"end\": end,\n",
    "                                    \"TSS\": TSS,\n",
    "                                    \"TES\": TES}\n",
    "        elif allowType == [\"exon\"]:\n",
    "            for line in file:\n",
    "                line = function.pickRefAnnotation(info=line, allowType=allowType)\n",
    "                if line is None:\n",
    "                    continue\n",
    "                # 提取数据\n",
    "                geneId = line[\"gene_id\"]\n",
    "                transcriptId = line[\"transcript_id\"]\n",
    "                [exonId, exonVersion] = function.strSplit(line[\"exon_id\"])\n",
    "                exonVersion = line[\"exon_version\"]\n",
    "                strand = line[\"strand\"]\n",
    "                chr = line[\"chr\"]\n",
    "                chr = (\"chr{}\".format(chr), \"chrM\")[chr == \"MT\"]\n",
    "                start = int(line[\"start\"])\n",
    "                end = int(line[\"end\"])\n",
    "                # 判断转录起点与转录终点\n",
    "                if strand == '+':\n",
    "                    TSS, TES = start, end\n",
    "                else:\n",
    "                    TSS, TES = end, start\n",
    "                # 在dataRef中添加信息\n",
    "                if exonId in dataRef.keys():\n",
    "                    dataRef[exonId][\"geneTranscriptIdSet\"].add((geneId, transcriptId))\n",
    "                else:\n",
    "                    dataRef[exonId] = {\"geneId\": geneId,\n",
    "                                        \"transcriptId\": transcriptId,\n",
    "                                        \"exonId\": exonId,\n",
    "                                        \"exonVersion\": exonVersion,\n",
    "                                        \"strand\": strand,\n",
    "                                        \"chr\": chr,\n",
    "                                        \"start\": start,\n",
    "                                        \"end\": end,\n",
    "                                        \"TSS\": TSS,\n",
    "                                        \"TES\": TES,\n",
    "                                        \"geneTranscriptIdSet\": {(geneId, transcriptId)}}\n",
    "        else:\n",
    "            raise ValueError(\"the allowType should be 'transcript' or 'gene', but it is {}\".format(allowType))\n",
    "\n",
    "    # 整理index\n",
    "    # 在positionRef中添加信息\n",
    "    if allowType == [\"transcript\"]:\n",
    "        for id, transcriptDict in dataRef.items():\n",
    "            chr = transcriptDict[\"chr\"]\n",
    "            TSS = transcriptDict[\"TSS\"]\n",
    "            TES = transcriptDict[\"TES\"]\n",
    "            if chr not in positionRef[\"TSS\"].keys():\n",
    "                positionRef[\"TSS\"][chr] = {TSS: {TES: [id]}}\n",
    "            else:\n",
    "                if TSS not in positionRef[\"TSS\"][chr].keys():\n",
    "                    positionRef[\"TSS\"][chr][TSS] = {TES: [id]}\n",
    "                else:\n",
    "                    if TES not in positionRef[\"TSS\"][chr][TSS].keys():\n",
    "                        positionRef[\"TSS\"][chr][TSS][TES] = [id]\n",
    "                    else:\n",
    "                        positionRef[\"TSS\"][chr][TSS][TES].append(id)\n",
    "            if chr not in positionRef[\"TES\"].keys():\n",
    "                positionRef[\"TES\"][chr] = {TES: {TSS: [id]}}\n",
    "            else:\n",
    "                if TES not in positionRef[\"TES\"][chr].keys():\n",
    "                    positionRef[\"TES\"][chr][TES] = {TSS: [id]}\n",
    "                else:\n",
    "                    if TSS not in positionRef[\"TES\"][chr][TES].keys():\n",
    "                        positionRef[\"TES\"][chr][TES][TSS] = [id]\n",
    "                    else:\n",
    "                        positionRef[\"TES\"][chr][TES][TSS].append(id)\n",
    "    elif allowType == [\"gene\"] or allowType == [\"exon\"]:\n",
    "        for id, dataDict in dataRef.items():\n",
    "            chr = dataDict[\"chr\"]\n",
    "            TSS = dataDict[\"TSS\"]\n",
    "            TES = dataDict[\"TES\"]\n",
    "            if chr not in positionRef[\"TSS\"].keys():\n",
    "                positionRef[\"TSS\"][chr] = {TSS: {TES: id}}\n",
    "            else:\n",
    "                if TSS not in positionRef[\"TSS\"][chr].keys():\n",
    "                    positionRef[\"TSS\"][chr][TSS] = {TES: id}\n",
    "                else:\n",
    "                    if TES not in positionRef[\"TSS\"][chr][TSS].keys():\n",
    "                        positionRef[\"TSS\"][chr][TSS][TES] = id\n",
    "                    else:\n",
    "                        pass\n",
    "                        #print(\"[Warning]{id}-->{e}, but {e} is existed\".format(id=id, e=positionRef[\"TSS\"][chr][TSS][TES]))\n",
    "            if chr not in positionRef[\"TES\"].keys():\n",
    "                positionRef[\"TES\"][chr] = {TES: {TSS: id}}\n",
    "            else:\n",
    "                if TES not in positionRef[\"TES\"][chr].keys():\n",
    "                    positionRef[\"TES\"][chr][TES] = {TSS: id}\n",
    "                else:\n",
    "                    if TSS not in positionRef[\"TES\"][chr][TES].keys():\n",
    "                        positionRef[\"TES\"][chr][TES][TSS] = id\n",
    "                    else:\n",
    "                        pass\n",
    "                        #print(\"[Warning]{id}-->{e}, but {e} is existed\".format(id=id, e=positionRef[\"TSS\"][chr][TSS][TES]))\n",
    "    else:\n",
    "        raise ValueError(\"the allowType should be 'transcript' or 'gene', but it is {}\".format(allowType))\n",
    "    \n",
    "    return {\"positionRef\": positionRef, \"dataRef\": dataRef}\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "with open(INPUTTOTAL, \"rb\") as file:\n",
    "    total = pickle.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 更新total中的exonDict及exonIndex\n",
    "total.refresh()\n",
    "total.reIndex()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 声明变量\n",
    "exonCorrected = {\"mapped\": {}, \"TSS-TES\": {}, \"TSS\": {}, \"TES\": {}, \"integrate TSS\": {}, \"integrate TES\": {}}  # {\"\": {<existed exon id>: <corrected exon id>}, ...}\n",
    "readyCorList = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# 根据参考基因组中exon数据对exon的TSS及TES进行校正\n",
    "# 读取参考基因组注释中exon的数据, 并建立index\n",
    "temp = loadRefAnnotation(filename=INPUTREFANNOTATION, allowType=\"exon\")\n",
    "exonRef = temp[\"dataRef\"]\n",
    "exonIndex = temp[\"positionRef\"]\n",
    "# 遍历total.exonDict中所有exon, 记录这些exon\n",
    "for exonId, exonObject in total.exonDict.items():\n",
    "    readyCorList.append(exonId)\n",
    "# 统计校正前exon及transcript的数量\n",
    "infoDict = {}\n",
    "infoDict[\"total exon num before exon correct\"] = len(total.exonDict)\n",
    "num = 0\n",
    "for exonId, exonObject in total.exonDict.items():\n",
    "    if exonObject.status == \"NOVEL\":\n",
    "        num += 1\n",
    "infoDict[\"novel exon num before exon correct\"] = num\n",
    "num = 0\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptId in geneObject.transcriptDict.keys():\n",
    "        num += 1\n",
    "infoDict[\"total transcript num before exon correct\"] = num\n",
    "num = 0\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptObject in geneObject.transcriptDict.values():\n",
    "        if transcriptObject.status == \"NOVEL\":\n",
    "            num +=1\n",
    "infoDict[\"novel transcript num before exon correct\"] = num\n",
    "\n",
    "# 根据参考基因组中exon数据对exon的TSS及TES进行校正\n",
    "for exonId in tqdm.tqdm(readyCorList, desc=\"Exon\", leave=True):\n",
    "    exonObject = total.exonDict[exonId]\n",
    "    chr = exonObject.chr\n",
    "    TSS = exonObject.TSS\n",
    "    TES = exonObject.TES\n",
    "\n",
    "    # 声明list类型变量exonList\n",
    "    exonList = []  # 存储可进行映射的exon [(<exonId>, <dif>, <startMarker>, <endMarker>), ...]\n",
    "\n",
    "    # 准备校正exon的TSS及TES\n",
    "    # 获取[TSS-5, TES+5]范围的所有exon\n",
    "    temp = [exonIndex[\"TSS\"][chr].get(i, None) for i in range(TSS-EXONRANGE, TSS+EXONRANGE+1)]\n",
    "    # 去除空值\n",
    "    while None in temp:\n",
    "        temp.remove(None)\n",
    "    while {} in temp:\n",
    "        temp.remove({})\n",
    "        pass\n",
    "    for d in temp:\n",
    "        for tempTES, tempExonId in d.items():\n",
    "            if abs(tempTES-TES) <= EXONRANGE:\n",
    "                # 从中筛选出[TES-5, TES+5]范围的所有exon, 认为这些exon为可双端映射的exon\n",
    "                num = abs(TSS-exonRef[tempExonId][\"TSS\"]) + abs(TES-tempTES)\n",
    "                exonList.append((tempExonId, num, True, True))\n",
    "            else:\n",
    "                # 仅TSS可映射到tempExon\n",
    "                num = abs(TSS-exonRef[tempExonId][\"TSS\"])\n",
    "                exonList.append((tempExonId, num, True, False))\n",
    "\n",
    "    # 获取[TES-5, TES+5]范围的所有exon\n",
    "    temp = [exonIndex[\"TES\"][chr].get(i, None) for i in range(TES-EXONRANGE, TES+EXONRANGE+1)]\n",
    "    # 去除空值\n",
    "    while None in temp:\n",
    "        temp.remove(None)\n",
    "    while {} in temp:\n",
    "        temp.remove({})\n",
    "    for d in temp:\n",
    "        for tempTSS, tempExonId in d.items():\n",
    "            # 从中筛选出(-oo, TSS-5)U(TSS+5, +oo)的所有exon, 认为这些exon为仅可映射到TES的exon\n",
    "            if abs(tempTSS-TSS) > EXONRANGE:\n",
    "                # 仅end可映射到tempExon\n",
    "                num = abs(TES-exonRef[tempExonId][\"TES\"])\n",
    "                exonList.append((tempExonId, num, False, True))\n",
    "\n",
    "    # 不分析无法映射的exon\n",
    "    if len(exonList) == 0:\n",
    "        continue\n",
    "\n",
    "    # 校正过程\n",
    "    mode = False  # 是否存在最佳映射\n",
    "    for (newExonId, num, startMarker, endMarker) in exonList:\n",
    "        if startMarker == endMarker == True:\n",
    "            mode = True\n",
    "            break\n",
    "    # 开始映射\n",
    "    if mode is True:\n",
    "        # 若exon可双端映射到一个exon\n",
    "        # 获取相差碱基数最少的exon\n",
    "        temp = []\n",
    "        for (newExonId, num, TSSMarker, TESMarker) in exonList:\n",
    "            if TSSMarker == TESMarker == True:\n",
    "                temp.append((newExonId, num))\n",
    "        temp = sorted(temp, key=lambda x: x[1])\n",
    "        (newExonId, num) = temp[0]\n",
    "        '''if num == 0:\n",
    "            # 表明不必进行start, end的修改\n",
    "            continue'''\n",
    "        # 判断exonId是否相同\n",
    "        if exonId == newExonId:\n",
    "            # exonId相同\n",
    "            # 修改status, exonVersion, start, TSS, end, TES\n",
    "            total.exonDict[exonId].status = \"KNOWN\"\n",
    "            total.exonDict[exonId].exonVersion = exonRef[newExonId][\"exonVersion\"]\n",
    "            total.exonDict[exonId].start = exonRef[newExonId][\"start\"]\n",
    "            total.exonDict[exonId].end = exonRef[newExonId][\"end\"]\n",
    "            total.exonDict[exonId].TSS = exonRef[newExonId][\"TSS\"]\n",
    "            total.exonDict[exonId].TES = exonRef[newExonId][\"TES\"]\n",
    "        else:\n",
    "            # exonId不同\n",
    "            # 查询参考exonId是否已记录\n",
    "            if newExonId in total.exonDict.keys():\n",
    "                # 参考exonId已记录, 合并两exon\n",
    "                total._exonMerge(exonId1=newExonId, exonId2=exonId)\n",
    "            else:\n",
    "                # 参考exonId未记录\n",
    "                # exonDict中新增参考exon对象\n",
    "                total.exonDict[newExonId] = Exon(status=\"KNOWN\",\n",
    "                                                 exonId=exonRef[newExonId][\"exonId\"],\n",
    "                                                 exonVersion=exonRef[newExonId][\"exonVersion\"],\n",
    "                                                 chr=exonRef[newExonId][\"chr\"],\n",
    "                                                 strand=exonRef[newExonId][\"strand\"],\n",
    "                                                 start=exonRef[newExonId][\"start\"],\n",
    "                                                 end=exonRef[newExonId][\"end\"],\n",
    "                                                 TSS=exonRef[newExonId][\"TSS\"],\n",
    "                                                 TES=exonRef[newExonId][\"TES\"])\n",
    "                # 合并原exon与参考exon\n",
    "                total._exonMerge(exonId1=newExonId, exonId2=exonId)\n",
    "            # 将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"mapped\"]中\n",
    "            exonCorrected[\"mapped\"][exonId] = newExonId\n",
    "    else:\n",
    "        # 若exon不可双端映射到一个exon\n",
    "        # 如果status为KNOWN就不再进行映射, 跳过该exon\n",
    "        if total.exonDict[exonId].status == \"KNOWN\":\n",
    "            continue\n",
    "        # 修改exon的status为CORRECTED\n",
    "        total.exonDict[exonId].status = \"CORRECTED\"\n",
    "        # 判断exon能否双端映射到两个exon\n",
    "        marker=False  # 标记该exon能否映射到两个exon, 1-可根据start映射 2-可根据end映射 3-可分别根据start,end映射\n",
    "        for (newExonId, num, startMarker, endMarker) in exonList:\n",
    "            if startMarker is True:\n",
    "                marker = 3 if marker==2 else 1\n",
    "            else:\n",
    "                marker = 3 if marker==1 else 2\n",
    "            if marker == 3:\n",
    "                break\n",
    "        if marker == 3:\n",
    "            # 能映射到两个exon\n",
    "            # 获取TSS端相差碱基数最少的exon\n",
    "            temp = [(newExonId, num, startMarker, endMarker) for (newExonId, num, startMarker, endMarker) in exonList if startMarker is True]\n",
    "            temp = sorted(temp, key=lambda x: x[1])\n",
    "            (newExonId, num, startMarker, endMarker) = temp[0]\n",
    "            # 在exon的status中添加-TSS-TES后缀\n",
    "            total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TSS-TES\"\n",
    "            if num != 0:\n",
    "                # 表明需要进行TSS的修改\n",
    "                # 修改exon的TSS的位置\n",
    "                if total.exonDict[exonId].strand == '+':\n",
    "                    total.exonDict[exonId].start = exonRef[newExonId][\"start\"]\n",
    "                    total.exonDict[exonId].TSS = exonRef[newExonId][\"TSS\"]\n",
    "                else:\n",
    "                    total.exonDict[exonId].end = exonRef[newExonId][\"end\"]\n",
    "                    total.exonDict[exonId].TSS = exonRef[newExonId][\"TSS\"]\n",
    "                # 将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TSS-TES\"]中\n",
    "                exonCorrected[\"TSS-TES\"][exonId] = {\"TSS\": newExonId, \"TES\": None}\n",
    "            # 获取TES端相差碱基数最少的exon\n",
    "            temp = [(newExonId, num, startMarker, endMarker) for (newExonId, num, startMarker, endMarker) in exonList if endMarker is True]\n",
    "            temp = sorted(temp, key=lambda x: x[1])\n",
    "            (newExonId, num, startMarker, endMarker) = temp[0]\n",
    "            if num != 0:\n",
    "                # 表明需要进行TES的修改\n",
    "                # 修改exon的TES的位置\n",
    "                if total.exonDict[exonId].strand == '+':\n",
    "                    total.exonDict[exonId].end = exonRef[newExonId][\"end\"]\n",
    "                    total.exonDict[exonId].TES = exonRef[newExonId][\"TES\"]\n",
    "                else:\n",
    "                    total.exonDict[exonId].start = exonRef[newExonId][\"start\"]\n",
    "                    total.exonDict[exonId].TES = exonRef[newExonId][\"TES\"]\n",
    "                # 将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TSS-TES\"]中\n",
    "                if exonId not in exonCorrected[\"TSS-TES\"].keys():\n",
    "                    exonCorrected[\"TSS-TES\"][exonId] = {\"TSS\": None, \"TES\": newExonId}\n",
    "                else:\n",
    "                    exonCorrected[\"TSS-TES\"][exonId][\"TES\"] = newExonId\n",
    "        else:\n",
    "            # 仅能单端映射到一个exon\n",
    "            # 筛选出相差碱基数最少的exon进行映射\n",
    "            exonList = sorted(exonList, key=lambda x: x[1])\n",
    "            tempNum = exonList[0][1]\n",
    "            exonList = [(newExonId, num, startMarker, endMarker) for (newExonId, num, startMarker, endMarker) in exonList if num == tempNum]\n",
    "            (newExonId, num, startMarker, endMarker) = exonList[0]\n",
    "            # 判断哪一端可进行映射\n",
    "            if startMarker is True:\n",
    "                # TSS端可映射\n",
    "                # 在exon的status中添加-TSS后缀\n",
    "                total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TSS\"\n",
    "                # 将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TSS\"]中\n",
    "                exonCorrected[\"TSS\"][exonId] = newExonId\n",
    "                if num == 0:\n",
    "                    # 表明不必进行TSS的修改\n",
    "                    continue\n",
    "                # 修改exon的TSS的位置\n",
    "                if total.exonDict[exonId].strand == '+':\n",
    "                    total.exonDict[exonId].start = exonRef[newExonId][\"start\"]\n",
    "                    total.exonDict[exonId].TSS = exonRef[newExonId][\"TSS\"]\n",
    "                else:\n",
    "                    total.exonDict[exonId].end = exonRef[newExonId][\"end\"]\n",
    "                    total.exonDict[exonId].TSS = exonRef[newExonId][\"TSS\"]\n",
    "            else:\n",
    "                # TES端可映射\n",
    "                # 在exon的status中添加-TES后缀\n",
    "                total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TES\"\n",
    "                # 将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TES\"]中\n",
    "                exonCorrected[\"TES\"][exonId] = newExonId\n",
    "                if num == 0:\n",
    "                    # 表明不必进行TES的修改\n",
    "                    continue\n",
    "                # 修改exon的TES的位置\n",
    "                if total.exonDict[exonId].strand == '+':\n",
    "                    total.exonDict[exonId].end = exonRef[newExonId][\"end\"]\n",
    "                    total.exonDict[exonId].TES = exonRef[newExonId][\"TES\"]\n",
    "                else:\n",
    "                    total.exonDict[exonId].start = exonRef[newExonId][\"start\"]\n",
    "                    total.exonDict[exonId].TES = exonRef[newExonId][\"TES\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 更新total中的exonDict及exonIndex\n",
    "total.refresh()\n",
    "total.reIndex()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 统计校正后exon及transcript的数量\n",
    "infoDict[\"exon mapped\"] = len(exonCorrected[\"mapped\"])\n",
    "infoDict[\"exon TSS-TES\"] = len(exonCorrected[\"TSS-TES\"])\n",
    "infoDict[\"exon TSS\"] = len(exonCorrected[\"TSS\"])\n",
    "infoDict[\"exon TES\"] = len(exonCorrected[\"TES\"])\n",
    "infoDict[\"total exon num after exon correct\"] = len(total.exonDict)\n",
    "num=0\n",
    "for exonId, exonObject in total.exonDict.items():\n",
    "    if exonObject.status == \"NOVEL\":\n",
    "        num+=1\n",
    "infoDict[\"novel exon num after exon correct\"] = num\n",
    "num = 0\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptId in geneObject.transcriptDict.keys():\n",
    "        num += 1\n",
    "infoDict[\"total transcript num after exon correct\"] = num\n",
    "num = 0\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptObject in geneObject.transcriptDict.values():\n",
    "        if transcriptObject.status == \"NOVEL\":\n",
    "            num +=1\n",
    "infoDict[\"novel transcript num after exon correct\"] = num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 计算exon, transcript, gene的相对表达量\n",
    "total.computeRelativeExpression()\n",
    "total.computeCellLineExpression()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 准备记录合并记录\n",
    "infoDict[\"integrate TSS\"] = 0\n",
    "infoDict[\"integrate TES\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 合并相邻的exon, 即合并相差EXONRANGE范围内的NOVEL及未校正端的exon\n",
    "if True:\n",
    "    # 建立index\n",
    "    indexTSS = total.exonIndexBuild(siteType=\"TSS\")  # {<chr>: {<TSS>: {<TES>: <exonId>, ...}, ...}, ...}\n",
    "    indexTES = total.exonIndexBuild(siteType=\"TES\")  # {<chr>: {<TES>: {<TSS>: <exonId>, ...}, ...}, ...}\n",
    "\n",
    "    # 处理TSS\n",
    "    # 声明set变量exonSet\n",
    "    exonSet = set()  # {<exonId>, ...}\n",
    "    # 筛选符合校正条件的exon\n",
    "    for exonId, exonObject in total.exonDict.items():\n",
    "        if exonObject.status == \"KNOWN\":\n",
    "            continue\n",
    "        elif \"-TSS\" in exonObject.status:\n",
    "            continue\n",
    "        else:\n",
    "            exonSet.add(exonId)\n",
    "    # 只要exonSet不为空集, 就随机提取exonSet中的一个exonId, 寻找其TSS上下游各EXONRANGE范围内的exon\n",
    "    while len(exonSet) != 0:\n",
    "        exonId = exonSet.pop()\n",
    "        exonObject = total.exonDict[exonId]\n",
    "        chr = exonObject.chr\n",
    "        TSS = exonObject.TSS\n",
    "        # 声明list变量temp, [(<exonId>, <expression>), ...]\n",
    "        temp = []  # [(<exonId>, <expression>), ...]\n",
    "        # 寻找所有与该exon的start相同的exon\n",
    "        for exonId in indexTSS[chr][TSS].values():\n",
    "            temp = temp + [(exonId, numpy.mean([total.exonDict[exonId].cellLineExpression.get(cellLine, 0) for cellLine in total.celllineInfo.keys()]))]\n",
    "        # 递减寻找EXONRANGE范围内的exon\n",
    "        temp = temp + findId(total=total, index=indexTSS, chr=chr, location=TSS, rangeMax=EXONRANGE, orientation='-', allowType=\"exon\")\n",
    "        # 递增寻找EXONRANGE范围内的exon\n",
    "        temp = temp + findId(total=total, index=indexTSS, chr=chr, location=TSS, rangeMax=EXONRANGE, orientation='+', allowType=\"exon\")\n",
    "        # 去重\n",
    "        temp = list(set(temp))\n",
    "        # 判断是否寻找到了其他的exon\n",
    "        if len(temp)==1:\n",
    "            # 未寻找到其他的exon, 认为该exon的start无法修改, 跳过该exon\n",
    "            exonSet.discard(temp[0][0])\n",
    "            continue\n",
    "        else:\n",
    "            # 已寻找到其他的exon\n",
    "            pass\n",
    "        # 判断在temp中是否存在已校正端\n",
    "        corSet = set()  # {(exonId, TSS), ...}\n",
    "        nonCorSet = set()  # {(exonId, TSS), ...}\n",
    "        # 分离其中的未校正端与已校正端\n",
    "        for (exonId, expression) in temp:\n",
    "            status = total.exonDict[exonId].status\n",
    "            TSS = total.exonDict[exonId].TSS\n",
    "            if status==\"KNOWN\" or \"-TSS\" in status:\n",
    "                # TSS端可信\n",
    "                corSet.add((exonId, TSS))\n",
    "            else:\n",
    "                # TSS端不可信\n",
    "                nonCorSet.add((exonId, TSS))\n",
    "        if len(corSet) != 0:\n",
    "            # 存在已校正端\n",
    "            # 对每一个未校正端选择其距离最近的已校正端作为参考位点进行校正\n",
    "            for (exonId, TSS) in nonCorSet:\n",
    "                # 寻找最近的已校正端\n",
    "                diffTSS = [(exonIdTemp, tempTSS, abs(TSS-tempTSS)) for (exonIdTemp, tempTSS) in corSet]\n",
    "                diffTSS = sorted(diffTSS, key=lambda x: x[2])\n",
    "                targetExonId = diffTSS[0][0]\n",
    "                targetTSS = diffTSS[0][1]\n",
    "                # 修改TSS, start/end位置\n",
    "                total.exonDict[exonId].TSS = targetTSS\n",
    "                if total.exonDict[exonId].strand == '+':\n",
    "                    total.exonDict[exonId].start = targetTSS\n",
    "                else:\n",
    "                    total.exonDict[exonId].end = targetTSS\n",
    "                # 改变status\n",
    "                total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TSS\"\n",
    "                # 删除exonSet中该exonId\n",
    "                exonSet.discard(exonId)\n",
    "                # 记录这一次校正\n",
    "                exonCorrected[\"integrate TSS\"][exonId] = targetExonId\n",
    "            # 结束此次循环\n",
    "            continue\n",
    "        else:\n",
    "            # 不存在已校正端, 进行下一步\n",
    "            pass\n",
    "        # 以expression最大的exon为参照, 修改其他exon的TSS\n",
    "        exonIdTarget = None\n",
    "        expressionMax = 0\n",
    "        # 寻找expression最大的exon\n",
    "        for (exonId, expression) in temp:\n",
    "            if expression >= expressionMax:\n",
    "                expressionMax = expression\n",
    "                exonIdTarget = exonId\n",
    "            else:\n",
    "                continue\n",
    "        # 修改其他的exon\n",
    "        for (exonId, expression) in temp:\n",
    "            # 修改其他exon的TSS, start/end\n",
    "            total.exonDict[exonId].TSS = total.exonDict[exonIdTarget].TSS\n",
    "            if total.exonDict[exonId].strand == '+':\n",
    "                total.exonDict[exonId].start = total.exonDict[exonIdTarget].start\n",
    "            else:\n",
    "                total.exonDict[exonId].end = total.exonDict[exonIdTarget].end\n",
    "            # 修改status\n",
    "            # 不应该修改status, 因为无已校正TSS作为参考, 所以TSS即使进行了合并也并不是可靠的\n",
    "            #total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TSS\"\n",
    "            # 删除exonSet中的这些exon\n",
    "            exonSet.discard(exonId)\n",
    "            # 记录这一次校正\n",
    "            exonCorrected[\"integrate TSS\"][exonId] = exonIdTarget\n",
    "\n",
    "    # 处理TES\n",
    "    # 声明set变量exonSet\n",
    "    exonSet = set()  # {<exonId>, ...}\n",
    "    # 筛选符合校正条件的exon\n",
    "    for exonId, exonObject in total.exonDict.items():\n",
    "        if exonObject.status == \"KNOWN\":\n",
    "            continue\n",
    "        elif \"-TES\" in exonObject.status:\n",
    "            continue\n",
    "        else:\n",
    "            exonSet.add(exonId)\n",
    "    # 只要exonSet不为空集, 就随机提取exonSet中的一个exonId, 寻找其TES上下游各EXONRANGE范围内的exon\n",
    "    while len(exonSet) != 0:\n",
    "        exonId = exonSet.pop()\n",
    "        exonObject = total.exonDict[exonId]\n",
    "        chr = exonObject.chr\n",
    "        TES = exonObject.TES\n",
    "        # 声明list变量temp, [(<exonId>, <expression>), ...]\n",
    "        temp = []  # [(<exonId>, <expression>), ...]\n",
    "        # 寻找所有与该exon的TES相同的exon\n",
    "        for exonId in indexTES[chr][TES].values():\n",
    "            temp = temp + [(exonId, numpy.mean([total.exonDict[exonId].cellLineExpression.get(cellLine, 0) for cellLine in total.celllineInfo.keys()]))]\n",
    "        # 递减寻找EXONRANGE范围内的exon\n",
    "        temp = temp + findId(total=total, index=indexTES, chr=chr, location=TES, rangeMax=EXONRANGE, orientation='-', allowType=\"exon\")\n",
    "        # 递增寻找EXONRANGE范围内的exon\n",
    "        temp = temp + findId(total=total, index=indexTES, chr=chr, location=TES, rangeMax=EXONRANGE, orientation='+', allowType=\"exon\")\n",
    "        # 去重\n",
    "        temp = list(set(temp))\n",
    "        # 判断是否寻找到了其他的exon\n",
    "        if len(temp)==1:\n",
    "            # 未寻找到其他的exon, 认为该exon的end无法修改, 跳过该exon\n",
    "            exonSet.discard(temp[0][0])\n",
    "            continue\n",
    "        else:\n",
    "            # 已寻找到其他的exon, 继续\n",
    "            pass\n",
    "        # 判断在temp中是否存在已校正端\n",
    "        corSet = set()  # {(exonId, TES), ...}\n",
    "        nonCorSet = set()  # {(exonId, TES), ...}\n",
    "        # 分离其中的未校正端与已校正端\n",
    "        for (exonId, expression) in temp:\n",
    "            status = total.exonDict[exonId].status\n",
    "            TES = total.exonDict[exonId].TES\n",
    "            if status==\"KNOWN\" or \"-TES\" in status:\n",
    "                # TES端可信\n",
    "                corSet.add((exonId, TES))\n",
    "            else:\n",
    "                # TES端不可信\n",
    "                nonCorSet.add((exonId, TES))\n",
    "        if len(corSet) != 0:\n",
    "            # 存在已校正端\n",
    "            # 对每一个未校正端选择其距离最近的已校正端作为参考位点进行校正\n",
    "            for (exonId, TES) in nonCorSet:\n",
    "                # 寻找最近的已校正端\n",
    "                diffTES = [(exonIdTemp, tempTES, abs(TES-tempTES)) for (exonIdTemp, tempTES) in corSet]\n",
    "                diffTES = sorted(diffTES, key=lambda x: x[2])\n",
    "                targetExonId = diffTES[0][0]\n",
    "                targetTES = diffTES[0][1]\n",
    "                # 修改TES, start/end位置\n",
    "                total.exonDict[exonId].TES = targetTES\n",
    "                if total.exonDict[exonId].strand == '+':\n",
    "                    total.exonDict[exonId].end = targetTES\n",
    "                else:\n",
    "                    total.exonDict[exonId].start =targetTES\n",
    "                # 改变status\n",
    "                total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TES\"\n",
    "                # 删除exonSet中该exonId\n",
    "                exonSet.discard(exonId)\n",
    "                # 记录这一次校正\n",
    "                exonCorrected[\"integrate TES\"][exonId] = targetExonId\n",
    "            # 结束此次循环\n",
    "            continue\n",
    "        else:\n",
    "            # 不存在已校正端, 进行下一步\n",
    "            pass\n",
    "        # 以expression最大的exon为参照, 修改其他exon的TES\n",
    "        exonIdTarget = None\n",
    "        expressionMax = 0\n",
    "        # 寻找expression最大的exon\n",
    "        for (exonId, expression) in temp:\n",
    "            if expression >= expressionMax:\n",
    "                expressionMax = expression\n",
    "                exonIdTarget = exonId\n",
    "        # 修改其他的exon\n",
    "        for (exonId, expression) in temp:\n",
    "            # 修改其他exon的TES, start/end\n",
    "            total.exonDict[exonId].TES = total.exonDict[exonIdTarget].TES\n",
    "            if total.exonDict[exonId].strand == '+':\n",
    "                total.exonDict[exonId].end = total.exonDict[exonIdTarget].end\n",
    "            else:\n",
    "                total.exonDict[exonId].start = total.exonDict[exonIdTarget].start\n",
    "            # 修改status\n",
    "            # 不应该修改status, 因为无已校正TSS作为参考, 所以TSS即使进行了合并也并不是可靠的\n",
    "            #total.exonDict[exonId].status = total.exonDict[exonId].status + \"-TES\"\n",
    "            # 删除exonSet中的这些exon\n",
    "            exonSet.discard(exonId)\n",
    "            # 记录这一次校正\n",
    "            exonCorrected[\"integrate TES\"][exonId] = exonIdTarget\n",
    "\n",
    "# 统计TSS或TES已进行合并的exon的数量\n",
    "infoDict[\"exon integrate TSS\"] = len(exonCorrected[\"integrate TSS\"])\n",
    "infoDict[\"exon integrate TES\"] = len(exonCorrected[\"integrate TSS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 更新exon的映射并合并重复的gene, transcript, exon并重新计算表达量\n",
    "total.refresh()\n",
    "total.reIndex()\n",
    "\n",
    "# 重新计算exon, transcript, gene的相对表达量\n",
    "total.computeRelativeExpression()\n",
    "total.computeCellLineExpression()\n",
    "\n",
    "# 整理exon的status\n",
    "for exonObject in total.exonDict.values():\n",
    "    status = exonObject.status\n",
    "    if status == \"CORRECTED-TES-TSS\":\n",
    "        exonObject.status = \"CORRECTED-TSS-TES\"\n",
    "    elif status == \"NOVEL-TES-TSS\":\n",
    "        exonObject.status = \"CORRECTED-TSS-TES\"\n",
    "    elif status == \"NOVEL-TSS\":\n",
    "        exonObject.status = \"CORRECTED-TSS\"\n",
    "    elif status == \"NOVEL-TES\":\n",
    "        exonObject.status = \"CORRECTED-TES\"\n",
    "\n",
    "# 统计exon合并后exon及transcript的数量\n",
    "infoDict[\"total exon num after exon integrate\"] = len(total.exonDict)\n",
    "num=0\n",
    "for exonId, exonObject in total.exonDict.items():\n",
    "    if exonObject.status == \"NOVEL\":\n",
    "        num+=1\n",
    "infoDict[\"novel exon num after exon integrate\"] = num\n",
    "num = 0\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptId in geneObject.transcriptDict.keys():\n",
    "        num += 1\n",
    "infoDict[\"total transcript num after exon integrate\"] = num\n",
    "num = 0\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptObject in geneObject.transcriptDict.values():\n",
    "        if transcriptObject.status == \"NOVEL\":\n",
    "            num +=1\n",
    "infoDict[\"novel transcript num after exon integrate\"] = num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 打印exon, transcript数量变化信息\n",
    "infoDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "之前的exon校正方法效果较差, 在经过校正后仍存在大量的NOVEL类型的exon, 现在更改了exon的校正及合并方法\n",
    "校正方法: \n",
    "    1. 对于每个exon, 寻找基因组注释中该exon的start与end上下游各5bp范围内的所有exon, 共有四种映射类型\n",
    "        1. 双端单映射, 该exon的start及end端都可映射到基因组注释中的同一个exon\n",
    "        2. 双端双映射, 该exon的start及end端可分别映射到基因组注释中的exon\n",
    "        3. 单端单映射, 该exon仅start端可映射到基因组注释中的exon\n",
    "        4. 单端单映射, 该exon仅end端可映射到基因组注释中的exon\n",
    "    2. 筛选参考exon\n",
    "        优先级: 双端单映射 > 双端双映射 > 单端单映射\n",
    "        1. 在这些参考exon的集合中, 优先筛选出双端单映射的exon, 并进一步筛选出相差碱基数最少的exon作为最终的参考exon\n",
    "        2. 其次才筛选双端双映射的exon, 并筛选出相差碱基数最少的exon作为最终的参考exon\n",
    "        3. 最后才筛选单端单映射的exon作为参考exon\n",
    "    3. 校正exon\n",
    "        根据参考exon的信息对当前exon的start及end进行修改\n",
    "合并方法:\n",
    "    1. 对于已记录的exon, 分别筛选出start, end未经过校正的exon\n",
    "    2. 以合并start端为例\n",
    "    3. 寻找其中每个exon的start上下游各5bp范围内的所有exon, 当在该范围内寻找到exon时再以该exon为中心继续寻找其附近的exon, 最终形成一个集合, 这个集合中所有exon的start上下游5bp范围内不再有其他exon\n",
    "    4. 若这个集合中存在已校正的start端, 则将所有未校正的exon的start修改为距离其最近的已校正的start\n",
    "    5. 若这个集合中不存在已校正的start端, 则对每个exon取其在所有细胞系中的平均表达值, 选择表达最高的exon的start作为集合中所有exon的start\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# debug\n",
    "#with open(\"10002_exonCorrected.pkl\", \"wb\") as file:\n",
    "#    pickle.dump(total, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "# debug\n",
    "with open(\"10002_exonCorrected.pkl\", \"rb\") as file:\n",
    "    total = pickle.load(file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "根据参考基因组进行校正的思路\n",
    "0. 处理exon\n",
    "    0.0 声明变量\n",
    "        exonCorrected, dict, {\"mapped\": {<novel exon id>: <known exon id>, ...},\n",
    "                              \"TSS\": {<novel exon id>: <known exon id>, ...},\n",
    "                              \"TES\": {<novel exon id>: <known exon id>, ...},\n",
    "                              \"TSS-TES\": {<novel exon id>: <known exon id>, ...}}\n",
    "        readyCorList, list, 存储status为NOVEL的exonId[<exonId>, ...]\n",
    "    0.1 引用变量\n",
    "        EXONRANGE, int, exon的start及end允许的偏差范围\n",
    "    已弃置 0.2 根据end建立索引\n",
    "    0.3 根据参考基因组中exon数据对exon的TSS及TES进行校正\n",
    "        因为三代测序中出现的错误绝大部分属于碱基的插入/缺失, 所以需要对exon的TSS/TES进行校正\n",
    "        读取参考基因组注释中exon的数据, 并建立index\n",
    "        统计校正前exon及transcript的数量\n",
    "        遍历total.exonDict中所有exon对象, 判断每个exon上下游exonRange范围内是否存在参考exon\n",
    "        声明list类型变量exonList, 存储该exon所能映射的所有参考exon, [[<exonRef>, <bool>, <bool>], ...]\n",
    "        准备校正exon的TSS及TES\n",
    "            获取[TSS-5, TES+5]范围的所有exon\n",
    "                从中筛选出[TES-5, TES+5]范围的所有exon, 认为这些exon为可双端映射的exon\n",
    "                未筛选出的exon则认为是仅可映射到TSS的exon\n",
    "            获取[TES-5, TES+5]范围的所有exon\n",
    "                从中筛选出(-oo, TSS-5)U(TSS+5, +oo)的所有exon, 认为这些exon为仅可映射到TES的exon\n",
    "            在获取的所有exon中\n",
    "                优先筛选出可双端映射的exon, 并在这些exon中选择总体距离最近的exon进行映射\n",
    "                其次才筛选出仅可单端映射的exon, 并在这些exon中选择离单端距离最近的exon进行映射\n",
    "        校正过程\n",
    "            若不存在映射, 则跳过该exon, 分析下一个exon\n",
    "            若存在双端映射exon\n",
    "                获取相差碱基数最少的exon作为参考exon\n",
    "                判断exonId是否相同\n",
    "                若exonId相同, 则修改status, exonVersion, start, TSS, end, TES\n",
    "                若exonId不同, 则查询参考exonId是否已记录\n",
    "                    若参考exonId已记录, 则合并两exon\n",
    "                    若参考exonId未记录\n",
    "                        exonDict中新增参考exon对象\n",
    "                        合并原exon与参考exon\n",
    "                    将这种映射关系添加到exonCorrected[\"mapped\"]中\n",
    "            若仅存在单端映射exon\n",
    "                如果status为KNOWN就不再进行映射, 跳过该exon\n",
    "                修改exon的status为CORRECTED\n",
    "                判断exon能否双端映射到两个exon\n",
    "                    能映射到两个exon\n",
    "                        在exon的status中添加-TSS-TES后缀\n",
    "                        获取TSS端相差碱基数最少的exon\n",
    "                        修改exon的start,end,TSS的位置\n",
    "                        获取TES端相差碱基数最少的exon\n",
    "                        修改exon的start,end,TES的位置\n",
    "                        将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TSS-TES\"]中\n",
    "                    不能映射到两个exon, 则进行下一步\n",
    "                筛选出相差碱基数最少的exon进行映射\n",
    "                若仅TSS端可映射\n",
    "                    在exon的status中添加-TSS后缀\n",
    "                    修改exon的start,end,TSS的位置\n",
    "                    将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TSS\"]中\n",
    "                若仅TES端可映射\n",
    "                    在exon的status中添加-TES后缀\n",
    "                    修改exon的start,end,TES的位置\n",
    "                    将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"TES\"]中\n",
    "    已弃置 因为已根据参考基因组进行了校正\n",
    "        0.3 处理total中的exon数据\n",
    "        遍历total.exonDict中所有.status==\"NOVEL\"的exon, 记录这些exon\n",
    "        声明list类型变量exonList, 存储该exon所能映射的所有参考exon, [[<exonRef>, <bool>, <bool>], ...]\n",
    "        准备校正exon的start及end\n",
    "            获取[start-5, start+5]范围的所有exon\n",
    "                从中筛选出[end-5, end+5]范围的所有exon, 认为这些exon为可双端映射的exon\n",
    "                未筛选出的exon则认为是仅可映射到start的exon\n",
    "            获取[end-5, end+5]范围的所有exon\n",
    "                从中筛选出(-oo, start-5)U(start+5, +oo)的所有exon, 认为这些exon为仅可映射到end的exon\n",
    "            在获取的所有exon中\n",
    "                优先筛选出可双端映射的exon, 并在这些exon中选择总体距离最近的exon进行映射\n",
    "                其次才筛选出仅可单端映射的exon, 并在这些exon中选择离单端距离最近的exon进行映射\n",
    "        校正过程\n",
    "            若exon可双端映射到一个exon\n",
    "                获取相差碱基数最少的exon\n",
    "                在total.exonExisted中添加该映射\n",
    "                删除total.exonDict中该exon对象\n",
    "                将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"mapped\"]中\n",
    "            若exon不可双端映射到一个exon\n",
    "                修改exon的status为CORRECTED\n",
    "                判断exon能否双端映射到两个exon\n",
    "                    能映射到两个exon\n",
    "                        在exon的status中添加-start-end后缀\n",
    "                        获取start端相差碱基数最少的exon\n",
    "                        修改exon的start的位置\n",
    "                        获取end端相差碱基数最少的exon\n",
    "                        修改exon的end的位置\n",
    "                        将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"start-end\"]中\n",
    "                    不能映射到两个exon, 则进行下一步\n",
    "                筛选出相差碱基数最少的exon进行映射\n",
    "                若仅start端可映射\n",
    "                    在exon的status中添加-start后缀\n",
    "                    修改exon的start的位置\n",
    "                    将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"start\"]中\n",
    "                若仅end端可映射\n",
    "                    在exon的status中添加-end后缀\n",
    "                    修改exon的end的位置\n",
    "                    将映射关系{<exonId>: <newExonId>}保存到exonCorrected[\"end\"]中\n",
    "    0.4 更新total中的exonDict及exonIndex\n",
    "        更新transcriptDict以合并可能重复的transcript(因为exon组成可能重复)\n",
    "        重建Index, 合并重复的exon及transcript\n",
    "    0.5 合并相邻的exon, 即合并相差EXONRANGE范围内的NOVEL的exon\n",
    "        情况说明\n",
    "            在合并距离较近的exon时可能会出现这么一种情况\n",
    "            当EXONRANGE=5时, 200可以合并到205, 205又可以合并到210, 所以200就可以合并到210\n",
    "            因此需要对这种情况进行处理, 而不是简单的将两exon根据EXONRANGE进行合并\n",
    "            所以需要先形成一个簇, 然后对这个簇中所有的exon的TSS或TES进行校正\n",
    "        准备记录合并记录 \n",
    "        建立index\n",
    "            声明indexTSS, {<chr>: {<TSS>: {<TES>: <exonId>, ...}, ...}, ...}\n",
    "            声明indexTES, {<chr>: {<TES>: {<TSS>: <exonId>, ...}, ...}, ...}\n",
    "            遍历total.exonDict\n",
    "                过滤掉所有status==KNOWN的exon\n",
    "                若status中不含-TSS, 则在indexTSS中添加该exon信息\n",
    "                若status中不含-TES, 则在indexTES中添加该exon信息\n",
    "                否则, continue\n",
    "        遍历total.exonDict中所有status不为KNOWN及其中一端未校正的exon\n",
    "            处理TSS\n",
    "                声明list变量exonSet, {<exonId>, ...}\n",
    "                遍历所有exon, 去除KNOWN或TSS已校正的exon, 将剩余的exon添加到exonSet\n",
    "                只要exonSet不为空集, 就随机提取exonSet中的一个exonId\n",
    "                    temp中添加所有与该exon的TSS相同的exon\n",
    "                    寻找其TSS上下游各EXONRANGE范围内的exon\n",
    "                        若寻找不到其他的exon, 就认为该exon的TSS无法修改, 跳过该exon\n",
    "                        若能寻找到其他的exon, 就进行下一步\n",
    "                判断在temp中是否存在已校正端\n",
    "                    分离其中的未校正端与已校正端\n",
    "                    若存在已校正端\n",
    "                        对每一个未校正端选择其距离最近的已校正端作为参考位点进行校正\n",
    "                            寻找最近的已校正端\n",
    "                            修改TSS, start/end位置\n",
    "                            改变status\n",
    "                            删除exonSet中该exonId\n",
    "                        结束此次循环\n",
    "                    若不存在已校正端, 则进行下一步\n",
    "                以expression最大的exon为参照, 修改其他exon的TSS\n",
    "                    寻找expression最大的exon\n",
    "                    修改其他exon的TSS, start/end\n",
    "                    修改status\n",
    "                    删除exonSet中的这些exon\n",
    "                删除exonSet中的这些exon\n",
    "                示例\n",
    "                    权重: 相对表达量\n",
    "                        假设有7个exon, 从start=100开始寻找EXONRANGE=5范围的exonStart\n",
    "                        exon0   95  200\n",
    "                        exon1   99  201\n",
    "                        exon2   99  200\n",
    "                        exon3   100 200\n",
    "                        exon4   104 204\n",
    "                        exon5   106 206\n",
    "                        exon6   107 210\n",
    "                        exon7   110 210\n",
    "                    分别递增, 递减寻找EXONRANGE范围内的exon\n",
    "                        当递减寻找[100-5, 100)的start时, 寻找到了99\n",
    "                            temp中添加[exon1, expression]\n",
    "                            temp中添加[exon2, expression]\n",
    "                            此时再根据99去递减寻找[99-5, 99)范围的start, 寻找到了95\n",
    "                                temp中添加[exon0, expression]\n",
    "                            此时再根据95去递减寻找[95-5, 95)范围的start, 寻找到了None\n",
    "                                当寻找到None时, 该寻找过程终止\n",
    "                        当递增寻找(100, 100+5]的start时, 寻找到了104\n",
    "                            temp中添加[exon4, expression]\n",
    "                            此时再根据104去递增寻找(104, 104+5]范围的start, 寻找到了106\n",
    "                                temp中添加[exon5, expression]\n",
    "                            此时再根据106去递增寻找(106, 106+5]范围的start, 寻找到了107\n",
    "                                temp中添加[exon6, expression]\n",
    "                            ...\n",
    "                            此时再根据***去递增寻找(***, ***+5]范围的start, 寻找到了None\n",
    "                                当寻找到None时, 寻找过程终止\n",
    "                    当两个寻找过程都停止后, 则开始对temp中的exon的start进行校正\n",
    "                        筛选temp中expression最大的exon, 作为参考exon\n",
    "                        其他exon的start都将修改为参考exon的start\n",
    "            处理TES\n",
    "                声明list变量temp, [[<end>, <expression>], ...]\n",
    "                权重: 相对表达量\n",
    "                分别递增, 递减寻找EXONRANGE范围内的exon, 在temp中添加[exon, expression]\n",
    "                当两个寻找过程都停止后, 则开始对temp中的exon的end进行校正\n",
    "                    与start处理过程相似\n",
    "    0.6 在所有exon都处理完后\n",
    "        更新transcriptDict以合并可能重复的transcript(因为exon组成可能重复)\n",
    "        重建Index\n",
    "1. 处理transcript\n",
    "    1.0 声明变量\n",
    "        transcriptSet, set, 用于transcriptId {<transcriptId>, ...}\n",
    "        transcriptCorrected, set, 用于存储校正过程中transcript的变化情况\n",
    "        transcriptGeneMap, dict, 存储transcriptId与geneId的映射关系\n",
    "        totalTranscriptIndex, dict, 准备total中transcript的index\n",
    "    1.1 读取参考基因组注释数据\n",
    "        获取所有transcript信息\n",
    "        补充transcript中的exon信息\n",
    "        获取index, {\"TSS\": {<chr>: {<TSS>: {<TES>: [<transcriptId>, ...],\n",
    "                                                         ...},\n",
    "                                                 ...},\n",
    "                                         ...},\n",
    "                                 \"TES\": {<chr>: {<TES>: {<TSS>: [<transcriptId>, ...],\n",
    "                                                         ...},\n",
    "                                                 ...},\n",
    "                                         ...}\n",
    "                                 }\n",
    "    1.2 统计校正前及transcript的数量\n",
    "    1.3 遍历所有transcript, 将所有transcriptId存储至变量transcriptSet中\n",
    "    1.4 寻找transcriptSet中可映射的transcript\n",
    "        声明变量transcriptList, 存储该transcript所能映射的所有参考transcript, [[<transcriptRef>, <num>, <bool>, <bool>], ...]\n",
    "        根据TSS获取其上下游rangeTSS范围内的transcript\n",
    "            在这些transcript中, 筛选TES上下游在rangeTES范围内的transcript, 认为这些transcript是可双端映射的transcript\n",
    "            剩余的transcript则被认为是仅可校正TSS端的transcript\n",
    "        根据TES获取其上下游rangeTES范围内的transcript\n",
    "            在这些transcript中, 筛选TSS上下游不在rangeTSS范围内的transcript, 认为这些transcript是仅可校正TES端的transcript\n",
    "    1.5 准备映射\n",
    "        判断是否存在映射, 若不存在映射则跳过该transcript, 若存在映射则进行下一步\n",
    "        判断映射的类型\n",
    "            判断是否仅一个映射\n",
    "                若仅单个映射就直接存储为transcriptRefSet={(<transcriptId>, <num>, <bool>, <bool>, <marker>, <mapped id>)}\n",
    "            优先筛选出双端映射的transcript\n",
    "                最优先筛选出transcriptId相同的transcript作为参考transcript\n",
    "                其次筛选出exon组成最相似的transcript作为参考transcript\n",
    "                    exon组成相似程度=参考transcript与该transcript的exon的重叠数量\n",
    "                    若exon组成相似程度相同, 则筛选出TSS与TES相差碱基数最少的transcript作为参考transcript\n",
    "                参考transcript存储为transcriptRefSet={(<transcriptId>, <num>, True, True, <marker>, <mapped id>)}\n",
    "            无双端映射transcript, 仅有单端映射transcript\n",
    "                transcriptRefSet=set()  # 可能有一个元素, 也可能有两个元素\n",
    "                TSS端\n",
    "                    最优先筛选出transcriptId相同的transcript作为参考transcript\n",
    "                    其次筛选出exon组成最相似的transcript作为参考transcript\n",
    "                        exon组成相似程度=参考transcript与该transcript的exon的重叠数量\n",
    "                        若exon组成相似程度相同, 则筛选出TSS与TES相差碱基数最少的transcript作为参考transcript\n",
    "                    transcriptRefSet中添加(<transcriptIdRef>, True, False, <marker>, <mapped id>)\n",
    "                TES端\n",
    "                    最优先筛选出transcriptId相同的transcript作为参考transcript\n",
    "                    其次筛选出exon组成最相似的transcript作为参考transcript\n",
    "                        exon组成相似程度=参考transcript与该transcript的exon的重叠数量\n",
    "                        若exon组成相似程度相同, 则筛选出TSS与TES相差碱基数最少的transcript作为参考transcript\n",
    "                    transcriptRefSet中添加(<transcriptIdRef>, False, True, <marker>, <mapped id>)\n",
    "        遍历transcriptRefSet, 检查参考transcript是否已记录\n",
    "            检查transcriptId是否已记录\n",
    "                若参考transcriptId已记录, 认为该transcript已记录\n",
    "                    准备合并两transcript\n",
    "                    marker = 1\n",
    "                    mapped id = None\n",
    "                若参考transcriptId未记录, 则进行下一步\n",
    "            检查geneObject.transcriptIndex中是否已记录该transcript\n",
    "                若已记录, 则认为该transcript已记录\n",
    "                    准备合并两transcript\n",
    "                    marker = 2\n",
    "                    mapped id = 参考transcript被记录的Id\n",
    "                若未记录, 则认为该transcript未记录, 进行下一步\n",
    "                    marker = 3\n",
    "                    mapped id = None\n",
    "    1.6 开始映射\n",
    "        检查transcriptRefSet的长度为2则分别映射到两transcript, 否则为映射到单个transcript\n",
    "        若长度为2, 分别映射到两transcript\n",
    "            检查status, 若status为KNOWN就不再映射, 否则进行下一步\n",
    "            遍历transcriptRefSet\n",
    "                修改transcript的TSS, start/end\n",
    "                修改transcript的TES, start/end\n",
    "                修改transcript的status为CORRECTED-TSS-TES\n",
    "                记录该映射到transcriptCorrected[\"TSS-TES\"]\n",
    "        若长度为1, 映射到单个transcript\n",
    "            若双端映射\n",
    "                检查参考transcript的记录状态marker\n",
    "                    若参考transcript已记录\n",
    "                        合并两transcript\n",
    "                        记录这一次合并到transcriptCorrected[\"mapped\"]\n",
    "                    若参考transcript未记录\n",
    "                        在对应的geneRefObject中添加参考transcript对象\n",
    "                        合并两transcript\n",
    "                        记录这一次合并到transcriptCorrected[\"mapped\"]\n",
    "            若单端映射\n",
    "                检查status, 若status为KNOWN就不再映射, 否则进行下一步\n",
    "                检查参考transcript的记录状态marker\n",
    "                    若仅TSS端可映射\n",
    "                        修改transcript的TSS, start/end\n",
    "                        修改transcript的status为CORRECTED-TSS\n",
    "                        记录该映射到transcriptCorrected[\"TSS\"]\n",
    "                    若仅TES端可映射\n",
    "                        修改transcript的TES, start/end\n",
    "                        修改transcript的status为CORRECTED-TES\n",
    "                        记录该映射到transcriptCorrected[\"TES\"]\n",
    "    1.7 重建index, 重新计算relativeExpression及cellLineExpression\n",
    "2. 处理gene\n",
    "3. 重新计算relativeExpression及cellLineExpression\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 声明变量\n",
    "transcriptSet = set()  # {(<geneId>, <transcriptId>), ...}\n",
    "transcriptCorrected = {\"mapped\": {}, \"TSS-TES\": {}, \"TSS\": {}, \"TES\": {}, \"integrate TSS\": {}, \"integrate TES\": {}}\n",
    "# 准备transcriptId与geneId的映射关系\n",
    "transcriptGeneMap = {}\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptId, transcriptObject in geneObject.transcriptDict.items():\n",
    "        transcriptGeneMap[transcriptId] = geneId\n",
    "transcriptGeneMapKeys = set(transcriptGeneMap.keys())\n",
    "# 准备total中transcript的index\n",
    "totalTranscriptIndex = total.transcriptIndexBuild(siteType=\"TSS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 读取参考基因组注释中transcript数据\n",
    "temp = loadRefAnnotation(filename=INPUTREFANNOTATION, allowType=\"transcript\")\n",
    "# 获取所有transcript信息\n",
    "transcriptRef = temp[\"dataRef\"]\n",
    "# 获取index\n",
    "transcriptIndex = temp[\"positionRef\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 读取参考基因组注释中gene数据\n",
    "temp = loadRefAnnotation(filename=INPUTREFANNOTATION, allowType=\"gene\")\n",
    "# 获取所有transcript信息\n",
    "geneRef = temp[\"dataRef\"]\n",
    "# 获取index\n",
    "geneIndex = temp[\"positionRef\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 读取参考基因组注释中exon数据\n",
    "temp = loadRefAnnotation(filename=INPUTREFANNOTATION, allowType=\"exon\")\n",
    "exonRef = temp[\"dataRef\"]\n",
    "exonIndex = temp[\"positionRef\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 补充transcript中的exon信息\n",
    "keys = set(transcriptRef)\n",
    "for exonId, exonDict in exonRef.items():\n",
    "    for geneId, transcriptId in exonDict[\"geneTranscriptIdSet\"]:\n",
    "        exonSet = transcriptRef[transcriptId].setdefault(\"exonSet\", set())\n",
    "        exonSet.add(exonId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 统计校正前transcript的数量\n",
    "num = 0\n",
    "for geneObject in total.geneDict.values():\n",
    "    for transcriptObject in geneObject.transcriptDict.values():\n",
    "        num += 1\n",
    "infoDict[\"total transcript num before transcript correct\"] = num\n",
    "num = 0\n",
    "for geneObject in total.geneDict.values():\n",
    "    for transcriptObject in geneObject.transcriptDict.values():\n",
    "        if transcriptObject.status == \"NOVEL\":\n",
    "            num += 1\n",
    "infoDict[\"novel transcript num before transcript correct\"] = num"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 遍历所有transcript, 将所有transcriptId存储至变量transcriptSet中\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptId, transcriptObject in geneObject.transcriptDict.items():\n",
    "        if transcriptObject.status == \"NOVEL\":\n",
    "            transcriptSet.add((geneId, transcriptId))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "transcript:   1%|          | 1074/183842 [00:00<00:22, 8128.64it/s]\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'ENSG00000290832'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32mf:\\OneDrive\\Master\\Project\\trans\\scripts\\10002_correct.ipynb 单元格 31\u001b[0m line \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=197'>198</a>\u001b[0m \u001b[39mif\u001b[39;00m markerTSS \u001b[39m==\u001b[39m markerTES \u001b[39m==\u001b[39m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=198'>199</a>\u001b[0m     \u001b[39m# 双端映射\u001b[39;00m\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=199'>200</a>\u001b[0m     \u001b[39mif\u001b[39;00m marker \u001b[39m==\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=200'>201</a>\u001b[0m         \u001b[39m# 参考transcriptId已记录\u001b[39;00m\n\u001b[1;32m--> <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=201'>202</a>\u001b[0m         total\u001b[39m.\u001b[39;49m_transcriptMerge(transcriptId1\u001b[39m=\u001b[39;49mtranscriptIdRef,\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=202'>203</a>\u001b[0m                                transcriptId2\u001b[39m=\u001b[39;49mtranscriptId,\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=203'>204</a>\u001b[0m                                geneId1\u001b[39m=\u001b[39;49mtranscriptRef[transcriptIdRef][\u001b[39m\"\u001b[39;49m\u001b[39mgeneId\u001b[39;49m\u001b[39m\"\u001b[39;49m],\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=204'>205</a>\u001b[0m                                geneId2\u001b[39m=\u001b[39;49mgeneId)\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=205'>206</a>\u001b[0m     \u001b[39melif\u001b[39;00m marker \u001b[39m==\u001b[39m \u001b[39m2\u001b[39m:\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=206'>207</a>\u001b[0m         \u001b[39m# 参考transcript已记录, id为mapped\u001b[39;00m\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=207'>208</a>\u001b[0m         total\u001b[39m.\u001b[39m_transcriptMerge(transcriptId1\u001b[39m=\u001b[39mmapped,\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=208'>209</a>\u001b[0m                                transcriptId2\u001b[39m=\u001b[39mtranscriptId,\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=209'>210</a>\u001b[0m                                geneId1\u001b[39m=\u001b[39mtranscriptGeneMap[mapped],\n\u001b[0;32m    <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#X42sZmlsZQ%3D%3D?line=210'>211</a>\u001b[0m                                geneId2\u001b[39m=\u001b[39mgeneId)\n",
      "File \u001b[1;32mf:\\OneDrive\\Master\\Project\\trans\\scripts\\Class.py:451\u001b[0m, in \u001b[0;36mTotal._transcriptMerge\u001b[1;34m(self, transcriptId1, transcriptId2, geneId1, geneId2)\u001b[0m\n\u001b[0;32m    447\u001b[0m     \u001b[39mreturn\u001b[39;00m (deletedTranscriptId, remainedTranscriptId)\n\u001b[0;32m    448\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    449\u001b[0m     \u001b[39m# 两transcript位于不同gene\u001b[39;00m\n\u001b[0;32m    450\u001b[0m     \u001b[39m# 检查两transcriptId是否可映射\u001b[39;00m\n\u001b[1;32m--> 451\u001b[0m     \u001b[39mif\u001b[39;00m transcriptId1 \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgeneDict[geneId1]\u001b[39m.\u001b[39mtranscriptExisted\u001b[39m.\u001b[39mkeys():\n\u001b[0;32m    452\u001b[0m         transcriptId1 \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mgeneDict[geneId1]\u001b[39m.\u001b[39mtranscriptExisted[transcriptId1]\n\u001b[0;32m    453\u001b[0m     \u001b[39mif\u001b[39;00m transcriptId2 \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mgeneDict[geneId2]\u001b[39m.\u001b[39mtranscriptExisted\u001b[39m.\u001b[39mkeys():\n",
      "\u001b[1;31mKeyError\u001b[0m: 'ENSG00000290832'"
     ]
    }
   ],
   "source": [
    "for geneId, transcriptId in tqdm.tqdm(transcriptSet, leave=True, desc=\"transcript\"):\n",
    "    # 寻找transcriptSet中可映射的transcript\n",
    "    transcriptObject = total.geneDict[geneId].transcriptDict[transcriptId]\n",
    "    chr = total.geneDict[geneId].chr\n",
    "    TSS = transcriptObject.TSS\n",
    "    TES = transcriptObject.TES\n",
    "    exonSet = set(transcriptObject.exonList)\n",
    "    # 存储该transcript所能映射的所有参考transcript\n",
    "    transcriptList = []  # [[<transcriptRef>, <bool>, <bool>], ...]\n",
    "    # 根据TSS获取其上下游rangeTSS范围内的transcript\n",
    "    temp = [transcriptIndex[\"TSS\"][chr].get(i, None) for i in range(TSS-RANGETSS, TSS+RANGETSS+1)]\n",
    "    # 去除空值\n",
    "    while None in temp:\n",
    "        temp.remove(None)\n",
    "    while {} in temp:\n",
    "        temp.remove({})\n",
    "        pass\n",
    "    # 拆解list\n",
    "    if temp:\n",
    "        temp = [{tempTES: tempId} for d in temp  for tempTES, tempIdList in d.items() for tempId in tempIdList]\n",
    "    for d in temp:\n",
    "        for tempTES, tempTranscriptId in d.items():\n",
    "            if abs(tempTES-TES) <= RANGETES:\n",
    "                # 从中筛选出[TES-RANGETES, TES+RANGETES]范围的所有transcript, 认为这些transcript为可双端映射的transcript\n",
    "                num = abs(TSS-transcriptRef[tempTranscriptId][\"TSS\"]) + abs(TES-tempTES)\n",
    "                transcriptList.append((tempTranscriptId, num, True, True))\n",
    "            else:\n",
    "                # 仅TSS可映射到tempTranscript\n",
    "                num = abs(TSS-transcriptRef[tempTranscriptId][\"TSS\"])\n",
    "                transcriptList.append((tempTranscriptId, num, True, False))\n",
    "    # 根据TES获取其上下游rangeTES范围内的transcript\n",
    "    temp = [transcriptIndex[\"TES\"][chr].get(i, None) for i in range(TES-RANGETES, TES+RANGETES+1)]\n",
    "    # 去除空值\n",
    "    while None in temp:\n",
    "        temp.remove(None)\n",
    "    while {} in temp:\n",
    "        temp.remove({})\n",
    "        pass\n",
    "    # 拆解list\n",
    "    if temp:\n",
    "        temp = [{tempTES: tempId} for d in temp  for tempTES, tempIdList in d.items() for tempId in tempIdList]\n",
    "    for d in temp:\n",
    "        for tempTSS, tempTranscriptId in d.items():\n",
    "            # 从中筛选出(-oo, TSS-RANGETSS)U(TSS+RANGETSS, +oo)的所有transcript, 认为这些transcript为仅可映射到TES的transcript\n",
    "            if abs(tempTSS-TSS) > RANGETSS:\n",
    "                # 仅TES可映射到tempTranscript\n",
    "                num = abs(TES-transcriptRef[tempTranscriptId][\"TES\"])\n",
    "                transcriptList.append((tempTranscriptId, num, False, True))\n",
    "    # 准备映射\n",
    "    # 判断是否存在映射\n",
    "    if not transcriptList:\n",
    "        # 不存在映射, 跳过该transcript\n",
    "        continue\n",
    "    # 判断映射的类型\n",
    "    # 判断是否仅一个映射\n",
    "    if len(transcriptList) == 1:\n",
    "        # 仅单个映射仅单个映射\n",
    "        (transcriptIdRef, num, markerTSS, markerTES) = transcriptList[0]\n",
    "        transcriptRefSet = {(transcriptIdRef, num, markerTSS, markerTES, None, None)}\n",
    "    else:\n",
    "        # 优先筛选出双端映射的transcript\n",
    "        doubleMapped = []\n",
    "        for (transcriptIdRef, num, markerTSS, markerTES) in transcriptList:\n",
    "            if markerTSS == markerTES == True:\n",
    "                doubleMapped.append([transcriptIdRef, num, markerTSS, markerTES])\n",
    "        # 判断是否存在双端映射的transcript\n",
    "        if doubleMapped:\n",
    "            # 存在双端映射的transcript\n",
    "            doubleMapped = sorted(doubleMapped, key=lambda x: x[1])\n",
    "            # 最优先筛选出transcriptId相同的transcript作为参考transcript\n",
    "            sameTranscriptId = False\n",
    "            for [transcriptIdRef, num, markerTSS, markerTES] in doubleMapped:\n",
    "                if transcriptId == transcriptIdRef:\n",
    "                    # 存在transcriptId相同的项\n",
    "                    transcriptRefSet = {(transcriptIdRef, num, markerTSS, markerTES, None, None)}\n",
    "                    sameTranscriptId = True\n",
    "                    break\n",
    "            if sameTranscriptId is True:\n",
    "                pass\n",
    "            else:\n",
    "                # 不存在transcriptId相同的项\n",
    "                # 其次筛选出exon组成最相似的transcript\n",
    "                # exon组成相似程度=参考transcript与该transcript的exon的重叠数量\n",
    "                exonCombinationValueList = [(transcriptIdRef, num, len(transcriptRef[transcriptIdRef][\"exonSet\"].intersection(exonSet))) for [transcriptIdRef, num, markerTSS, markerTES] in doubleMapped]\n",
    "                # 若exon组成相似程度相同, 则筛选出TSS与TES相差碱基数最少的transcript\n",
    "                exonCombinationValueList = sorted(exonCombinationValueList, key=lambda x: (-x[2], x[1]))\n",
    "                transcriptRefSet = {(transcriptIdRef, num, markerTSS, markerTES, None, None) for [transcriptIdRef, num, markerTSS, markerTES] in doubleMapped if transcriptIdRef==exonCombinationValueList[0][0]}\n",
    "        else:\n",
    "            # 不存在双端映射的transcript\n",
    "            transcriptRefSet=set()  # 可能有一个元素, 也可能有两个元素\n",
    "            # TSS端\n",
    "            transcriptTSSList = [(transcriptIdRef, num, markerTSS, markerTES) for (transcriptIdRef, num, markerTSS, markerTES) in transcriptList if markerTSS is True]\n",
    "            # 最优先筛选出transcriptId相同的transcript作为参考transcript\n",
    "            sameTranscriptId = False\n",
    "            for (transcriptIdRef, num, markerTSS, markerTES) in transcriptTSSList:\n",
    "                if transcriptId == transcriptIdRef:\n",
    "                    # 存在transcriptId相同的项\n",
    "                    transcriptRefSet = transcriptRefSet.union({(transcriptId, num, markerTSS, markerTES, None, None)})\n",
    "                    sameTranscriptId = True\n",
    "                    break\n",
    "            if sameTranscriptId is True:\n",
    "                pass\n",
    "            else:\n",
    "                # 不存在transcriptId相同的项\n",
    "                # 其次筛选出exon组成最相似的transcript\n",
    "                # exon组成相似程度=参考transcript与该transcript的exon的重叠数量\n",
    "                exonCombinationValueList = [(transcriptIdRef, num, len(transcriptRef[transcriptIdRef][\"exonSet\"].intersection(exonSet))) for (transcriptIdRef, num, markerTSS, markerTES) in transcriptTSSList]\n",
    "                # 若exon组成相似程度相同, 则筛选出TSS与TES相差碱基数最少的transcript\n",
    "                exonCombinationValueList = sorted(exonCombinationValueList, key=lambda x: (-x[2], x[1]))\n",
    "                transcriptRefSet = transcriptRefSet.union({(transcriptIdRef, num, markerTSS, markerTES, None, None) for [transcriptIdRef, num, markerTSS, markerTES] in transcriptTSSList if transcriptIdRef==exonCombinationValueList[0][0]})\n",
    "            # TES端\n",
    "            transcriptTESList = [(transcriptIdRef, num, markerTSS, markerTES) for (transcriptIdRef, num, markerTSS, markerTES) in transcriptList if markerTES is True]\n",
    "            # 最优先筛选出transcriptId相同的transcript作为参考transcript\n",
    "            sameTranscriptId = False\n",
    "            for (transcriptIdRef, num, markerTSS, markerTES) in transcriptTESList:\n",
    "                if transcriptId == transcriptIdRef:\n",
    "                    # 存在transcriptId相同的项\n",
    "                    transcriptRefSet = transcriptRefSet.union({(transcriptId, num, markerTSS, markerTES, None, None)})\n",
    "                    sameTranscriptId = True\n",
    "                    break\n",
    "            if sameTranscriptId is True:\n",
    "                pass\n",
    "            else:\n",
    "                # 不存在transcriptId相同的项\n",
    "                # 其次筛选出exon组成最相似的transcript\n",
    "                # exon组成相似程度=参考transcript与该transcript的exon的重叠数量\n",
    "                exonCombinationValueList = [(transcriptIdRef, num, len(transcriptRef[transcriptIdRef][\"exonSet\"].intersection(exonSet))) for (transcriptIdRef, num, markerTSS, markerTES) in transcriptTESList]\n",
    "                # 若exon组成相似程度相同, 则筛选出TSS与TES相差碱基数最少的transcript\n",
    "                exonCombinationValueList = sorted(exonCombinationValueList, key=lambda x: (-x[2], x[1]))\n",
    "                transcriptRefSet = transcriptRefSet.union({(transcriptIdRef, num, markerTSS, markerTES, None, None) for [transcriptIdRef, num, markerTSS, markerTES] in transcriptTESList if transcriptIdRef==exonCombinationValueList[0][0]})\n",
    "    # 检查参考transcript是否已记录\n",
    "    temp = [[transcriptIdRef, num, markerTSS, markerTES, marker, mapped] for (transcriptIdRef, num, markerTSS, markerTES, marker, mapped) in transcriptRefSet]\n",
    "    for index, value in enumerate(temp):\n",
    "        transcriptIdRef = value[0]\n",
    "        if transcriptIdRef in transcriptGeneMapKeys:\n",
    "            # 参考transcriptId已记录, 认为该transcript已记录\n",
    "            temp[index][4] = 1\n",
    "            temp[index][5] = None\n",
    "        else:\n",
    "            # 检查geneObject.transcriptIndex中是否已记录该transcript\n",
    "            transcriptRefChr = transcriptRef[transcriptIdRef][\"chr\"]\n",
    "            transcriptRefTSS = transcriptRef[transcriptIdRef][\"TSS\"]\n",
    "            transcriptRefTES = transcriptRef[transcriptIdRef][\"TES\"]\n",
    "            transcriptRefExonSet = transcriptRef[transcriptIdRef][\"exonSet\"]\n",
    "            if transcriptRefTSS not in totalTranscriptIndex[transcriptRefChr].keys():\n",
    "                # 未记录该transcript\n",
    "                temp[index][4] = 3\n",
    "                temp[index][5] = None\n",
    "            else:\n",
    "                if transcriptRefTES not in totalTranscriptIndex[transcriptRefChr][transcriptRefTSS].keys():\n",
    "                    # 未记录该transcript\n",
    "                    temp[index][4] = 3\n",
    "                    temp[index][5] = None\n",
    "                else:\n",
    "                    markerBreak = False\n",
    "                    transcriptIdListTemp = totalTranscriptIndex[transcriptRefChr][transcriptRefTSS][transcriptRefTES]\n",
    "                    for transcriptIdTemp in transcriptIdListTemp:\n",
    "                        transcriptTempExon = set(total.geneDict[transcriptGeneMap[transcriptIdTemp]].transcriptDict[transcriptIdTemp].exonList)\n",
    "                        if transcriptTempExon == transcriptRefExonSet:\n",
    "                            # 已记录该transcript\n",
    "                            markerBreak = True\n",
    "                            temp[index][4] = 2\n",
    "                            temp[index][5] = transcriptIdTemp\n",
    "                            break\n",
    "                    if markerBreak is True:\n",
    "                        pass\n",
    "                    else:\n",
    "                        # 未记录该transcript\n",
    "                        temp[index][4] = 3\n",
    "                        temp[index][5] = None\n",
    "    transcriptRefSet = {(transcriptIdRef, num, markerTSS, markerTES, marker, mapped) for [transcriptIdRef, num, markerTSS, markerTES, marker, mapped] in temp}\n",
    "    # 开始映射\n",
    "    if len(transcriptRefSet) == 2:\n",
    "        # 分别映射到两transcript\n",
    "        if transcriptObject.status == \"KNOWN\":\n",
    "            continue\n",
    "        transcriptCorrected[\"TSS-TES\"] = {transcriptId: {\"TSS\": None, \"TES\": None}}\n",
    "        # 对transcript进行校正\n",
    "        total.geneDict[geneId].transcriptDict[transcriptId].status = \"CORRECTED-TSS-TES\"\n",
    "        for (transcriptIdRef, num, markerTSS, markerTES, marker, mapped) in transcriptRefSet:\n",
    "            if markerTSS is True:\n",
    "                total.geneDict[geneId].transcriptDict[transcriptId].TSS = transcriptRef[transcriptIdRef][\"TSS\"]\n",
    "                if total.geneDict[geneId].strand == '+':\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].start = transcriptRef[transcriptIdRef][\"start\"]\n",
    "                else:\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].end = transcriptRef[transcriptIdRef][\"end\"]\n",
    "                transcriptCorrected[\"TSS-TES\"][transcriptId][\"TSS\"] = transcriptIdRef\n",
    "            elif markerTES is True:\n",
    "                total.geneDict[geneId].transcriptDict[transcriptId].TES = transcriptRef[transcriptIdRef][\"TES\"]\n",
    "                if total.geneDict[geneId].strand == '+':\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].end = transcriptRef[transcriptIdRef][\"end\"]\n",
    "                else:\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].start = transcriptRef[transcriptIdRef][\"start\"]\n",
    "                transcriptCorrected[\"TSS-TES\"][transcriptId][\"TES\"] = transcriptIdRef\n",
    "    else:\n",
    "        # 映射到单个transcript\n",
    "        (transcriptIdRef, num, markerTSS, markerTES, marker, mapped) = transcriptRefSet.pop()\n",
    "        if markerTSS == markerTES == True:\n",
    "            # 双端映射\n",
    "            if marker == 1:\n",
    "                # 参考transcriptId已记录\n",
    "                # 查询参考transcript所属gene是否已记录\n",
    "                geneIdRef = transcriptRef[transcriptIdRef][\"geneId\"]\n",
    "                total._transcriptMerge(transcriptId1=transcriptIdRef,\n",
    "                                       transcriptId2=transcriptId,\n",
    "                                       geneId1=geneIdRef,\n",
    "                                       geneId2=geneId)\n",
    "            elif marker == 2:\n",
    "                # 参考transcript已记录, id为mapped\n",
    "                total._transcriptMerge(transcriptId1=mapped,\n",
    "                                       transcriptId2=transcriptId,\n",
    "                                       geneId1=transcriptGeneMap[mapped],\n",
    "                                       geneId2=geneId)\n",
    "            else:\n",
    "                # 参考transcript未记录\n",
    "                geneIdRef = transcriptRef[transcriptIdRef][\"geneId\"]\n",
    "                if geneIdRef not in total.geneDict.keys():\n",
    "                    # 参考transcript所属的参考gene未记录, 需要进行记录\n",
    "                    total.geneDict[geneIdRef] = Gene(status=\"KNOWN\",\n",
    "                                                     chr=transcriptRef[transcriptIdRef][\"chr\"],\n",
    "                                                     strand=transcriptRef[transcriptIdRef][\"strand\"],\n",
    "                                                     start=geneRef[geneIdRef][\"start\"],\n",
    "                                                     end=geneRef[geneIdRef][\"end\"],\n",
    "                                                     TSS=geneRef[geneIdRef][\"TSS\"],\n",
    "                                                     TES=geneRef[geneIdRef][\"TES\"],\n",
    "                                                     geneId=geneIdRef,\n",
    "                                                     geneVersion=geneRef[geneIdRef][\"geneVersion\"],\n",
    "                                                     geneName=geneRef[geneIdRef][\"geneName\"],\n",
    "                                                     geneBiotype=geneRef[geneIdRef][\"geneBiotype\"],)\n",
    "                total.geneDict[geneIdRef].transcriptDict[transcriptIdRef] = Transcript(status=\"KNOWN\",\n",
    "                                                                                        transcriptId=transcriptIdRef,\n",
    "                                                                                        transcriptVersion=transcriptRef[transcriptIdRef][\"transcriptVersion\"],\n",
    "                                                                                        transcriptName=transcriptRef[transcriptIdRef][\"transcriptName\"],\n",
    "                                                                                        transcriptBiotype=transcriptRef[transcriptIdRef][\"transcriptBiotype\"],\n",
    "                                                                                        start=transcriptRef[transcriptIdRef][\"start\"],\n",
    "                                                                                        end=transcriptRef[transcriptIdRef][\"end\"],\n",
    "                                                                                        exonList=list(transcriptRef[transcriptIdRef][\"exonSet\"]))\n",
    "                total._transcriptMerge(transcriptId1=transcriptIdRef,\n",
    "                                       transcriptId2=transcriptId,\n",
    "                                       geneId1=transcriptRef[transcriptIdRef][\"geneId\"],\n",
    "                                       geneId2=geneId)\n",
    "        else:\n",
    "            # 单端映射\n",
    "            # 若status为KNOWN就不再映射\n",
    "            if transcriptObject.status == \"KNOWN\":\n",
    "                continue\n",
    "            if markerTSS is True:\n",
    "                # 仅TSS端可映射\n",
    "                total.geneDict[geneId].transcriptDict[transcriptId].status = \"CORRECTED-TSS\"\n",
    "                total.geneDict[geneId].transcriptDict[transcriptId].TSS = transcriptRef[transcriptIdRef][\"TSS\"]\n",
    "                transcriptCorrected[\"TSS\"][transcriptId] = transcriptIdRef\n",
    "                if total.geneDict[geneId].strand == '+':\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].start = transcriptRef[transcriptIdRef][\"start\"]\n",
    "                else:\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].end = transcriptRef[transcriptIdRef][\"end\"]\n",
    "            else:\n",
    "                # 仅TES端可映射\n",
    "                total.geneDict[geneId].transcriptDict[transcriptId].status = \"CORRECTED-TES\"\n",
    "                total.geneDict[geneId].transcriptDict[transcriptId].TES = transcriptRef[transcriptIdRef][\"TES\"]\n",
    "                if total.geneDict[geneId].strand == '+':\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].end = transcriptRef[transcriptIdRef][\"end\"]\n",
    "                else:\n",
    "                    total.geneDict[geneId].transcriptDict[transcriptId].start = transcriptRef[transcriptIdRef][\"start\"]\n",
    "                transcriptCorrected[\"TES\"][transcriptId] = transcriptIdRef\n",
    "\n",
    "    #print('\\n', geneId, transcriptId)\n",
    "    #print('transcriptList', transcriptList)\n",
    "    #print(\"transcriptRefSet\", transcriptRefSet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'geneId': 'ENSG00000174353', 'geneVersion': 17, 'geneName': 'STAG3L3', 'geneBiotype': 'transcribed_unprocessed_pseudogene', 'status': 'KNOWN', 'strand': '-', 'chr': 'chr7', 'start': 72869658, 'end': 73005915, 'TSS': 73005915, 'TES': 72869658, 'transcript': 'ENCLB240AAXT000349248, ENCLB345FIST000374847, ENCLB240AAXT000349242, ENCLB315YZPT000366823, ENCLB315YZPT000366827, ENCLB315YZPT000366833, ENCLB345FIST000374837, ENST00000448173, ENCLB345FIST000374848, ENCLB315YZPT000366824, ENCLB345FIST000374851, ENCLB315YZPT000366831, ENST00000423834, ENCLB240AAXT000349247, ENCLB240AAXT000349241, ENCLB240AAXT000349246, ENCLB767XEKT000335181, ENST00000428423', 'exon': 'ENSE00003500777, ENSE00001721222, ENSE00001797769, ENSE00001635491, GSM6782552_1131077, GSM6783527_1108042, GSM6782551_1151377, ENSE00002477385, ENSE00002499054, GSM6783528_1159418, ENSE00003710880, ENSE00003709159, ENSE00003708642, ENSE00002578620, ENSE00001798213, ENSE00001599135, ENSE00003708645, GSM6782551_1151376, GSM6782551_1151375, GSM6782552_1131075, ENSE00002508957, GSM6782552_1131068, GSM6783528_1159401, GSM6782552_1131072'}\n",
      "{'geneId': 'ENSG00000174353', 'geneName': 'STAG3L3', 'geneVersion': '18', 'geneBiotype': 'transcribed_unprocessed_pseudogene', 'strand': '-', 'chr': 'chr7', 'start': 72998027, 'end': 73003730, 'TSS': 73003730, 'TES': 72998027}\n",
      "{'geneId': 'ENSG00000290832', 'geneName': 'None', 'geneVersion': '1', 'geneBiotype': 'lncRNA', 'strand': '-', 'chr': 'chr7', 'start': 72969696, 'end': 73005922, 'TSS': 73005922, 'TES': 72969696}\n"
     ]
    }
   ],
   "source": [
    "print(total.geneDict[\"ENSG00000174353\"].dictGet())\n",
    "print(geneRef[\"ENSG00000174353\"])\n",
    "print(geneRef[\"ENSG00000290832\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'geneId': 'ENSG00000174353', 'transcriptId': 'ENST00000423834'}"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total.transcriptIdGet(\"ENST00000423834\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000423834\n",
      "ENCLB240AAXT000349247\n",
      "ENSG00000290832\n",
      "ENSG00000174353\n"
     ]
    }
   ],
   "source": [
    "print(transcriptIdRef)\n",
    "print(transcriptId)\n",
    "print(transcriptRef[transcriptIdRef][\"geneId\"])\n",
    "print(geneId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "有部分transcript, 在total中该transcript属于geneTotal, 但在基因组注释中属于geneRef\n",
    "需要将该transcript所属的gene由geneTotal映射到geneRef\n",
    "    查询geneRef是否已记录\n",
    "        若geneRefId已记录\n",
    "            geneRef中新增该transcript对象\n",
    "            geneTotal中删除该transcript对象\n",
    "        若geneRefId未记录但geneRef已记录\n",
    "            获取其映射的geneRefMapped\n",
    "            查询该transcript是否已存在于\n",
    "        若geneRef未记录\n",
    "            记录geneRef对象\n",
    "            在geneRef中新增该transcript对象\n",
    "            删除geneTotal中的transcript对象\n",
    "去除不包含transcript的gene\n",
    "重建index\n",
    "保存这种错误映射的数据到tsv中\n",
    "'''\n",
    "totalMap = {}\n",
    "for geneId, geneObject in total.geneDict.items():\n",
    "    for transcriptId, transcriptObject in geneObject.transcriptDict.items():\n",
    "        totalMap[transcriptId] = geneId\n",
    "annotationMap = {}\n",
    "for transcriptId, transcriptDict in transcriptRef.items():\n",
    "    geneId = transcriptDict[\"geneId\"]\n",
    "    annotationMap[transcriptId] = geneId\n",
    "# 筛选出映射到错误gene的transcript\n",
    "errorMap = {}  # {<transcriptId>: {\"total\": <geneId>, \"ref\": <geneId>}}\n",
    "for transcriptId, geneId in totalMap.items():\n",
    "    marker = annotationMap.get(transcriptId, None)\n",
    "    if marker is None:\n",
    "        continue\n",
    "    if geneId != marker:\n",
    "        errorMap[transcriptId] = {\"total\": geneId, \"ref\": marker}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 ENST00000508856 ENSG00000248632 ENSG00000183439\n",
      "0 ENST00000463441 ENSG00000127399 ENSG00000188707\n",
      "0 ENST00000521369 ENSG00000254615 ENSG00000164830\n",
      "0 ENST00000577661 ENSG00000254615 ENSG00000164830\n",
      "0 ENST00000430721 ENSG00000239665 ENSG00000165630\n",
      "0 ENST00000359450 ENSG00000221995 ENSG00000196535\n",
      "0 ENST00000395693 ENSG00000130489 ENSG00000284194\n",
      "0 ENST00000442171 ENSG00000228106 ENSG00000154309\n",
      "0 ENST00000435733 ENSG00000226377 ENSG00000236377\n",
      "0 ENST00000598924 ENSG00000269814 ENSG00000142235\n",
      "0 ENST00000603308 ENSG00000271127 ENSG00000206195\n",
      "0 ENST00000624499 ENSG00000279355 ENSG00000026652\n",
      "0 ENST00000568479 ENSG00000260822 ENSG00000067992\n"
     ]
    }
   ],
   "source": [
    "for transcriptId, d in errorMap.items():\n",
    "    geneIdTotal = d[\"total\"]\n",
    "    geneIdRef = d[\"ref\"]\n",
    "    if geneIdRef in total.geneDict:\n",
    "        if transcriptId in total.geneDict[geneIdRef].transcriptDict:\n",
    "            print(1, transcriptId,geneIdTotal,geneIdRef)\n",
    "        else:\n",
    "            print(0, transcriptId,geneIdTotal,geneIdRef)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENST00000508856\n",
      "ENST00000463441\n",
      "ENST00000521369\n",
      "ENST00000577661\n",
      "ENST00000430721\n",
      "ENST00000359450\n",
      "ENST00000395693\n",
      "ENST00000442171\n",
      "ENST00000435733\n",
      "ENST00000598924\n",
      "ENST00000603308\n",
      "ENST00000624499\n",
      "ENST00000568479\n"
     ]
    }
   ],
   "source": [
    "# 纠正错误映射\n",
    "for transcriptId, d in errorMap.items():\n",
    "    geneIdTotal = d[\"total\"]\n",
    "    geneIdRef = d[\"ref\"]\n",
    "    # 查询geneRef是否已记录\n",
    "    #(marker, mappedId) = total._geneCheck(geneId=geneIdRef, chr=geneRef[geneIdRef][\"chr\"], strand=geneRef[geneIdRef][\"strand\"], start=geneRef[geneIdRef][\"start\"], end=geneRef[geneIdRef][\"end\"])\n",
    "    #geneIdRef = (geneIdRef, mappedId)[marker==3]\n",
    "    if geneIdRef in total.geneDict:\n",
    "        # geneRef已记录\n",
    "        if transcriptId in total.geneDict[geneIdRef].transcriptDict:\n",
    "            # 该transcript对象已存在于geneIdRef中\n",
    "            print(\"[Warning]{id} have been existed in {r}, failed {t} --> {r}\".format(id=transcriptId, r=geneIdRef, t=geneIdTotal))\n",
    "        else:\n",
    "            # 该transcript对象未存在于geneIdRef中\n",
    "            # geneRef中新增该transcript对象\n",
    "            total.geneDict[geneIdRef].transcriptDict[transcriptId] = total.geneDict[geneIdTotal].transcriptDict[transcriptId]\n",
    "        # geneTotal中删除该transcript对象\n",
    "        total.geneDict[geneIdTotal].transcriptDict.pop(transcriptId)\n",
    "        print(transcriptId)\n",
    "    else:\n",
    "        '''# geneRef未记录\n",
    "        # 记录geneRef对象\n",
    "        # 判断geneRef是否已记录\n",
    "        total.geneAdd(status=\"KNOWN\", chr=geneRef[geneIdRef][\"chr\"], strand=geneRef[geneIdRef][\"strand\"],\n",
    "                      start=geneRef[geneIdRef][\"start\"], end=geneRef[geneIdRef][\"end\"], geneId=geneRef[geneIdRef][\"geneId\"],\n",
    "                      geneVersion=geneRef[geneIdRef][\"geneVersion\"], geneName=geneRef[geneIdRef][\"geneName\"], geneBiotype=geneRef[geneIdRef][\"geneBiotype\"])\n",
    "        # 在geneRef中新增该transcript对象\n",
    "        total.geneDict[geneIdRef].transcriptDict[transcriptId] = total.geneDict[geneIdTotal].transcriptDict[transcriptId]\n",
    "        # 删除geneTotal中的transcript对象\n",
    "        total.geneDict[geneIdTotal].transcriptDict.pop(transcriptId)'''\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'total': 'ENSG00000226287', 'ref': 'ENSG00000291085'}"
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "errorMap[\"ENST00000452055\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'GSM6783527': 2}\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'ENST00000508856'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32mf:\\OneDrive\\Master\\Project\\trans\\scripts\\10002_correct.ipynb 单元格 39\u001b[0m line \u001b[0;36m<cell line: 2>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#Y164sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mprint\u001b[39m(total\u001b[39m.\u001b[39mgeneDict[\u001b[39m\"\u001b[39m\u001b[39mENSG00000248632\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m.\u001b[39mtranscriptDict[\u001b[39m\"\u001b[39m\u001b[39mENST00000508856\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m.\u001b[39mcountsExpression)\n\u001b[1;32m----> <a href='vscode-notebook-cell:/f%3A/OneDrive/Master/Project/trans/scripts/10002_correct.ipynb#Y164sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m \u001b[39mprint\u001b[39m(total\u001b[39m.\u001b[39;49mgeneDict[\u001b[39m\"\u001b[39;49m\u001b[39mENSG00000183439\u001b[39;49m\u001b[39m\"\u001b[39;49m]\u001b[39m.\u001b[39;49mtranscriptDict[\u001b[39m\"\u001b[39;49m\u001b[39mENST00000508856\u001b[39;49m\u001b[39m\"\u001b[39;49m]\u001b[39m.\u001b[39mcountsExpression)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'ENST00000508856'"
     ]
    }
   ],
   "source": [
    "print(total.geneDict[\"ENSG00000248632\"].transcriptDict[\"ENST00000508856\"].countsExpression)\n",
    "print(total.geneDict[\"ENSG00000183439\"].transcriptDict[\"ENST00000508856\"].countsExpression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'geneId': 'ENSG00000226287', 'geneVersion': 8, 'geneName': 'TMEM191A', 'geneBiotype': 'transcribed_unprocessed_pseudogene', 'status': 'KNOWN', 'strand': '+', 'chr': 'chr22', 'start': 20701114, 'end': 20704606, 'TSS': 20701114, 'TES': 20704606, 'transcript': 'ENCLB315YZPT000326343, ENST00000452055, ENCLB315YZPT000326345, ENCLB240AAXT000313140', 'exon': 'GSM6782551_1099831, ENSE00001641633, GSM6782551_1099833, ENSE00001645121, GSM6782552_1083956, ENSE00001788859, ENSE00003924989, ENSE00001679851, ENSE00001706912, ENSE00001712385, ENSE00001762770'}\n",
      "{'transcriptId': 'ENST00000452055', 'transcriptVersion': 5, 'transcriptName': 'TMEM191A-209', 'transcriptBiotype': 'retained_intron', 'status': 'KNOWN', 'start': 20702921, 'end': 20704606, 'TSS': 20702921, 'TES': 20704606, 'exonList': 'ENSE00001641633, ENSE00001645121, ENSE00001679851, ENSE00001706912, ENSE00001712385, ENSE00001762770'}\n",
      "{'geneId': 'ENSG00000226287', 'geneVersion': 8, 'geneName': 'TMEM191A', 'geneBiotype': 'transcribed_unprocessed_pseudogene', 'status': 'KNOWN', 'strand': '+', 'chr': 'chr22', 'start': 20701114, 'end': 20704606, 'TSS': 20701114, 'TES': 20704606, 'transcript': 'ENCLB315YZPT000326343, ENST00000452055, ENCLB315YZPT000326345, ENCLB240AAXT000313140', 'exon': 'GSM6782551_1099831, ENSE00001641633, GSM6782551_1099833, ENSE00001645121, GSM6782552_1083956, ENSE00001788859, ENSE00003924989, ENSE00001679851, ENSE00001706912, ENSE00001712385, ENSE00001762770'}\n",
      "{'transcriptId': 'ENST00000452055', 'transcriptVersion': 5, 'transcriptName': 'TMEM191A-209', 'transcriptBiotype': 'retained_intron', 'status': 'KNOWN', 'start': 20702921, 'end': 20704606, 'TSS': 20702921, 'TES': 20704606, 'exonList': 'ENSE00001641633, ENSE00001645121, ENSE00001679851, ENSE00001706912, ENSE00001712385, ENSE00001762770'}\n"
     ]
    }
   ],
   "source": [
    "print(total.geneDict[\"ENSG00000226287\"].dictGet())\n",
    "print(total.geneDict[\"ENSG00000226287\"].transcriptDict[\"ENST00000452055\"].dictGet())\n",
    "print(total.geneDict[\"ENSG00000226287\"].dictGet())\n",
    "print(total.geneDict[\"ENSG00000226287\"].transcriptDict[\"ENST00000452055\"].dictGet())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "暂存部分"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "debugFile = \"F:/OneDrive/Master/Project/trans/data/raw_data/GSM6782551/GSM6782551_ENCFF856FNN_transcriptome_annotations_GRCh38.gtf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 二级函数\n",
    "def findIdAdvanced(index, chr, TSS, TES, rangeTSS, rangeTES, allowType):\n",
    "    '''\n",
    "    input:\n",
    "        index, dict\n",
    "        chr, str\n",
    "        TSS, int\n",
    "        TES, int\n",
    "        rangeTSS, int\n",
    "        rangeTES, int\n",
    "        allowType, str, \"gene\" or transcript\n",
    "    change:\n",
    "        查询index中, 与TSS或TES上下游分别相距rangeTSS或rangeTES的id\n",
    "    return:\n",
    "        tuple, (id, markerTSS, markerTES)\n",
    "    '''\n",
    "    index = index = transcriptRef\n",
    "    chr = chr\n",
    "    TSS = TSS\n",
    "    TES = TES\n",
    "    rangeTSS = rangeTSS = RANGETSS\n",
    "    rangeTES = rangeTES = RANGETES\n",
    "\n",
    "    a = findId(index=index, chr=chr, location=location, rangeMax=rangeMax, orientation='-', allowType=allowType)\n",
    "    return a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcriptSet = set()\n",
    "for geneObject in total.geneDict.values():\n",
    "    for transcriptId, transcriptObject in geneObject.transcriptDict.items():\n",
    "        if transcriptObject.status == \"NOVEL\":\n",
    "            transcriptSet.add(transcriptId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 根据参考基因组注释数据对start及end进行校正\n",
    "for geneObject in total.geneDict.values():\n",
    "    for transcriptObject in geneObject.transcriptDict.values():\n",
    "        strand = transcriptObject.strand\n",
    "        start = transcriptObject.start\n",
    "        end = transcriptObject.end\n",
    "        (TSS, TES) = ((end, start), (start, end))[strand is '+']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "end\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = set()\n",
    "for exonObject in total.exonDict.values():\n",
    "    status = exonObject.status\n",
    "    a.add(status)\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "debug = {}\n",
    "for exonObject in total.exonDict.values():\n",
    "    status = exonObject.status\n",
    "    if status == \"KNOWN\":\n",
    "        debug[\"KNOWN\"] = debug.setdefault(\"KNOWN\", 0) + 1\n",
    "    elif status == \"NOVEL\":\n",
    "        debug[\"NOVEL\"] = debug.setdefault(\"NOVEL\", 0) + 1\n",
    "    elif status == \"CORRECTED-TSS-TES\":\n",
    "        debug[\"CORRECTED-TSS-TES\"] = debug.setdefault(\"CORRECTED-TSS-TES\", 0) + 1\n",
    "    elif status == \"CORRECTED-TES-TSS\":\n",
    "        debug[\"CORRECTED-TSS-TES\"] = debug.setdefault(\"CORRECTED-TSS-TES\", 0) + 1\n",
    "    elif status == \"CORRECTED-TSS\":\n",
    "        debug[\"CORRECTED-TSS\"] = debug.setdefault(\"CORRECTED-TSS\", 0) + 1\n",
    "    elif status == \"CORRECTED-TES\":\n",
    "        debug[\"CORRECTED-TES\"] = debug.setdefault(\"CORRECTED-TES\", 0) + 1\n",
    "    elif status == \"NOVEL-TSS\":\n",
    "        debug[\"NOVEL-TSS\"] = debug.setdefault(\"NOVEL-TSS\", 0) + 1\n",
    "    elif status == \"NOVEL-TES\":\n",
    "        debug[\"NOVEL-TES\"] = debug.setdefault(\"NOVEL-TES\", 0) + 1\n",
    "    elif status == \"NOVEL-TSS-TES\":\n",
    "        debug[\"NOVEL-TSS-TES\"] = debug.setdefault(\"NOVEL-TSS-TES\", 0) + 1\n",
    "    elif status == \"NOVEL-TES-TSS\":\n",
    "        debug[\"NOVEL-TSS-TES\"] = debug.setdefault(\"NOVEL-TSS-TES\", 0) + 1\n",
    "    else:\n",
    "        print(status)\n",
    "debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "trace = go.Pie(labels=[\"novel exon\", \"start-end corrected\", \"start corrected\", \"end corrected\", \"known exon\"],\n",
    "\t   values=[67087, 13543, 20126, 23812, 190845],\n",
    "\t   marker={\"colors\": [\"#505168\", \"#2A7F62\", \"#75DDDD\", \"#E8E1EF\", \"lightgrey\"],\n",
    "\t\t\t   \"line\": {\"color\": \"white\",\n",
    "\t\t\t\t\t\t\"width\": 1}},\n",
    "\t   sort=False,\n",
    "\t   textinfo=\"value\",\n",
    "\t   textfont_size=12,\n",
    "\t   insidetextorientation='auto',\n",
    "\t   hole=0,\n",
    "\t   pull=[0.1, 0, 0, 0, 0],\n",
    "\t   )\n",
    "fig = go.Figure(trace)\n",
    "\n",
    "layout={\"width\": 500, \"height\": 500,\n",
    "\t\t\"margin\": {'l':25, 'r':25, 't':25, 'b':25},\n",
    "        \"title\": {\"text\": \"Correction of exon<br>Range: 5\",\n",
    "\t\t\t\t  \"font\": {\"family\": \"Arial\",\n",
    "\t\t\t\t\t\t   \"color\": \"black\",\n",
    "\t\t\t\t\t\t   \"size\": 16},\n",
    "\t\t\t\t   \"x\": 0.4},\n",
    "        \"font\": {\"family\": \"Arial\",  # 作用所有地方\n",
    "\t\t\t\t \"color\": \"black\",\n",
    "\t\t\t\t \"size\": 12,\n",
    "\t\t\t\t },\n",
    "        \"paper_bgcolor\": \"white\",\n",
    "        \"plot_bgcolor\": \"white\",\n",
    "        }\n",
    "fig = fig.update_layout(layout)\n",
    "'''\n",
    "将最大校正范围设定在5bp时, 在标注为novel的74504个exon中, 有3558个exon可以完全映射到已注释exon上, 有16229个exon的start附近存在已注释exon, 有10967个exon的end附近存在已注释exon。\n",
    "已根据其映射情况修改了exon的start或end的位置\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig.show()\n",
    "fig.write_image(\"5exonBeforeIntegrate.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from plotly.subplots import make_subplots\n",
    "\n",
    "# 创建子图，两个 Y 轴\n",
    "fig = make_subplots(specs=[[{\"secondary_y\": True}]])\n",
    "\n",
    "trace1 = go.Scatter(x=[1,2,3,4,5,6,7,8,9,10,\n",
    "                       11,12,13,14,15,16,17,18,19,20],\n",
    "                   y=[213,934,1826,3197,3558,3867,4071,4233,4416,4559,\n",
    "                      4677,4825,4914,5005,5138,5225,5308,5437,5546,5630],\n",
    "                   mode=\"lines+markers\",\n",
    "                   name=\"start-end exon\",\n",
    "                   line={\"color\": \"#2A7F62\"})\n",
    "trace2 = go.Scatter(x=[1,2,3,4,5,6,7,8,9,10,\n",
    "                       11,12,13,14,15,16,17,18,19,20],\n",
    "                   y=[44356,44197,44197,44197,43750,43665,43603,43549,43549,43476,\n",
    "                      43445,43410,43410,43346,43327,43298,43278,43257,43232,43206],\n",
    "                   mode=\"lines+markers\",\n",
    "                   name=\"novel exon\",\n",
    "                   line={\"color\": \"#505168\"})\n",
    "\n",
    "# 添加第一个 Y 轴的数据\n",
    "fig.add_trace(trace1, secondary_y=False)\n",
    "# 添加第二个 Y 轴的数据\n",
    "fig.add_trace(trace2, secondary_y=True)\n",
    "\n",
    "# 设置图形布局\n",
    "layout = {\"width\": 700, \"height\": 400,\n",
    "\t\t    \"margin\": {'l':25, 'r':25, 't':25, 'b':25},\n",
    "          \"title\": {\"text\": \"\",\n",
    "\t\t\t\t  \"font\": {\"family\": \"Arial\",\n",
    "\t\t\t\t\t\t   \"color\": \"black\",\n",
    "\t\t\t\t\t\t   \"size\": 16},\n",
    "\t\t\t\t   \"x\": 0.4},\n",
    "        \"font\": {\"family\": \"Arial\",  # 作用所有地方\n",
    "\t\t\t\t \"color\": \"black\",\n",
    "\t\t\t\t \"size\": 12,\n",
    "\t\t\t\t },\n",
    "         \"xaxis\": {\"showline\": True,\n",
    "                   \"linecolor\": \"black\",\n",
    "                   \"linewidth\": 1,\n",
    "                   \"showticklabels\": True,\n",
    "                   \"dtick\": 2},\n",
    "         \"yaxis\": {\"showline\": True,\n",
    "                   \"linecolor\": \"black\",\n",
    "                   \"linewidth\": 1,\n",
    "                   \"showticklabels\": True,},\n",
    "         \"yaxis2\": {\"showline\": True,\n",
    "                   \"linecolor\": \"black\",\n",
    "                   \"linewidth\": 1,\n",
    "                   \"showticklabels\": True,},\n",
    "        \"paper_bgcolor\": \"white\",\n",
    "        \"plot_bgcolor\": \"white\",}\n",
    "fig = fig.update_layout(layout)\n",
    "fig = fig.update_layout(title_text=\"Correction of exons in different ranges\",\n",
    "                  xaxis_title=\"Range\",\n",
    "                  yaxis_title=\"Counts of start-end exon\",\n",
    "                  yaxis2_title=\"Counts of novel exon\")\n",
    "\n",
    "# 显示图形\n",
    "#fig.write_image(\"line.png\")\n",
    "fig.show()\n",
    "'''\n",
    "之所以选择exon的最大校正范围为5bp, 是因为已经分析了在最大校正范围为1~20时的校正情况。\n",
    "当范围设定在5bp时, novel exon的减少速度趋于稳定, 且在所有exon中双端可映射到已注释exon的数量的增加速度趋于稳定。\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "不根据参考基因组注释进行校正:\n",
    "\n",
    "校正前total exon num: 323363\n",
    "校正前novel exon num: 133311\n",
    "校正前total transcript num: 213387\n",
    "校正前novel transcript num: 185981\n",
    "mapped 6095\n",
    "start-end 6503\n",
    "start 22997\n",
    "end 19860\n",
    "校正后total exon num: 309350\n",
    "校正后novel exon num: 72014\n",
    "校正后total transcript num: 211068\n",
    "校正后novel transcript num: 183662\n",
    "\n",
    "根据参考基因组注释进行校正:\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\tnovel exon\tstart-end exon\tstart exon\tend exon\tremained exon\n",
    "1\t74504\t213\t17626\t12309\t44356\n",
    "2\t74504\t934\t17311\t12062\t44197\n",
    "3\t74504\t1826\t16935\t11713\t44197\n",
    "4\t74504\t3197\t16355\t11107\t44197\n",
    "5\t74504\t3558\t16229\t10967\t43750\n",
    "6\t74504\t3867\t16119\t10853\t43665\n",
    "7\t74504\t4071\t16029\t10801\t43603\n",
    "8\t74504\t4233\t15961\t10761\t43549\n",
    "9\t74504\t4416\t15877\t10704\t43549\n",
    "10\t74504\t4559\t15802\t10667\t43476\n",
    "11\t74504\t4677\t15753\t10629\t43445\n",
    "12\t74504\t4825\t15694\t10575\t43410\n",
    "13\t74504\t4914\t15651\t10557\t43410\n",
    "14\t74504\t5005\t15621\t10532\t43346\n",
    "15\t74504\t5138\t15543\t10496\t43327\n",
    "16\t74504\t5225\t15501\t10480\t43298\n",
    "17\t74504\t5308\t15452\t10466\t43278\n",
    "18\t74504\t5437\t15395\t10415\t43257\n",
    "19\t74504\t5546\t15337\t10389\t43232\n",
    "20\t74504\t5630\t15297\t10371\t43206'''\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
